{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment in GWAS, TWAS, and DE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:08:03.760403Z",
     "iopub.status.busy": "2021-09-08T01:08:03.754729Z",
     "iopub.status.idle": "2021-09-08T01:08:04.661325Z",
     "shell.execute_reply": "2021-09-08T01:08:04.659350Z"
    }
   },
   "outputs": [],
   "source": [
    "import functools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import collections as cx\n",
    "from pybiomart import Dataset\n",
    "from gtfparse import read_gtf\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "# GO analysis\n",
    "from goatools.base import download_go_basic_obo\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:08:04.672728Z",
     "iopub.status.busy": "2021-09-08T01:08:04.671185Z",
     "iopub.status.idle": "2021-09-08T01:08:04.676434Z",
     "shell.execute_reply": "2021-09-08T01:08:04.674820Z"
    }
   },
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_gtf_genes_df():\n",
    "    gtf_df = read_gtf(\"/ceph/genome/human/gencode25/gtf.CHR/_m/gencode.v25.annotation.gtf\")\n",
    "    return gtf_df[gtf_df[\"feature\"] == \"gene\"][['gene_id', 'gene_name']]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_wgcna_modules():\n",
    "    return pd.read_csv(\"../../_m/modules.csv\", index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_database():\n",
    "    dataset = Dataset(name=\"hsapiens_gene_ensembl\", \n",
    "                      host=\"http://www.ensembl.org\",\n",
    "                      use_cache=True)\n",
    "    db = dataset.query(attributes=[\"ensembl_gene_id\", \n",
    "                                   \"external_gene_name\", \n",
    "                                   \"entrezgene_id\"], \n",
    "                       use_attr_names=True).dropna(subset=['entrezgene_id'])\n",
    "    return db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:08:04.696207Z",
     "iopub.status.busy": "2021-09-08T01:08:04.688157Z",
     "iopub.status.idle": "2021-09-08T01:08:04.701742Z",
     "shell.execute_reply": "2021-09-08T01:08:04.700274Z"
    }
   },
   "outputs": [],
   "source": [
    "def fet(a, b, u):\n",
    "    # a, b, u are sets\n",
    "    # u is the universe\n",
    "    yes_a = u.intersection(a)\n",
    "    yes_b = u.intersection(b)\n",
    "    no_a = u - a\n",
    "    no_b = u - b\n",
    "    m = [[len(yes_a.intersection(yes_b)), len(no_a.intersection(yes_b)) ], \n",
    "         [len(yes_a.intersection(no_b)), len(no_a.intersection(no_b))]]\n",
    "    return fisher_exact(m)\n",
    "\n",
    "\n",
    "def enrichment_rows():\n",
    "    mod = get_wgcna_modules().module.unique()\n",
    "    u = set(get_wgcna_modules().index) \n",
    "    for ii in range(len(mod)): # for each module\n",
    "        a = set(get_wgcna_modules()[(get_wgcna_modules().module) == mod[ii]].index)\n",
    "        yield (mod[ii],\n",
    "               len(a),\n",
    "               *fet(a, gwas_genes, u),\n",
    "               *fet(a, twas_genes, u),\n",
    "               *fet(a, de_genes, u),\n",
    "               )\n",
    "        \n",
    "\n",
    "def enrichment_rows_nomhc():\n",
    "    mod = get_wgcna_modules().module.unique()\n",
    "    u = set(get_wgcna_modules().index) - mhc_genes\n",
    "    for ii in range(len(mod)): # for each module\n",
    "        a = set(get_wgcna_modules()[(get_wgcna_modules().module) == mod[ii]].index) - mhc_genes\n",
    "        yield (mod[ii],\n",
    "               len(a),\n",
    "               *fet(a, gwas_genes - mhc_genes, u),\n",
    "               *fet(a, twas_genes - mhc_genes, u),\n",
    "               *fet(a, de_genes - mhc_genes, u),\n",
    "              )\n",
    "        \n",
    "\n",
    "def convert2entrez(mod):\n",
    "    df = get_wgcna_modules()[(get_wgcna_modules().module) == mod].copy()\n",
    "    df[\"ensemblID\"] = df.index.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "    return df.merge(get_database(), left_on='ensemblID', \n",
    "                    right_on='ensembl_gene_id')\n",
    "\n",
    "\n",
    "def obo_annotation(alpha=0.05):\n",
    "    # database annotation\n",
    "    fn_obo = download_go_basic_obo()\n",
    "    fn_gene2go = download_ncbi_associations() # must be gunzip to work\n",
    "    obodag = GODag(fn_obo) # downloads most up-to-date\n",
    "    anno_hs = Gene2GoReader(fn_gene2go, taxids=[9606])\n",
    "    # get associations\n",
    "    ns2assoc = anno_hs.get_ns2assc()\n",
    "    for nspc, id2gos in ns2assoc.items():\n",
    "        print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "    goeaobj = GOEnrichmentStudyNS(\n",
    "        get_database()['entrezgene_id'], # List of human genes with entrez IDs\n",
    "        ns2assoc, # geneid/GO associations\n",
    "        obodag, # Ontologies\n",
    "        propagate_counts = False,\n",
    "        alpha = alpha, # default significance cut-off\n",
    "        methods = ['fdr_bh'])\n",
    "    return goeaobj\n",
    "\n",
    "\n",
    "def run_goea(mod):\n",
    "    df = convert2entrez(mod)\n",
    "    geneids_study = {z[0]:z[1] for z in zip(df['entrezgene_id'], df['external_gene_name'])}\n",
    "    goeaobj = obo_annotation()\n",
    "    goea_results_all = goeaobj.run_study(geneids_study)\n",
    "    goea_results_sig = [r for r in goea_results_all if r.p_fdr_bh < 0.05]\n",
    "    ctr = cx.Counter([r.NS for r in goea_results_sig])\n",
    "    print('Significant results[{TOTAL}] = {BP} BP + {MF} MF + {CC} CC'.format(\n",
    "        TOTAL=len(goea_results_sig),\n",
    "        BP=ctr['BP'],  # biological_process\n",
    "        MF=ctr['MF'],  # molecular_function\n",
    "        CC=ctr['CC'])) # cellular_component\n",
    "    goeaobj.wr_xlsx(\"GO_analysis_module_%s.xlsx\" % mod, goea_results_sig)\n",
    "    goeaobj.wr_txt(\"GO_analysis_module_%s.txt\" % mod, goea_results_sig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:08:04.711401Z",
     "iopub.status.busy": "2021-09-08T01:08:04.708822Z",
     "iopub.status.idle": "2021-09-08T01:09:37.583039Z",
     "shell.execute_reply": "2021-09-08T01:09:37.584114Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_status', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_status', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id gene_name\n",
       "0   ENSG00000223972.5   DDX11L1\n",
       "12  ENSG00000227232.5    WASH7P"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf = get_gtf_genes_df()\n",
    "gtf.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GWAS, TWAS and DE enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load DE, TWAS, and GWAS genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:37.593405Z",
     "iopub.status.busy": "2021-09-08T01:09:37.592130Z",
     "iopub.status.idle": "2021-09-08T01:09:37.609197Z",
     "shell.execute_reply": "2021-09-08T01:09:37.608355Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2701"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_genes = set(pd.read_csv('../../../differential_expression/_m/genes/diffExpr_szVctl_FDR05.txt',\n",
    "                           sep='\\t', usecols=[0], index_col=0).index)\n",
    "len(de_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:37.616944Z",
     "iopub.status.busy": "2021-09-08T01:09:37.615773Z",
     "iopub.status.idle": "2021-09-08T01:09:37.662469Z",
     "shell.execute_reply": "2021-09-08T01:09:37.663489Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_genes = set(pd.read_csv('/ceph/projects/v3_phase3_paper/inputs/gwas/PGC2_CLOZUK/table_s3/hg38/genes/_m/gwas_genes.csv')['gene_id'])\n",
    "len(gwas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:37.673671Z",
     "iopub.status.busy": "2021-09-08T01:09:37.672416Z",
     "iopub.status.idle": "2021-09-08T01:09:37.684337Z",
     "shell.execute_reply": "2021-09-08T01:09:37.683601Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "383"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhc_genes = set(pd.read_csv('/ceph/projects/v4_phase3_paper/inputs/counts/mhc_region_genes/_m/mhc_genes.csv')['gene_id'])\n",
    "len(mhc_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:37.708086Z",
     "iopub.status.busy": "2021-09-08T01:09:37.700488Z",
     "iopub.status.idle": "2021-09-08T01:09:38.460320Z",
     "shell.execute_reply": "2021-09-08T01:09:38.459295Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:numexpr.utils:Note: NumExpr detected 60 cores but \"NUMEXPR_MAX_THREADS\" not set, so enforcing safe limit of 8.\n",
      "INFO:numexpr.utils:NumExpr defaulting to 8 threads.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = pd.read_csv(\"/ceph/projects/v4_phase3_paper/inputs/counts/text_files_counts/_m/caudate/gene.bed\", \n",
    "                    sep='\\t', index_col=0)\n",
    "annot[\"Feature\"] = annot.gene_id.str.replace(\"\\\\..*\", \"\", regex=True)\n",
    "twas = pd.read_csv(\"../../../twas/feature_comparison/manuscript_supp_data/_m/\"+\\\n",
    "                   \"BrainSeq_Phase3_Caudate_TWAS_associations_allFeatures.txt.gz\", sep='\\t')\n",
    "twas = twas[(twas[\"FDR\"] < 0.05) & (twas[\"Type\"] == \"Gene\")].merge(annot, on=\"Feature\")\n",
    "twas_genes = set(twas['gene_id'])\n",
    "len(twas_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load WGCNA module"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:38.468393Z",
     "iopub.status.busy": "2021-09-08T01:09:38.467488Z",
     "iopub.status.idle": "2021-09-08T01:09:38.610058Z",
     "shell.execute_reply": "2021-09-08T01:09:38.609081Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12.0</th>\n",
       "      <td>grey</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25.0</th>\n",
       "      <td>yellow</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      module            gene_id  gene_name\n",
       "12.0    grey  ENSG00000227232.5     WASH7P\n",
       "25.0  yellow  ENSG00000278267.1  MIR6859-1"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df = get_wgcna_modules().merge(gtf, left_index=True, right_on=\"gene_id\", how=\"left\")\n",
    "wgcna_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:38.629104Z",
     "iopub.status.busy": "2021-09-08T01:09:38.627893Z",
     "iopub.status.idle": "2021-09-08T01:09:38.644308Z",
     "shell.execute_reply": "2021-09-08T01:09:38.645073Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113412884-113414374(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113412884-113415420(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NaN</th>\n",
       "      <td>lightgreen</td>\n",
       "      <td>chr11:113414462-113415420(-)</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         module                       gene_id gene_name\n",
       "NaN  lightgreen  chr11:113412884-113414374(-)       NaN\n",
       "NaN  lightgreen  chr11:113412884-113415420(-)       NaN\n",
       "NaN  lightgreen  chr11:113414462-113415420(-)       NaN"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df[(wgcna_df.gene_id.str.startswith(\"chr\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:38.669240Z",
     "iopub.status.busy": "2021-09-08T01:09:38.661351Z",
     "iopub.status.idle": "2021-09-08T01:09:38.681619Z",
     "shell.execute_reply": "2021-09-08T01:09:38.680733Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1445279.0</th>\n",
       "      <td>lightcyan</td>\n",
       "      <td>ENSG00000149295.13</td>\n",
       "      <td>DRD2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              module             gene_id gene_name\n",
       "1445279.0  lightcyan  ENSG00000149295.13      DRD2"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df[(wgcna_df.gene_name == 'DRD2')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:38.701930Z",
     "iopub.status.busy": "2021-09-08T01:09:38.694550Z",
     "iopub.status.idle": "2021-09-08T01:09:38.706971Z",
     "shell.execute_reply": "2021-09-08T01:09:38.706032Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>module</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1914770.0</th>\n",
       "      <td>yellow</td>\n",
       "      <td>ENSG00000099381.16</td>\n",
       "      <td>SETD1A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           module             gene_id gene_name\n",
       "1914770.0  yellow  ENSG00000099381.16    SETD1A"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wgcna_df[(wgcna_df.gene_name == 'SETD1A')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:38.727580Z",
     "iopub.status.busy": "2021-09-08T01:09:38.725213Z",
     "iopub.status.idle": "2021-09-08T01:09:40.393890Z",
     "shell.execute_reply": "2021-09-08T01:09:40.392073Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>module_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>grey</th>\n",
       "      <td>3940</td>\n",
       "      <td>0.892699</td>\n",
       "      <td>0.214026</td>\n",
       "      <td>0.523174</td>\n",
       "      <td>1.082137</td>\n",
       "      <td>0.439890</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.676330</td>\n",
       "      <td>2.572857e-11</td>\n",
       "      <td>8.086121e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>1587</td>\n",
       "      <td>1.127312</td>\n",
       "      <td>0.308776</td>\n",
       "      <td>0.617552</td>\n",
       "      <td>1.111805</td>\n",
       "      <td>0.444625</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.567600</td>\n",
       "      <td>1.189145e-09</td>\n",
       "      <td>2.906798e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>turquoise</th>\n",
       "      <td>3334</td>\n",
       "      <td>0.864652</td>\n",
       "      <td>0.142167</td>\n",
       "      <td>0.390959</td>\n",
       "      <td>0.877289</td>\n",
       "      <td>0.270678</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.303677</td>\n",
       "      <td>3.095585e-57</td>\n",
       "      <td>3.405144e-56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyan</th>\n",
       "      <td>375</td>\n",
       "      <td>1.172909</td>\n",
       "      <td>0.522217</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.983961</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.867858</td>\n",
       "      <td>7.274883e-06</td>\n",
       "      <td>1.150205e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>1568</td>\n",
       "      <td>1.245199</td>\n",
       "      <td>0.063396</td>\n",
       "      <td>0.232452</td>\n",
       "      <td>0.889608</td>\n",
       "      <td>0.537755</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.379828</td>\n",
       "      <td>4.231908e-21</td>\n",
       "      <td>1.862039e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brown</th>\n",
       "      <td>2253</td>\n",
       "      <td>0.767852</td>\n",
       "      <td>0.025971</td>\n",
       "      <td>0.142843</td>\n",
       "      <td>0.864100</td>\n",
       "      <td>0.327368</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>2.935529</td>\n",
       "      <td>7.804697e-79</td>\n",
       "      <td>1.717033e-77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black</th>\n",
       "      <td>1022</td>\n",
       "      <td>1.108971</td>\n",
       "      <td>0.480575</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>1.127104</td>\n",
       "      <td>0.509117</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>2.301877</td>\n",
       "      <td>1.466545e-23</td>\n",
       "      <td>8.065996e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>midnightblue</th>\n",
       "      <td>370</td>\n",
       "      <td>1.463223</td>\n",
       "      <td>0.093245</td>\n",
       "      <td>0.293057</td>\n",
       "      <td>1.189579</td>\n",
       "      <td>0.535008</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>1.491874</td>\n",
       "      <td>7.027371e-03</td>\n",
       "      <td>8.136956e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightyellow</th>\n",
       "      <td>226</td>\n",
       "      <td>0.908099</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.584393</td>\n",
       "      <td>0.426782</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>1.826881</td>\n",
       "      <td>8.007519e-04</td>\n",
       "      <td>9.786968e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tan</th>\n",
       "      <td>411</td>\n",
       "      <td>1.003817</td>\n",
       "      <td>0.903400</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>1.415578</td>\n",
       "      <td>0.184201</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>1.956593</td>\n",
       "      <td>3.824870e-07</td>\n",
       "      <td>7.649741e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>royalblue</th>\n",
       "      <td>217</td>\n",
       "      <td>0.728610</td>\n",
       "      <td>0.505137</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.609334</td>\n",
       "      <td>0.422714</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.844996</td>\n",
       "      <td>5.253517e-01</td>\n",
       "      <td>5.253517e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>magenta</th>\n",
       "      <td>679</td>\n",
       "      <td>0.623757</td>\n",
       "      <td>0.044510</td>\n",
       "      <td>0.195843</td>\n",
       "      <td>0.528732</td>\n",
       "      <td>0.037884</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.449134</td>\n",
       "      <td>1.162519e-07</td>\n",
       "      <td>2.557542e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>darkred</th>\n",
       "      <td>204</td>\n",
       "      <td>1.879159</td>\n",
       "      <td>0.023150</td>\n",
       "      <td>0.142843</td>\n",
       "      <td>1.688871</td>\n",
       "      <td>0.140537</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>2.019585</td>\n",
       "      <td>1.683286e-04</td>\n",
       "      <td>2.178370e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>793</td>\n",
       "      <td>1.012345</td>\n",
       "      <td>0.929431</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>1.293422</td>\n",
       "      <td>0.167399</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>3.497723</td>\n",
       "      <td>1.153651e-46</td>\n",
       "      <td>8.460107e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>blue</th>\n",
       "      <td>2532</td>\n",
       "      <td>1.351950</td>\n",
       "      <td>0.001631</td>\n",
       "      <td>0.035874</td>\n",
       "      <td>1.165965</td>\n",
       "      <td>0.192978</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.729929</td>\n",
       "      <td>7.319488e-06</td>\n",
       "      <td>1.150205e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightgreen</th>\n",
       "      <td>227</td>\n",
       "      <td>0.193191</td>\n",
       "      <td>0.004850</td>\n",
       "      <td>0.053345</td>\n",
       "      <td>0.731567</td>\n",
       "      <td>0.692431</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>2.029721</td>\n",
       "      <td>4.559218e-05</td>\n",
       "      <td>6.686852e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightcyan</th>\n",
       "      <td>339</td>\n",
       "      <td>1.014813</td>\n",
       "      <td>0.893464</td>\n",
       "      <td>0.973690</td>\n",
       "      <td>0.784640</td>\n",
       "      <td>0.628696</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>2.727243</td>\n",
       "      <td>1.836597e-13</td>\n",
       "      <td>6.734189e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>484</td>\n",
       "      <td>1.146919</td>\n",
       "      <td>0.499134</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.611986</td>\n",
       "      <td>0.174787</td>\n",
       "      <td>0.707588</td>\n",
       "      <td>0.437539</td>\n",
       "      <td>4.324439e-06</td>\n",
       "      <td>7.928138e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>red</th>\n",
       "      <td>1260</td>\n",
       "      <td>0.841721</td>\n",
       "      <td>0.287101</td>\n",
       "      <td>0.617552</td>\n",
       "      <td>1.072040</td>\n",
       "      <td>0.669600</td>\n",
       "      <td>0.801762</td>\n",
       "      <td>0.683561</td>\n",
       "      <td>1.292860e-04</td>\n",
       "      <td>1.777682e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>grey60</th>\n",
       "      <td>302</td>\n",
       "      <td>1.229691</td>\n",
       "      <td>0.394382</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>1.234897</td>\n",
       "      <td>0.492128</td>\n",
       "      <td>0.739413</td>\n",
       "      <td>0.150158</td>\n",
       "      <td>3.957232e-10</td>\n",
       "      <td>1.088239e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salmon</th>\n",
       "      <td>398</td>\n",
       "      <td>1.162782</td>\n",
       "      <td>0.534400</td>\n",
       "      <td>0.691577</td>\n",
       "      <td>0.837017</td>\n",
       "      <td>0.764880</td>\n",
       "      <td>0.813108</td>\n",
       "      <td>1.393123</td>\n",
       "      <td>2.245696e-02</td>\n",
       "      <td>2.470266e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>greenyellow</th>\n",
       "      <td>440</td>\n",
       "      <td>0.878693</td>\n",
       "      <td>0.723219</td>\n",
       "      <td>0.883934</td>\n",
       "      <td>1.071803</td>\n",
       "      <td>0.776149</td>\n",
       "      <td>0.813108</td>\n",
       "      <td>0.767239</td>\n",
       "      <td>1.164459e-01</td>\n",
       "      <td>1.219909e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_genes   gwas_or    gwas_p  gwas_fdr_bh   twas_or    twas_p  \\\n",
       "module_id                                                                    \n",
       "grey             3940  0.892699  0.214026     0.523174  1.082137  0.439890   \n",
       "yellow           1587  1.127312  0.308776     0.617552  1.111805  0.444625   \n",
       "turquoise        3334  0.864652  0.142167     0.390959  0.877289  0.270678   \n",
       "cyan              375  1.172909  0.522217     0.691577  0.983961  1.000000   \n",
       "green            1568  1.245199  0.063396     0.232452  0.889608  0.537755   \n",
       "brown            2253  0.767852  0.025971     0.142843  0.864100  0.327368   \n",
       "black            1022  1.108971  0.480575     0.691577  1.127104  0.509117   \n",
       "midnightblue      370  1.463223  0.093245     0.293057  1.189579  0.535008   \n",
       "lightyellow       226  0.908099  1.000000     1.000000  0.584393  0.426782   \n",
       "tan               411  1.003817  0.903400     0.973690  1.415578  0.184201   \n",
       "royalblue         217  0.728610  0.505137     0.691577  0.609334  0.422714   \n",
       "magenta           679  0.623757  0.044510     0.195843  0.528732  0.037884   \n",
       "darkred           204  1.879159  0.023150     0.142843  1.688871  0.140537   \n",
       "pink              793  1.012345  0.929431     0.973690  1.293422  0.167399   \n",
       "blue             2532  1.351950  0.001631     0.035874  1.165965  0.192978   \n",
       "lightgreen        227  0.193191  0.004850     0.053345  0.731567  0.692431   \n",
       "lightcyan         339  1.014813  0.893464     0.973690  0.784640  0.628696   \n",
       "purple            484  1.146919  0.499134     0.691577  0.611986  0.174787   \n",
       "red              1260  0.841721  0.287101     0.617552  1.072040  0.669600   \n",
       "grey60            302  1.229691  0.394382     0.691577  1.234897  0.492128   \n",
       "salmon            398  1.162782  0.534400     0.691577  0.837017  0.764880   \n",
       "greenyellow       440  0.878693  0.723219     0.883934  1.071803  0.776149   \n",
       "\n",
       "              twas_fdr_bh     de_or          de_p     de_fdr_bh  \n",
       "module_id                                                        \n",
       "grey             0.739413  0.676330  2.572857e-11  8.086121e-11  \n",
       "yellow           0.739413  0.567600  1.189145e-09  2.906798e-09  \n",
       "turquoise        0.739413  0.303677  3.095585e-57  3.405144e-56  \n",
       "cyan             1.000000  1.867858  7.274883e-06  1.150205e-05  \n",
       "green            0.739413  0.379828  4.231908e-21  1.862039e-20  \n",
       "brown            0.739413  2.935529  7.804697e-79  1.717033e-77  \n",
       "black            0.739413  2.301877  1.466545e-23  8.065996e-23  \n",
       "midnightblue     0.739413  1.491874  7.027371e-03  8.136956e-03  \n",
       "lightyellow      0.739413  1.826881  8.007519e-04  9.786968e-04  \n",
       "tan              0.707588  1.956593  3.824870e-07  7.649741e-07  \n",
       "royalblue        0.739413  0.844996  5.253517e-01  5.253517e-01  \n",
       "magenta          0.707588  0.449134  1.162519e-07  2.557542e-07  \n",
       "darkred          0.707588  2.019585  1.683286e-04  2.178370e-04  \n",
       "pink             0.707588  3.497723  1.153651e-46  8.460107e-46  \n",
       "blue             0.707588  0.729929  7.319488e-06  1.150205e-05  \n",
       "lightgreen       0.801762  2.029721  4.559218e-05  6.686852e-05  \n",
       "lightcyan        0.801762  2.727243  1.836597e-13  6.734189e-13  \n",
       "purple           0.707588  0.437539  4.324439e-06  7.928138e-06  \n",
       "red              0.801762  0.683561  1.292860e-04  1.777682e-04  \n",
       "grey60           0.739413  0.150158  3.957232e-10  1.088239e-09  \n",
       "salmon           0.813108  1.393123  2.245696e-02  2.470266e-02  \n",
       "greenyellow      0.813108  0.767239  1.164459e-01  1.219909e-01  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edf1 = pd.DataFrame.from_records(enrichment_rows(), \n",
    "                                 columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', \n",
    "                                          'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                 index='module_id')\n",
    "edf1['twas_fdr_bh'] = multipletests(edf1['twas_p'], method='fdr_bh')[1]\n",
    "edf1['gwas_fdr_bh'] = multipletests(edf1['gwas_p'], method='fdr_bh')[1]\n",
    "edf1['de_fdr_bh'] = multipletests(edf1['de_p'], method='fdr_bh')[1]\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('wgcna_module_enrichment.csv')\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### No MHC region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:40.405572Z",
     "iopub.status.busy": "2021-09-08T01:09:40.403984Z",
     "iopub.status.idle": "2021-09-08T01:09:41.505802Z",
     "shell.execute_reply": "2021-09-08T01:09:41.504284Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>module_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>grey</th>\n",
       "      <td>3909</td>\n",
       "      <td>0.850262</td>\n",
       "      <td>0.102202</td>\n",
       "      <td>0.281057</td>\n",
       "      <td>1.032990</td>\n",
       "      <td>0.746956</td>\n",
       "      <td>0.887359</td>\n",
       "      <td>0.673036</td>\n",
       "      <td>1.788290e-11</td>\n",
       "      <td>5.620340e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>yellow</th>\n",
       "      <td>1569</td>\n",
       "      <td>1.025675</td>\n",
       "      <td>0.834775</td>\n",
       "      <td>0.931585</td>\n",
       "      <td>1.103254</td>\n",
       "      <td>0.521159</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>0.558072</td>\n",
       "      <td>4.513232e-10</td>\n",
       "      <td>1.241139e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>turquoise</th>\n",
       "      <td>3313</td>\n",
       "      <td>0.865699</td>\n",
       "      <td>0.177595</td>\n",
       "      <td>0.390709</td>\n",
       "      <td>0.834726</td>\n",
       "      <td>0.150140</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>0.305742</td>\n",
       "      <td>2.263035e-56</td>\n",
       "      <td>2.489338e-55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cyan</th>\n",
       "      <td>373</td>\n",
       "      <td>1.256521</td>\n",
       "      <td>0.331638</td>\n",
       "      <td>0.561233</td>\n",
       "      <td>0.976909</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.880481</td>\n",
       "      <td>6.734897e-06</td>\n",
       "      <td>1.229777e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>green</th>\n",
       "      <td>1563</td>\n",
       "      <td>1.415003</td>\n",
       "      <td>0.006436</td>\n",
       "      <td>0.070794</td>\n",
       "      <td>0.901696</td>\n",
       "      <td>0.629698</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>0.381027</td>\n",
       "      <td>7.589137e-21</td>\n",
       "      <td>3.339220e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>brown</th>\n",
       "      <td>2245</td>\n",
       "      <td>0.829071</td>\n",
       "      <td>0.155947</td>\n",
       "      <td>0.381205</td>\n",
       "      <td>0.916562</td>\n",
       "      <td>0.586088</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>2.945581</td>\n",
       "      <td>6.003082e-79</td>\n",
       "      <td>1.320678e-77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>black</th>\n",
       "      <td>1014</td>\n",
       "      <td>1.109538</td>\n",
       "      <td>0.494980</td>\n",
       "      <td>0.725971</td>\n",
       "      <td>1.124947</td>\n",
       "      <td>0.491598</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>2.312840</td>\n",
       "      <td>1.036007e-23</td>\n",
       "      <td>5.698037e-23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>midnightblue</th>\n",
       "      <td>367</td>\n",
       "      <td>1.522798</td>\n",
       "      <td>0.090576</td>\n",
       "      <td>0.281057</td>\n",
       "      <td>1.309193</td>\n",
       "      <td>0.331022</td>\n",
       "      <td>0.809166</td>\n",
       "      <td>1.506621</td>\n",
       "      <td>5.327744e-03</td>\n",
       "      <td>6.168966e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightyellow</th>\n",
       "      <td>226</td>\n",
       "      <td>1.087747</td>\n",
       "      <td>0.722394</td>\n",
       "      <td>0.882926</td>\n",
       "      <td>0.636869</td>\n",
       "      <td>0.535715</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>1.826750</td>\n",
       "      <td>8.010333e-04</td>\n",
       "      <td>9.790407e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tan</th>\n",
       "      <td>408</td>\n",
       "      <td>1.000158</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.557566</td>\n",
       "      <td>0.089448</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>1.944618</td>\n",
       "      <td>5.322901e-07</td>\n",
       "      <td>1.064580e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>royalblue</th>\n",
       "      <td>214</td>\n",
       "      <td>0.496723</td>\n",
       "      <td>0.199549</td>\n",
       "      <td>0.399097</td>\n",
       "      <td>0.501996</td>\n",
       "      <td>0.293814</td>\n",
       "      <td>0.807988</td>\n",
       "      <td>0.814611</td>\n",
       "      <td>4.548135e-01</td>\n",
       "      <td>4.548135e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>magenta</th>\n",
       "      <td>676</td>\n",
       "      <td>0.628150</td>\n",
       "      <td>0.076511</td>\n",
       "      <td>0.281057</td>\n",
       "      <td>0.471059</td>\n",
       "      <td>0.022054</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>0.438756</td>\n",
       "      <td>5.544032e-08</td>\n",
       "      <td>1.219687e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>darkred</th>\n",
       "      <td>202</td>\n",
       "      <td>1.967530</td>\n",
       "      <td>0.021940</td>\n",
       "      <td>0.120667</td>\n",
       "      <td>1.663843</td>\n",
       "      <td>0.128938</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>1.924064</td>\n",
       "      <td>4.104262e-04</td>\n",
       "      <td>5.311398e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pink</th>\n",
       "      <td>788</td>\n",
       "      <td>1.038477</td>\n",
       "      <td>0.846895</td>\n",
       "      <td>0.931585</td>\n",
       "      <td>1.372737</td>\n",
       "      <td>0.118340</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>3.509209</td>\n",
       "      <td>1.168000e-46</td>\n",
       "      <td>8.565331e-46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>blue</th>\n",
       "      <td>2511</td>\n",
       "      <td>1.381774</td>\n",
       "      <td>0.001941</td>\n",
       "      <td>0.042706</td>\n",
       "      <td>1.227438</td>\n",
       "      <td>0.091969</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>0.729536</td>\n",
       "      <td>7.825853e-06</td>\n",
       "      <td>1.229777e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightgreen</th>\n",
       "      <td>227</td>\n",
       "      <td>0.231090</td>\n",
       "      <td>0.019341</td>\n",
       "      <td>0.120667</td>\n",
       "      <td>0.797370</td>\n",
       "      <td>0.837113</td>\n",
       "      <td>0.920824</td>\n",
       "      <td>2.029600</td>\n",
       "      <td>4.564248e-05</td>\n",
       "      <td>6.694230e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>lightcyan</th>\n",
       "      <td>338</td>\n",
       "      <td>1.134521</td>\n",
       "      <td>0.660435</td>\n",
       "      <td>0.854680</td>\n",
       "      <td>0.747402</td>\n",
       "      <td>0.612372</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>2.738411</td>\n",
       "      <td>1.682037e-13</td>\n",
       "      <td>6.167469e-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>purple</th>\n",
       "      <td>475</td>\n",
       "      <td>0.851828</td>\n",
       "      <td>0.622896</td>\n",
       "      <td>0.854680</td>\n",
       "      <td>0.525295</td>\n",
       "      <td>0.088751</td>\n",
       "      <td>0.471869</td>\n",
       "      <td>0.446436</td>\n",
       "      <td>7.295151e-06</td>\n",
       "      <td>1.229777e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>red</th>\n",
       "      <td>1253</td>\n",
       "      <td>0.857296</td>\n",
       "      <td>0.395329</td>\n",
       "      <td>0.621231</td>\n",
       "      <td>1.148942</td>\n",
       "      <td>0.422294</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>0.673181</td>\n",
       "      <td>8.420111e-05</td>\n",
       "      <td>1.157765e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>grey60</th>\n",
       "      <td>291</td>\n",
       "      <td>0.454860</td>\n",
       "      <td>0.082625</td>\n",
       "      <td>0.281057</td>\n",
       "      <td>0.744392</td>\n",
       "      <td>0.589158</td>\n",
       "      <td>0.814904</td>\n",
       "      <td>0.156007</td>\n",
       "      <td>1.187735e-09</td>\n",
       "      <td>2.903353e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>salmon</th>\n",
       "      <td>397</td>\n",
       "      <td>1.324492</td>\n",
       "      <td>0.225311</td>\n",
       "      <td>0.413070</td>\n",
       "      <td>0.915317</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.397243</td>\n",
       "      <td>2.227414e-02</td>\n",
       "      <td>2.450156e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>greenyellow</th>\n",
       "      <td>438</td>\n",
       "      <td>0.927932</td>\n",
       "      <td>0.897975</td>\n",
       "      <td>0.940735</td>\n",
       "      <td>1.086932</td>\n",
       "      <td>0.766356</td>\n",
       "      <td>0.887359</td>\n",
       "      <td>0.771059</td>\n",
       "      <td>1.336408e-01</td>\n",
       "      <td>1.400046e-01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              n_genes   gwas_or    gwas_p  gwas_fdr_bh   twas_or    twas_p  \\\n",
       "module_id                                                                    \n",
       "grey             3909  0.850262  0.102202     0.281057  1.032990  0.746956   \n",
       "yellow           1569  1.025675  0.834775     0.931585  1.103254  0.521159   \n",
       "turquoise        3313  0.865699  0.177595     0.390709  0.834726  0.150140   \n",
       "cyan              373  1.256521  0.331638     0.561233  0.976909  1.000000   \n",
       "green            1563  1.415003  0.006436     0.070794  0.901696  0.629698   \n",
       "brown            2245  0.829071  0.155947     0.381205  0.916562  0.586088   \n",
       "black            1014  1.109538  0.494980     0.725971  1.124947  0.491598   \n",
       "midnightblue      367  1.522798  0.090576     0.281057  1.309193  0.331022   \n",
       "lightyellow       226  1.087747  0.722394     0.882926  0.636869  0.535715   \n",
       "tan               408  1.000158  1.000000     1.000000  1.557566  0.089448   \n",
       "royalblue         214  0.496723  0.199549     0.399097  0.501996  0.293814   \n",
       "magenta           676  0.628150  0.076511     0.281057  0.471059  0.022054   \n",
       "darkred           202  1.967530  0.021940     0.120667  1.663843  0.128938   \n",
       "pink              788  1.038477  0.846895     0.931585  1.372737  0.118340   \n",
       "blue             2511  1.381774  0.001941     0.042706  1.227438  0.091969   \n",
       "lightgreen        227  0.231090  0.019341     0.120667  0.797370  0.837113   \n",
       "lightcyan         338  1.134521  0.660435     0.854680  0.747402  0.612372   \n",
       "purple            475  0.851828  0.622896     0.854680  0.525295  0.088751   \n",
       "red              1253  0.857296  0.395329     0.621231  1.148942  0.422294   \n",
       "grey60            291  0.454860  0.082625     0.281057  0.744392  0.589158   \n",
       "salmon            397  1.324492  0.225311     0.413070  0.915317  1.000000   \n",
       "greenyellow       438  0.927932  0.897975     0.940735  1.086932  0.766356   \n",
       "\n",
       "              twas_fdr_bh     de_or          de_p     de_fdr_bh  \n",
       "module_id                                                        \n",
       "grey             0.887359  0.673036  1.788290e-11  5.620340e-11  \n",
       "yellow           0.814904  0.558072  4.513232e-10  1.241139e-09  \n",
       "turquoise        0.471869  0.305742  2.263035e-56  2.489338e-55  \n",
       "cyan             1.000000  1.880481  6.734897e-06  1.229777e-05  \n",
       "green            0.814904  0.381027  7.589137e-21  3.339220e-20  \n",
       "brown            0.814904  2.945581  6.003082e-79  1.320678e-77  \n",
       "black            0.814904  2.312840  1.036007e-23  5.698037e-23  \n",
       "midnightblue     0.809166  1.506621  5.327744e-03  6.168966e-03  \n",
       "lightyellow      0.814904  1.826750  8.010333e-04  9.790407e-04  \n",
       "tan              0.471869  1.944618  5.322901e-07  1.064580e-06  \n",
       "royalblue        0.807988  0.814611  4.548135e-01  4.548135e-01  \n",
       "magenta          0.471869  0.438756  5.544032e-08  1.219687e-07  \n",
       "darkred          0.471869  1.924064  4.104262e-04  5.311398e-04  \n",
       "pink             0.471869  3.509209  1.168000e-46  8.565331e-46  \n",
       "blue             0.471869  0.729536  7.825853e-06  1.229777e-05  \n",
       "lightgreen       0.920824  2.029600  4.564248e-05  6.694230e-05  \n",
       "lightcyan        0.814904  2.738411  1.682037e-13  6.167469e-13  \n",
       "purple           0.471869  0.446436  7.295151e-06  1.229777e-05  \n",
       "red              0.814904  0.673181  8.420111e-05  1.157765e-04  \n",
       "grey60           0.814904  0.156007  1.187735e-09  2.903353e-09  \n",
       "salmon           1.000000  1.397243  2.227414e-02  2.450156e-02  \n",
       "greenyellow      0.887359  0.771059  1.336408e-01  1.400046e-01  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "edf2 = pd.DataFrame.from_records(enrichment_rows_nomhc(), \n",
    "                                columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', \n",
    "                                         'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                index='module_id')\n",
    "edf2['twas_fdr_bh'] = multipletests(edf2['twas_p'], method='fdr_bh')[1]\n",
    "edf2['gwas_fdr_bh'] = multipletests(edf2['gwas_p'], method='fdr_bh')[1]\n",
    "edf2['de_fdr_bh'] = multipletests(edf2['de_p'], method='fdr_bh')[1]\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('wgcna_module_enrichment_excluding_mhc_region.csv')\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', \n",
    "      'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GO enrichment for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-08T01:09:41.515567Z",
     "iopub.status.busy": "2021-09-08T01:09:41.513993Z",
     "iopub.status.idle": "2021-09-08T01:18:46.321806Z",
     "shell.execute_reply": "2021-09-08T01:18:46.320968Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "requests.get(http://purl.obolibrary.org/obo/go/go-basic.obo, stream=True)\n",
      "  WROTE: go-basic.obo\n",
      "\n",
      "FTP RETR ftp.ncbi.nlm.nih.gov gene/DATA gene2go.gz -> gene2go.gz\n",
      "  gunzip gene2go.gz\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.711431 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2789 IDs ... 76%  2,113 of  2,789 study items found in association\n",
      "100%  2,789 of  2,789 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   5,611 GO terms are associated with  2,113 of  2,789 study items\n",
      "  METHOD fdr_bh:\n",
      "      26 GO terms found significant (< 0.05=alpha) ( 24 enriched +   2 purified): statsmodels fdr_bh\n",
      "     741 study items associated with significant GO IDs (enriched)\n",
      "       9 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2789 IDs ... 80%  2,240 of  2,789 study items found in association\n",
      "100%  2,789 of  2,789 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     817 GO terms are associated with  2,240 of  2,789 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 20 enriched +   0 purified): statsmodels fdr_bh\n",
      "   2,037 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2789 IDs ... 78%  2,164 of  2,789 study items found in association\n",
      "100%  2,789 of  2,789 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "   1,750 GO terms are associated with  2,164 of  2,789 study items\n",
      "  METHOD fdr_bh:\n",
      "      18 GO terms found significant (< 0.05=alpha) ( 16 enriched +   2 purified): statsmodels fdr_bh\n",
      "   1,728 study items associated with significant GO IDs (enriched)\n",
      "       6 study items associated with significant GO IDs (purified)\n",
      "Significant results[64] = 26 BP + 18 MF + 20 CC\n",
      "     64 items WROTE: GO_analysis_module_grey.xlsx\n",
      "     64 GOEA results for  2201 study items. WROTE: GO_analysis_module_grey.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.221850 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1109 IDs ... 81%    901 of  1,109 study items found in association\n",
      "100%  1,109 of  1,109 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,996 GO terms are associated with    901 of  1,109 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 18 enriched +   2 purified): statsmodels fdr_bh\n",
      "     280 study items associated with significant GO IDs (enriched)\n",
      "      13 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1109 IDs ... 84%    935 of  1,109 study items found in association\n",
      "100%  1,109 of  1,109 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     563 GO terms are associated with    935 of  1,109 study items\n",
      "  METHOD fdr_bh:\n",
      "      18 GO terms found significant (< 0.05=alpha) ( 18 enriched +   0 purified): statsmodels fdr_bh\n",
      "     733 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1109 IDs ... 81%    903 of  1,109 study items found in association\n",
      "100%  1,109 of  1,109 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     911 GO terms are associated with    903 of  1,109 study items\n",
      "  METHOD fdr_bh:\n",
      "      16 GO terms found significant (< 0.05=alpha) ( 14 enriched +   2 purified): statsmodels fdr_bh\n",
      "     759 study items associated with significant GO IDs (enriched)\n",
      "       4 study items associated with significant GO IDs (purified)\n",
      "Significant results[54] = 20 BP + 16 MF + 18 CC\n",
      "     54 items WROTE: GO_analysis_module_yellow.xlsx\n",
      "     54 GOEA results for   884 study items. WROTE: GO_analysis_module_yellow.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.852936 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1232 IDs ... 60%    742 of  1,232 study items found in association\n",
      "100%  1,232 of  1,232 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,063 GO terms are associated with    742 of  1,232 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  2 enriched +   0 purified): statsmodels fdr_bh\n",
      "     135 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1232 IDs ... 64%    792 of  1,232 study items found in association\n",
      "100%  1,232 of  1,232 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     474 GO terms are associated with    792 of  1,232 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  1 enriched +   3 purified): statsmodels fdr_bh\n",
      "      27 study items associated with significant GO IDs (enriched)\n",
      "     225 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1232 IDs ... 59%    723 of  1,232 study items found in association\n",
      "100%  1,232 of  1,232 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     718 GO terms are associated with    723 of  1,232 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  5 enriched +   1 purified): statsmodels fdr_bh\n",
      "     182 study items associated with significant GO IDs (enriched)\n",
      "     440 study items associated with significant GO IDs (purified)\n",
      "Significant results[12] = 2 BP + 6 MF + 4 CC\n",
      "     12 items WROTE: GO_analysis_module_turquoise.xlsx\n",
      "     12 GOEA results for   606 study items. WROTE: GO_analysis_module_turquoise.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.340478 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 349 IDs ... 85%    298 of    349 study items found in association\n",
      "100%    349 of    349 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,384 GO terms are associated with    298 of    349 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 349 IDs ... 91%    317 of    349 study items found in association\n",
      "100%    349 of    349 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     323 GO terms are associated with    317 of    349 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  9 enriched +   0 purified): statsmodels fdr_bh\n",
      "     222 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 349 IDs ... 88%    308 of    349 study items found in association\n",
      "100%    349 of    349 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     445 GO terms are associated with    308 of    349 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  2 enriched +   0 purified): statsmodels fdr_bh\n",
      "     233 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[11] = 0 BP + 2 MF + 9 CC\n",
      "     11 items WROTE: GO_analysis_module_cyan.xlsx\n",
      "     11 GOEA results for   294 study items. WROTE: GO_analysis_module_cyan.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.666287 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 909 IDs ... 72%    655 of    909 study items found in association\n",
      "100%    909 of    909 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,226 GO terms are associated with    655 of    909 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "     122 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 909 IDs ... 75%    684 of    909 study items found in association\n",
      "100%    909 of    909 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     489 GO terms are associated with    684 of    909 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  5 enriched +   1 purified): statsmodels fdr_bh\n",
      "     336 study items associated with significant GO IDs (enriched)\n",
      "      24 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 909 IDs ... 74%    671 of    909 study items found in association\n",
      "100%    909 of    909 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     727 GO terms are associated with    671 of    909 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  8 enriched +   1 purified): statsmodels fdr_bh\n",
      "     560 study items associated with significant GO IDs (enriched)\n",
      "       1 study items associated with significant GO IDs (purified)\n",
      "Significant results[18] = 3 BP + 9 MF + 6 CC\n",
      "     18 items WROTE: GO_analysis_module_green.xlsx\n",
      "     18 GOEA results for   605 study items. WROTE: GO_analysis_module_green.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.327052 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2007 IDs ... 86%  1,723 of  2,007 study items found in association\n",
      "100%  2,007 of  2,007 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   4,742 GO terms are associated with  1,723 of  2,007 study items\n",
      "  METHOD fdr_bh:\n",
      "     135 GO terms found significant (< 0.05=alpha) (131 enriched +   4 purified): statsmodels fdr_bh\n",
      "     988 study items associated with significant GO IDs (enriched)\n",
      "      20 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2007 IDs ... 91%  1,819 of  2,007 study items found in association\n",
      "100%  2,007 of  2,007 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     851 GO terms are associated with  1,819 of  2,007 study items\n",
      "  METHOD fdr_bh:\n",
      "     104 GO terms found significant (< 0.05=alpha) (103 enriched +   1 purified): statsmodels fdr_bh\n",
      "   1,750 study items associated with significant GO IDs (enriched)\n",
      "      84 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2007 IDs ... 88%  1,759 of  2,007 study items found in association\n",
      "100%  2,007 of  2,007 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "   1,567 GO terms are associated with  1,759 of  2,007 study items\n",
      "  METHOD fdr_bh:\n",
      "      63 GO terms found significant (< 0.05=alpha) ( 61 enriched +   2 purified): statsmodels fdr_bh\n",
      "   1,512 study items associated with significant GO IDs (enriched)\n",
      "       3 study items associated with significant GO IDs (purified)\n",
      "Significant results[302] = 135 BP + 63 MF + 104 CC\n",
      "    302 items WROTE: GO_analysis_module_brown.xlsx\n",
      "    302 GOEA results for  1858 study items. WROTE: GO_analysis_module_brown.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.482915 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 859 IDs ... 81%    695 of    859 study items found in association\n",
      "100%    859 of    859 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,852 GO terms are associated with    695 of    859 study items\n",
      "  METHOD fdr_bh:\n",
      "      17 GO terms found significant (< 0.05=alpha) ( 15 enriched +   2 purified): statsmodels fdr_bh\n",
      "     216 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 859 IDs ... 84%    725 of    859 study items found in association\n",
      "100%    859 of    859 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     515 GO terms are associated with    725 of    859 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 20 enriched +   0 purified): statsmodels fdr_bh\n",
      "     608 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 859 IDs ... 84%    722 of    859 study items found in association\n",
      "100%    859 of    859 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     861 GO terms are associated with    722 of    859 study items\n",
      "  METHOD fdr_bh:\n",
      "      22 GO terms found significant (< 0.05=alpha) ( 20 enriched +   2 purified): statsmodels fdr_bh\n",
      "     617 study items associated with significant GO IDs (enriched)\n",
      "       4 study items associated with significant GO IDs (purified)\n",
      "Significant results[59] = 17 BP + 22 MF + 20 CC\n",
      "     59 items WROTE: GO_analysis_module_black.xlsx\n",
      "     59 GOEA results for   705 study items. WROTE: GO_analysis_module_black.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.462110 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 348 IDs ... 87%    304 of    348 study items found in association\n",
      "100%    348 of    348 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,619 GO terms are associated with    304 of    348 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      18 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 348 IDs ... 90%    314 of    348 study items found in association\n",
      "100%    348 of    348 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     389 GO terms are associated with    314 of    348 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 20 enriched +   0 purified): statsmodels fdr_bh\n",
      "     251 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 348 IDs ... 89%    309 of    348 study items found in association\n",
      "100%    348 of    348 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     496 GO terms are associated with    309 of    348 study items\n",
      "  METHOD fdr_bh:\n",
      "       5 GO terms found significant (< 0.05=alpha) (  5 enriched +   0 purified): statsmodels fdr_bh\n",
      "     250 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[26] = 1 BP + 5 MF + 20 CC\n",
      "     26 items WROTE: GO_analysis_module_midnightblue.xlsx\n",
      "     26 GOEA results for   306 study items. WROTE: GO_analysis_module_midnightblue.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.162584 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 220 IDs ... 90%    198 of    220 study items found in association\n",
      "100%    220 of    220 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "     965 GO terms are associated with    198 of    220 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 220 IDs ... 94%    206 of    220 study items found in association\n",
      "100%    220 of    220 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     255 GO terms are associated with    206 of    220 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "     107 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 220 IDs ... 93%    204 of    220 study items found in association\n",
      "100%    220 of    220 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     347 GO terms are associated with    204 of    220 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "     174 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[6] = 0 BP + 3 MF + 3 CC\n",
      "      6 items WROTE: GO_analysis_module_lightyellow.xlsx\n",
      "      6 GOEA results for   196 study items. WROTE: GO_analysis_module_lightyellow.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.574018 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 346 IDs ... 84%    292 of    346 study items found in association\n",
      "100%    346 of    346 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,441 GO terms are associated with    292 of    346 study items\n",
      "  METHOD fdr_bh:\n",
      "       7 GO terms found significant (< 0.05=alpha) (  7 enriched +   0 purified): statsmodels fdr_bh\n",
      "      42 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 346 IDs ... 88%    305 of    346 study items found in association\n",
      "100%    346 of    346 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     333 GO terms are associated with    305 of    346 study items\n",
      "  METHOD fdr_bh:\n",
      "      17 GO terms found significant (< 0.05=alpha) ( 17 enriched +   0 purified): statsmodels fdr_bh\n",
      "     249 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 346 IDs ... 86%    296 of    346 study items found in association\n",
      "100%    346 of    346 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     487 GO terms are associated with    296 of    346 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  9 enriched +   0 purified): statsmodels fdr_bh\n",
      "     240 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[33] = 7 BP + 9 MF + 17 CC\n",
      "     33 items WROTE: GO_analysis_module_tan.xlsx\n",
      "     33 GOEA results for   295 study items. WROTE: GO_analysis_module_tan.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.420147 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 197 IDs ... 85%    167 of    197 study items found in association\n",
      "100%    197 of    197 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "     994 GO terms are associated with    167 of    197 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 197 IDs ... 91%    179 of    197 study items found in association\n",
      "100%    197 of    197 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     251 GO terms are associated with    179 of    197 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  4 enriched +   0 purified): statsmodels fdr_bh\n",
      "     130 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 197 IDs ... 94%    185 of    197 study items found in association\n",
      "100%    197 of    197 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     348 GO terms are associated with    185 of    197 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  2 enriched +   0 purified): statsmodels fdr_bh\n",
      "     157 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[6] = 0 BP + 2 MF + 4 CC\n",
      "      6 items WROTE: GO_analysis_module_royalblue.xlsx\n",
      "      6 GOEA results for   176 study items. WROTE: GO_analysis_module_royalblue.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.293213 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 625 IDs ... 75%    471 of    625 study items found in association\n",
      "100%    625 of    625 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,050 GO terms are associated with    471 of    625 study items\n",
      "  METHOD fdr_bh:\n",
      "      30 GO terms found significant (< 0.05=alpha) ( 30 enriched +   0 purified): statsmodels fdr_bh\n",
      "     170 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 625 IDs ... 81%    507 of    625 study items found in association\n",
      "100%    625 of    625 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     370 GO terms are associated with    507 of    625 study items\n",
      "  METHOD fdr_bh:\n",
      "      31 GO terms found significant (< 0.05=alpha) ( 30 enriched +   1 purified): statsmodels fdr_bh\n",
      "     353 study items associated with significant GO IDs (enriched)\n",
      "       8 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 625 IDs ... 77%    480 of    625 study items found in association\n",
      "100%    625 of    625 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     563 GO terms are associated with    480 of    625 study items\n",
      "  METHOD fdr_bh:\n",
      "      15 GO terms found significant (< 0.05=alpha) ( 14 enriched +   1 purified): statsmodels fdr_bh\n",
      "     367 study items associated with significant GO IDs (enriched)\n",
      "       8 study items associated with significant GO IDs (purified)\n",
      "Significant results[76] = 30 BP + 15 MF + 31 CC\n",
      "     76 items WROTE: GO_analysis_module_magenta.xlsx\n",
      "     76 GOEA results for   465 study items. WROTE: GO_analysis_module_magenta.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.139482 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 193 IDs ... 82%    159 of    193 study items found in association\n",
      "100%    193 of    193 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "     924 GO terms are associated with    159 of    193 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 193 IDs ... 91%    176 of    193 study items found in association\n",
      "100%    193 of    193 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     233 GO terms are associated with    176 of    193 study items\n",
      "  METHOD fdr_bh:\n",
      "       7 GO terms found significant (< 0.05=alpha) (  7 enriched +   0 purified): statsmodels fdr_bh\n",
      "      79 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 193 IDs ... 88%    169 of    193 study items found in association\n",
      "100%    193 of    193 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     303 GO terms are associated with    169 of    193 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "     141 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[8] = 0 BP + 1 MF + 7 CC\n",
      "      8 items WROTE: GO_analysis_module_darkred.xlsx\n",
      "      8 GOEA results for   150 study items. WROTE: GO_analysis_module_darkred.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.413095 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 700 IDs ... 85%    598 of    700 study items found in association\n",
      "100%    700 of    700 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,482 GO terms are associated with    598 of    700 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  2 enriched +   0 purified): statsmodels fdr_bh\n",
      "      15 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 700 IDs ... 90%    628 of    700 study items found in association\n",
      "100%    700 of    700 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     539 GO terms are associated with    628 of    700 study items\n",
      "  METHOD fdr_bh:\n",
      "      28 GO terms found significant (< 0.05=alpha) ( 28 enriched +   0 purified): statsmodels fdr_bh\n",
      "     589 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 700 IDs ... 89%    622 of    700 study items found in association\n",
      "100%    700 of    700 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     821 GO terms are associated with    622 of    700 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  4 enriched +   0 purified): statsmodels fdr_bh\n",
      "     488 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[34] = 2 BP + 4 MF + 28 CC\n",
      "     34 items WROTE: GO_analysis_module_pink.xlsx\n",
      "     34 GOEA results for   629 study items. WROTE: GO_analysis_module_pink.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.579850 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2211 IDs ... 87%  1,915 of  2,211 study items found in association\n",
      "100%  2,211 of  2,211 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   4,067 GO terms are associated with  1,915 of  2,211 study items\n",
      "  METHOD fdr_bh:\n",
      "     140 GO terms found significant (< 0.05=alpha) (131 enriched +   9 purified): statsmodels fdr_bh\n",
      "   1,132 study items associated with significant GO IDs (enriched)\n",
      "      58 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2211 IDs ... 92%  2,035 of  2,211 study items found in association\n",
      "100%  2,211 of  2,211 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     861 GO terms are associated with  2,035 of  2,211 study items\n",
      "  METHOD fdr_bh:\n",
      "     181 GO terms found significant (< 0.05=alpha) (173 enriched +   8 purified): statsmodels fdr_bh\n",
      "   1,936 study items associated with significant GO IDs (enriched)\n",
      "     305 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2211 IDs ... 91%  2,008 of  2,211 study items found in association\n",
      "100%  2,211 of  2,211 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "   1,387 GO terms are associated with  2,008 of  2,211 study items\n",
      "  METHOD fdr_bh:\n",
      "      95 GO terms found significant (< 0.05=alpha) ( 85 enriched +  10 purified): statsmodels fdr_bh\n",
      "   1,865 study items associated with significant GO IDs (enriched)\n",
      "     108 study items associated with significant GO IDs (purified)\n",
      "Significant results[416] = 140 BP + 95 MF + 181 CC\n",
      "    416 items WROTE: GO_analysis_module_blue.xlsx\n",
      "    416 GOEA results for  2074 study items. WROTE: GO_analysis_module_blue.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.100211 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 184 IDs ... 78%    144 of    184 study items found in association\n",
      "100%    184 of    184 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "     871 GO terms are associated with    144 of    184 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 184 IDs ... 82%    150 of    184 study items found in association\n",
      "100%    184 of    184 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     214 GO terms are associated with    150 of    184 study items\n",
      "  METHOD fdr_bh:\n",
      "       2 GO terms found significant (< 0.05=alpha) (  2 enriched +   0 purified): statsmodels fdr_bh\n",
      "      20 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 184 IDs ... 74%    136 of    184 study items found in association\n",
      "100%    184 of    184 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     254 GO terms are associated with    136 of    184 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[2] = 0 BP + 0 MF + 2 CC\n",
      "      2 items WROTE: GO_analysis_module_lightgreen.xlsx\n",
      "      2 GOEA results for    20 study items. WROTE: GO_analysis_module_lightgreen.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.036047 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 307 IDs ... 89%    272 of    307 study items found in association\n",
      "100%    307 of    307 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,478 GO terms are associated with    272 of    307 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 20 enriched +   0 purified): statsmodels fdr_bh\n",
      "      77 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 307 IDs ... 92%    281 of    307 study items found in association\n",
      "100%    307 of    307 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     353 GO terms are associated with    281 of    307 study items\n",
      "  METHOD fdr_bh:\n",
      "      45 GO terms found significant (< 0.05=alpha) ( 45 enriched +   0 purified): statsmodels fdr_bh\n",
      "     232 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 307 IDs ... 88%    270 of    307 study items found in association\n",
      "100%    307 of    307 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     430 GO terms are associated with    270 of    307 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  6 enriched +   0 purified): statsmodels fdr_bh\n",
      "     195 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[71] = 20 BP + 6 MF + 45 CC\n",
      "     71 items WROTE: GO_analysis_module_lightcyan.xlsx\n",
      "     71 GOEA results for   271 study items. WROTE: GO_analysis_module_lightcyan.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.291330 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 471 IDs ... 92%    431 of    471 study items found in association\n",
      "100%    471 of    471 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   2,889 GO terms are associated with    431 of    471 study items\n",
      "  METHOD fdr_bh:\n",
      "     232 GO terms found significant (< 0.05=alpha) (231 enriched +   1 purified): statsmodels fdr_bh\n",
      "     348 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 471 IDs ... 95%    449 of    471 study items found in association\n",
      "100%    471 of    471 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     422 GO terms are associated with    449 of    471 study items\n",
      "  METHOD fdr_bh:\n",
      "      68 GO terms found significant (< 0.05=alpha) ( 68 enriched +   0 purified): statsmodels fdr_bh\n",
      "     439 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 471 IDs ... 93%    440 of    471 study items found in association\n",
      "100%    471 of    471 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     705 GO terms are associated with    440 of    471 study items\n",
      "  METHOD fdr_bh:\n",
      "      32 GO terms found significant (< 0.05=alpha) ( 32 enriched +   0 purified): statsmodels fdr_bh\n",
      "     392 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[332] = 232 BP + 32 MF + 68 CC\n",
      "    332 items WROTE: GO_analysis_module_purple.xlsx\n",
      "    332 GOEA results for   447 study items. WROTE: GO_analysis_module_purple.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.513058 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1233 IDs ... 91%  1,121 of  1,233 study items found in association\n",
      "100%  1,233 of  1,233 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   3,575 GO terms are associated with  1,121 of  1,233 study items\n",
      "  METHOD fdr_bh:\n",
      "      61 GO terms found significant (< 0.05=alpha) ( 58 enriched +   3 purified): statsmodels fdr_bh\n",
      "     607 study items associated with significant GO IDs (enriched)\n",
      "       5 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1233 IDs ... 95%  1,176 of  1,233 study items found in association\n",
      "100%  1,233 of  1,233 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     736 GO terms are associated with  1,176 of  1,233 study items\n",
      "  METHOD fdr_bh:\n",
      "     110 GO terms found significant (< 0.05=alpha) (108 enriched +   2 purified): statsmodels fdr_bh\n",
      "   1,147 study items associated with significant GO IDs (enriched)\n",
      "      77 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1233 IDs ... 94%  1,159 of  1,233 study items found in association\n",
      "100%  1,233 of  1,233 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "   1,113 GO terms are associated with  1,159 of  1,233 study items\n",
      "  METHOD fdr_bh:\n",
      "      46 GO terms found significant (< 0.05=alpha) ( 43 enriched +   3 purified): statsmodels fdr_bh\n",
      "   1,068 study items associated with significant GO IDs (enriched)\n",
      "       3 study items associated with significant GO IDs (purified)\n",
      "Significant results[217] = 61 BP + 46 MF + 110 CC\n",
      "    217 items WROTE: GO_analysis_module_red.xlsx\n",
      "    217 GOEA results for  1196 study items. WROTE: GO_analysis_module_red.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.322407 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 290 IDs ... 90%    261 of    290 study items found in association\n",
      "100%    290 of    290 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,768 GO terms are associated with    261 of    290 study items\n",
      "  METHOD fdr_bh:\n",
      "     251 GO terms found significant (< 0.05=alpha) (251 enriched +   0 purified): statsmodels fdr_bh\n",
      "     222 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 290 IDs ... 96%    277 of    290 study items found in association\n",
      "100%    290 of    290 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     269 GO terms are associated with    277 of    290 study items\n",
      "  METHOD fdr_bh:\n",
      "      65 GO terms found significant (< 0.05=alpha) ( 65 enriched +   0 purified): statsmodels fdr_bh\n",
      "     265 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 290 IDs ... 92%    267 of    290 study items found in association\n",
      "100%    290 of    290 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     467 GO terms are associated with    267 of    290 study items\n",
      "  METHOD fdr_bh:\n",
      "      35 GO terms found significant (< 0.05=alpha) ( 35 enriched +   0 purified): statsmodels fdr_bh\n",
      "     231 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[351] = 251 BP + 35 MF + 65 CC\n",
      "    351 items WROTE: GO_analysis_module_grey60.xlsx\n",
      "    351 GOEA results for   275 study items. WROTE: GO_analysis_module_grey60.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:06.066688 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 234 IDs ... 68%    159 of    234 study items found in association\n",
      "100%    234 of    234 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "     635 GO terms are associated with    159 of    234 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 234 IDs ... 71%    167 of    234 study items found in association\n",
      "100%    234 of    234 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     167 GO terms are associated with    167 of    234 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 234 IDs ... 65%    152 of    234 study items found in association\n",
      "100%    234 of    234 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     280 GO terms are associated with    152 of    234 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      41 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[1] = 0 BP + 1 MF + 0 CC\n",
      "      1 items WROTE: GO_analysis_module_salmon.xlsx\n",
      "      1 GOEA results for    41 study items. WROTE: GO_analysis_module_salmon.txt\n",
      "  EXISTS: go-basic.obo\n",
      "  EXISTS: gene2go\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:05.791595 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "MF 18,194 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,231 of 29,107 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 74% 21,438 of 29,107 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 70% 20,357 of 29,107 population items found in association\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 417 IDs ... 88%    369 of    417 study items found in association\n",
      "100%    417 of    417 study items found in population(29107)\n",
      "Calculating 12,416 uncorrected p-values using fisher_scipy_stats\n",
      "  12,416 GO terms are associated with 17,843 of 29,107 population items\n",
      "   1,923 GO terms are associated with    369 of    417 study items\n",
      "  METHOD fdr_bh:\n",
      "      10 GO terms found significant (< 0.05=alpha) ( 10 enriched +   0 purified): statsmodels fdr_bh\n",
      "     118 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 417 IDs ... 93%    387 of    417 study items found in association\n",
      "100%    417 of    417 study items found in population(29107)\n",
      "Calculating 1,752 uncorrected p-values using fisher_scipy_stats\n",
      "   1,752 GO terms are associated with 18,721 of 29,107 population items\n",
      "     350 GO terms are associated with    387 of    417 study items\n",
      "  METHOD fdr_bh:\n",
      "      18 GO terms found significant (< 0.05=alpha) ( 18 enriched +   0 purified): statsmodels fdr_bh\n",
      "     323 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 417 IDs ... 92%    382 of    417 study items found in association\n",
      "100%    417 of    417 study items found in population(29107)\n",
      "Calculating 4,418 uncorrected p-values using fisher_scipy_stats\n",
      "   4,418 GO terms are associated with 17,841 of 29,107 population items\n",
      "     610 GO terms are associated with    382 of    417 study items\n",
      "  METHOD fdr_bh:\n",
      "      12 GO terms found significant (< 0.05=alpha) ( 12 enriched +   0 purified): statsmodels fdr_bh\n",
      "     314 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "Significant results[40] = 10 BP + 12 MF + 18 CC\n",
      "     40 items WROTE: GO_analysis_module_greenyellow.xlsx\n",
      "     40 GOEA results for   367 study items. WROTE: GO_analysis_module_greenyellow.txt\n"
     ]
    }
   ],
   "source": [
    "for mod in get_wgcna_modules().module.unique():\n",
    "    run_goea(mod)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
