{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:58.775015Z",
     "iopub.status.busy": "2021-09-10T13:40:58.769595Z",
     "iopub.status.idle": "2021-09-10T13:40:59.922641Z",
     "shell.execute_reply": "2021-09-10T13:40:59.921307Z"
    }
   },
   "outputs": [],
   "source": [
    "# Leiden and igraph\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.neighbors import kneighbors_graph\n",
    "import igraph as ig\n",
    "import leidenalg\n",
    "import re\n",
    "import scipy.stats as stats\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "\n",
    "# Ensembl ID to Entrez ID conversion\n",
    "import rpy2.robjects as ro\n",
    "from rpy2.robjects import pandas2ri\n",
    "from rpy2.robjects.conversion import localconverter\n",
    "\n",
    "\n",
    "# GO analysis\n",
    "from goatools.base import download_go_basic_obo\n",
    "from goatools.base import download_ncbi_associations\n",
    "from goatools.obo_parser import GODag\n",
    "from goatools.anno.genetogo_reader import Gene2GoReader\n",
    "from goatools.goea.go_enrichment_ns import GOEnrichmentStudyNS\n",
    "\n",
    "\n",
    "# Word cloud\n",
    "import wordcloud\n",
    "import math\n",
    "import random\n",
    "\n",
    "# GTF parser for Ensembl ID to gene symbol conversion\n",
    "from gtfparse import read_gtf\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:59.931923Z",
     "iopub.status.busy": "2021-09-10T13:40:59.930653Z",
     "iopub.status.idle": "2021-09-10T13:40:59.933202Z",
     "shell.execute_reply": "2021-09-10T13:40:59.934223Z"
    }
   },
   "outputs": [],
   "source": [
    "def latent_variables_to_leiden(df, n_neighbors, directed, seed):\n",
    "    \n",
    "    adjacency = kneighbors_graph(df, n_neighbors, mode='connectivity', include_self=False)\n",
    "    \n",
    "    distance = kneighbors_graph(df, n_neighbors, mode='distance', include_self=False)\n",
    "    \n",
    "    sources, targets = adjacency.nonzero()\n",
    "\n",
    "    # weight is inverse distance squared\n",
    "\n",
    "    weights = 1 /  distance[sources, targets].A1 ** 2\n",
    "\n",
    "    \n",
    "    g = ig.Graph(directed=directed)\n",
    "    g.add_vertices(adjacency.shape[0])  # this adds adjacency.shape[0] vertices\n",
    "    g.add_edges(list(zip(sources, targets)))\n",
    "    g.es['weight'] = weights\n",
    "    \n",
    "    part = leidenalg.find_partition(g, leidenalg.ModularityVertexPartition, weights='weight', seed=seed)\n",
    "    \n",
    "    \n",
    "    membership_df = pd.DataFrame({'cluster_id':part.membership}, index=df.index)\n",
    "    \n",
    "    return (membership_df, part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:59.949800Z",
     "iopub.status.busy": "2021-09-10T13:40:59.948550Z",
     "iopub.status.idle": "2021-09-10T13:40:59.951065Z",
     "shell.execute_reply": "2021-09-10T13:40:59.952149Z"
    }
   },
   "outputs": [],
   "source": [
    "# Ensembl to Entrez\n",
    "\n",
    "class EnsemblIDToEntrezIDConverter():\n",
    "    def __init__(self):\n",
    "        with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "            df_a = ro.conversion.rpy2py(ro.r('''\n",
    "                                             library(org.Hs.eg.db)\n",
    "                                             as.data.frame(org.Hs.egENSEMBL)\n",
    "                                            '''))\n",
    "            self.entrez_ensembl_df = df_a.groupby(['ensembl_id']).first()\n",
    "    \n",
    "    def convert(self, gene_list_or_set):\n",
    "        \n",
    "        ensembl_id_df = pd.DataFrame(index = { re.sub(\"\\..*$\",\"\",x) for x in gene_list_or_set })\n",
    "        entrez_ids = set(ensembl_id_df.merge(self.entrez_ensembl_df, left_index=True, right_index=True)['gene_id'].astype(int))\n",
    "        return entrez_ids\n",
    "        \n",
    "    \n",
    "    \n",
    "    \n",
    "# Gene Ontology\n",
    "    \n",
    "def get_entrez_ensembl_df():\n",
    "    with localconverter(ro.default_converter + pandas2ri.converter):\n",
    "        df_a = ro.conversion.rpy2py(ro.r('''\n",
    "                                         library(org.Hs.eg.db)\n",
    "                                         as.data.frame(org.Hs.egENSEMBL)\n",
    "                                        '''))\n",
    "        entrez_ensembl_df = df_a.groupby(['ensembl_id']).first()\n",
    "        return entrez_ensembl_df\n",
    "\n",
    "\n",
    "class MyGeneOntologyAnalysis():\n",
    "    def __init__(self):\n",
    "        \n",
    "\n",
    "        obo_fname = download_go_basic_obo()\n",
    "        file_gene2go = download_ncbi_associations()\n",
    "        obodag = GODag(\"go-basic.obo\")\n",
    "\n",
    "        # Read NCBI's gene2go. Store annotations in a list of namedtuples\n",
    "        objanno = Gene2GoReader(file_gene2go, taxids=[9606])\n",
    "\n",
    "        # Get associations for each branch of the GO DAG (BP, MF, CC)\n",
    "        ns2assoc = objanno.get_ns2assc()\n",
    "\n",
    "        for nspc, id2gos in ns2assoc.items():\n",
    "            print(\"{NS} {N:,} annotated human genes\".format(NS=nspc, N=len(id2gos)))\n",
    "          \n",
    "        genes_with_annotation = set.union(*(set(x.keys()) for x in ns2assoc.values()))\n",
    "\n",
    "        \n",
    "        self.alpha = 0.05\n",
    "        \n",
    "        self.goeaobj = GOEnrichmentStudyNS(\n",
    "            genes_with_annotation, # List of human genes\n",
    "            ns2assoc, # geneid/GO associations\n",
    "            obodag, # Ontologies\n",
    "            propagate_counts = False,\n",
    "            alpha = self.alpha, # default significance cut-off\n",
    "            methods = ['fdr_bh']) # defult multipletest correction method\n",
    "        \n",
    "        \n",
    "\n",
    "    def goea_results_all(self, gene_set):\n",
    "        return self.goeaobj.run_study(gene_set)\n",
    "    \n",
    "    \n",
    "    def goea_results_significant(self, gene_set):\n",
    "        all_results = self.goea_results_all(gene_set)\n",
    "        return [r for r in all_results if r.p_fdr_bh < self.alpha]\n",
    "    \n",
    "    \n",
    "\n",
    "# Word cloud\n",
    "    \n",
    "class GeneOntologyWordCloud():\n",
    "    def __init__(self):\n",
    "        self.wc = wordcloud.WordCloud(colormap='rainbow', \n",
    "                             stopwords=['integral', 'component', 'of', 'process', 'activity', 'to'],\n",
    "                             collocations = True,\n",
    "                             ranks_only=True,\n",
    "                         )\n",
    "        \n",
    "    def gen_random_text(self):\n",
    "        # Generate a spacer between gene ontology terms\n",
    "        return ' '.join(''.join((random.choice('abcdefghijklmnopqrstuvwxyz') for _ in range(150))) for _ in range(3))\n",
    "        \n",
    "    def goea_to_text(self, goea_results):\n",
    "        # Generate a text concatenating GO term names, each name being repeated proportionally to the minus log of its pvalue\n",
    "        return ' '.join((''.join( ' ' + self.gen_random_text() + ' ' + x.name for _ in range(int((-10) * math.log2(x.p_uncorrected))))\n",
    "                for x in goea_results  if x.enrichment == 'e'))\n",
    "    \n",
    "    def generate_image(self, goea_results):\n",
    "        return self.wc.generate(self.goea_to_text(goea_results)).to_image()\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:59.959206Z",
     "iopub.status.busy": "2021-09-10T13:40:59.957886Z",
     "iopub.status.idle": "2021-09-10T13:40:59.960442Z",
     "shell.execute_reply": "2021-09-10T13:40:59.961503Z"
    }
   },
   "outputs": [],
   "source": [
    "def get_gtf_genes_df():\n",
    "    try:\n",
    "        gtf_df_genes = pd.read_csv(\"gtf_df_genes.csv\")\n",
    "    except:\n",
    "        gtf_df = read_gtf(\"/ceph/genome/human/gencode25/gtf.CHR/_m/gencode.v25.annotation.gtf\")\n",
    "        gtf_df_genes = gtf_df[gtf_df[\"feature\"] == \"gene\"][['gene_id', 'gene_name']]\n",
    "        gtf_df_genes.to_csv(\"gtf_df_genes.csv\", index=None)\n",
    "        \n",
    "    return gtf_df_genes\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:59.971787Z",
     "iopub.status.busy": "2021-09-10T13:40:59.970390Z",
     "iopub.status.idle": "2021-09-10T13:40:59.973050Z",
     "shell.execute_reply": "2021-09-10T13:40:59.974071Z"
    }
   },
   "outputs": [],
   "source": [
    "class ClusterGOWordCloud():\n",
    "    def __init__(self, n_neighbors, directed, seed):\n",
    "        \n",
    "        self.n_neighbors = n_neighbors\n",
    "        expression_df = pd.read_csv('../../_m/latent_variables.csv', index_col=0)\n",
    "        mucols = [x for x in expression_df.columns if 'mu' in x]\n",
    "        self.expression_df = expression_df[mucols]\n",
    "        \n",
    "        \n",
    "        \n",
    "        (self.mdf, self.part) =  latent_variables_to_leiden(self.expression_df, n_neighbors, directed, seed)\n",
    "        \n",
    "        \n",
    "        self.gtf_df_genes = get_gtf_genes_df()\n",
    "\n",
    "        self.e2e = EnsemblIDToEntrezIDConverter()\n",
    "        self.mygoa = MyGeneOntologyAnalysis()\n",
    "        self.gowc = GeneOntologyWordCloud()\n",
    "    \n",
    "    \n",
    "    def cluster_df(self, cluster_id):\n",
    "        \n",
    "        return self.expression_df.iloc[self.part[cluster_id]][[]]\\\n",
    "        .merge(cgowc.gtf_df_genes, left_index=True, right_on='gene_id', how='left')\\\n",
    "        .set_index('gene_id')\n",
    "        \n",
    "    \n",
    "    def pipeline(self, cluster_id, filename_prefix):\n",
    "        \n",
    "        #with open(\"%s_neighbors.txt\" % filename_prefix, \"wt\") as f:\n",
    "        #    for x in nn:\n",
    "        #        print(x, file=f)\n",
    "\n",
    "        #self.gn.neighbors_df(gene_id, self.n_neighbors)\\\n",
    "        #.to_csv(\"%s_neighbors.csv\" % filename_prefix)\n",
    "        \n",
    "        #nn = self.gn.neighbors(gene_id, self.n_neighbors)\n",
    "        \n",
    "        self.cluster_df(cluster_id).to_csv(\"%s_genes.csv\" % filename_prefix)\n",
    "        \n",
    "        nn = set((self.expression_df.index[x] for x in self.part[cluster_id]))\n",
    "                \n",
    "        go_r = self.mygoa.goea_results_significant(self.e2e.convert(nn))\n",
    "        self.mygoa.goeaobj.wr_tsv(\"%s_go_enrichment.tsv\" % filename_prefix, go_r)\n",
    "        \n",
    "        if len(go_r) > 0:\n",
    "            p = self.gowc.generate_image(go_r)\n",
    "            p.save(\"%s_go_wordcloud.png\" % filename_prefix)\n",
    "        \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:40:59.980740Z",
     "iopub.status.busy": "2021-09-10T13:40:59.979575Z",
     "iopub.status.idle": "2021-09-10T13:42:22.321638Z",
     "shell.execute_reply": "2021-09-10T13:42:22.320532Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/apua/.local/lib/python3.9/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "/home/apua/.local/lib/python3.9/site-packages/gtfparse/read_gtf.py:151: FutureWarning: The warn_bad_lines argument has been deprecated and will be removed in a future version.\n",
      "\n",
      "\n",
      "  result = parse_gtf(\n",
      "INFO:root:Extracted GTF attributes: ['gene_id', 'gene_type', 'gene_status', 'gene_name', 'level', 'havana_gene', 'transcript_id', 'transcript_type', 'transcript_status', 'transcript_name', 'transcript_support_level', 'tag', 'havana_transcript', 'exon_number', 'exon_id', 'ont', 'protein_id', 'ccdsid']\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: AnnotationDbi\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: stats4\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: BiocGenerics\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: parallel\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: Biobase\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: IRanges\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: Loading required package: S4Vectors\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "WARNING:rpy2.rinterface_lib.callbacks:R[write to console]: \n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "requests.get(http://purl.obolibrary.org/obo/go/go-basic.obo, stream=True)\n",
      "  WROTE: go-basic.obo\n",
      "\n",
      "FTP RETR ftp.ncbi.nlm.nih.gov gene/DATA gene2go.gz -> gene2go.gz\n",
      "  gunzip gene2go.gz\n",
      "go-basic.obo: fmt(1.2) rel(2021-09-01) 47,191 GO Terms\n",
      "HMS:0:00:04.667423 330,404 annotations, 20,688 genes, 18,642 GOs, 1 taxids READ: gene2go \n",
      "MF 18,194 annotated human genes\n",
      "CC 19,433 annotated human genes\n",
      "BP 18,501 annotated human genes\n",
      "\n",
      "Load BP Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 90% 18,501 of 20,652 population items found in association\n",
      "\n",
      "Load CC Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 94% 19,433 of 20,652 population items found in association\n",
      "\n",
      "Load MF Gene Ontology Analysis ...\n",
      "fisher module not installed.  Falling back on scipy.stats.fisher_exact\n",
      " 88% 18,194 of 20,652 population items found in association\n"
     ]
    }
   ],
   "source": [
    "cgowc = ClusterGOWordCloud(8, False, 1092333)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.329633Z",
     "iopub.status.busy": "2021-09-10T13:42:22.328449Z",
     "iopub.status.idle": "2021-09-10T13:42:22.336137Z",
     "shell.execute_reply": "2021-09-10T13:42:22.337178Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cgowc.part)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.362063Z",
     "iopub.status.busy": "2021-09-10T13:42:22.351316Z",
     "iopub.status.idle": "2021-09-10T13:42:22.441411Z",
     "shell.execute_reply": "2021-09-10T13:42:22.440349Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "D2 junction 5-6 is in module 0\n",
      "D2 junction 5-7 is in module 0\n",
      "SETD1A is in module 5\n",
      "DRD2 is in module 0\n"
     ]
    }
   ],
   "source": [
    "a = [x for x in range(len(cgowc.part)) if 1 in cgowc.part[x]][0]\n",
    "print(\"D2 junction 5-6 is in module\", a)\n",
    "\n",
    "a = [x for x in range(len(cgowc.part)) if 0 in cgowc.part[x]][0]\n",
    "print(\"D2 junction 5-7 is in module\", a)\n",
    "\n",
    "d = {z[0]:z[1]  for z in zip(cgowc.gtf_df_genes['gene_name'], cgowc.gtf_df_genes['gene_id'])}\n",
    "a = cgowc.mdf.loc[d['SETD1A'], 'cluster_id']\n",
    "print(\"SETD1A is in module\", a)\n",
    "\n",
    "a = cgowc.mdf.loc[d['DRD2'], 'cluster_id']\n",
    "print(\"DRD2 is in module\", a)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GWAS, TWAS and DE enrichment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.448629Z",
     "iopub.status.busy": "2021-09-10T13:42:22.447396Z",
     "iopub.status.idle": "2021-09-10T13:42:22.459126Z",
     "shell.execute_reply": "2021-09-10T13:42:22.457916Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2696"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "de_genes = set(pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/differential_expression/_m/genes/diffExpr_szVctl_FDR05.txt',\n",
    "                      sep='\\t', usecols=[0], index_col=0).index)\n",
    "len(de_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.466471Z",
     "iopub.status.busy": "2021-09-10T13:42:22.465211Z",
     "iopub.status.idle": "2021-09-10T13:42:22.508072Z",
     "shell.execute_reply": "2021-09-10T13:42:22.507065Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "684"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_twas_genes():\n",
    "    df = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas/results_tables/pgc2/_m/gene_twas_results_including_MHC.csv')\n",
    "    return set(df[df['TWAS.P.fdr_bh'] < .05]['gene_id'])\n",
    "\n",
    "\n",
    "twas_genes = get_twas_genes()\n",
    "len(twas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.514815Z",
     "iopub.status.busy": "2021-09-10T13:42:22.513625Z",
     "iopub.status.idle": "2021-09-10T13:42:22.521910Z",
     "shell.execute_reply": "2021-09-10T13:42:22.520811Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas_genes = set(pd.read_csv('/ceph/projects/v4_phase3_paper/inputs/sz_gwas/pgc2_clozuk/table_s3/hg38/genes/_m/gwas_genes.csv')['gene_id'])\n",
    "len(gwas_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.529363Z",
     "iopub.status.busy": "2021-09-10T13:42:22.526592Z",
     "iopub.status.idle": "2021-09-10T13:42:22.533746Z",
     "shell.execute_reply": "2021-09-10T13:42:22.532664Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "383"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mhc_genes = set(pd.read_csv('/ceph/projects/v4_phase3_paper/inputs/sz_gwas/pgc2_clozuk/table_s3/hg38/mhc_region_genes/_m/mhc_genes.csv')['gene_id'])\n",
    "len(mhc_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.540990Z",
     "iopub.status.busy": "2021-09-10T13:42:22.539805Z",
     "iopub.status.idle": "2021-09-10T13:42:22.543663Z",
     "shell.execute_reply": "2021-09-10T13:42:22.542649Z"
    }
   },
   "outputs": [],
   "source": [
    "def fet(a, b, u):\n",
    "    # a, b, u are sets\n",
    "    # u is the universe\n",
    "    \n",
    "    yes_a = u.intersection(a)\n",
    "    yes_b = u.intersection(b)\n",
    "    no_a = u - a\n",
    "    no_b = u - b\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "    m = [[len(yes_a.intersection(yes_b)), len(no_a.intersection(yes_b)) ], \n",
    "                               [len(yes_a.intersection(no_b)), len(no_a.intersection(no_b))]]\n",
    "    return stats.fisher_exact(m) #, m, len(yes_b)/len(u), len(yes_a)/len(u), len(yes_a.intersection(yes_b)) * len(no_a.intersection(no_b)) / (len(yes_a.intersection(no_b)) * len(no_a.intersection(yes_b)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:22.566609Z",
     "iopub.status.busy": "2021-09-10T13:42:22.559469Z",
     "iopub.status.idle": "2021-09-10T13:42:23.181362Z",
     "shell.execute_reply": "2021-09-10T13:42:23.180173Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>module_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4932</td>\n",
       "      <td>1.195380</td>\n",
       "      <td>0.017789</td>\n",
       "      <td>1.515280</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>2.970279</td>\n",
       "      <td>1.148426e-124</td>\n",
       "      <td>0.000073</td>\n",
       "      <td>0.085240</td>\n",
       "      <td>2.296851e-123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3681</td>\n",
       "      <td>0.977286</td>\n",
       "      <td>0.827135</td>\n",
       "      <td>0.569069</td>\n",
       "      <td>0.000021</td>\n",
       "      <td>0.343346</td>\n",
       "      <td>3.497782e-46</td>\n",
       "      <td>0.000137</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>3.497782e-45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3047</td>\n",
       "      <td>1.048770</td>\n",
       "      <td>0.602387</td>\n",
       "      <td>1.686867</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>1.204706</td>\n",
       "      <td>2.526951e-03</td>\n",
       "      <td>0.000030</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>3.158689e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2688</td>\n",
       "      <td>1.110855</td>\n",
       "      <td>0.270985</td>\n",
       "      <td>1.294817</td>\n",
       "      <td>0.032309</td>\n",
       "      <td>0.957064</td>\n",
       "      <td>5.414012e-01</td>\n",
       "      <td>0.080773</td>\n",
       "      <td>0.677462</td>\n",
       "      <td>5.414012e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1533</td>\n",
       "      <td>1.046916</td>\n",
       "      <td>0.697807</td>\n",
       "      <td>0.839276</td>\n",
       "      <td>0.401404</td>\n",
       "      <td>0.908283</td>\n",
       "      <td>3.126862e-01</td>\n",
       "      <td>0.573434</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>3.474292e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1488</td>\n",
       "      <td>1.170352</td>\n",
       "      <td>0.211855</td>\n",
       "      <td>1.185547</td>\n",
       "      <td>0.269657</td>\n",
       "      <td>1.559550</td>\n",
       "      <td>3.047559e-08</td>\n",
       "      <td>0.414857</td>\n",
       "      <td>0.605301</td>\n",
       "      <td>5.541017e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1353</td>\n",
       "      <td>0.471844</td>\n",
       "      <td>0.000014</td>\n",
       "      <td>0.564796</td>\n",
       "      <td>0.009668</td>\n",
       "      <td>1.069158</td>\n",
       "      <td>4.564584e-01</td>\n",
       "      <td>0.038673</td>\n",
       "      <td>0.000275</td>\n",
       "      <td>4.804825e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1157</td>\n",
       "      <td>0.722880</td>\n",
       "      <td>0.054457</td>\n",
       "      <td>0.494499</td>\n",
       "      <td>0.003970</td>\n",
       "      <td>0.183562</td>\n",
       "      <td>3.232025e-27</td>\n",
       "      <td>0.019850</td>\n",
       "      <td>0.181522</td>\n",
       "      <td>1.616013e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1103</td>\n",
       "      <td>0.545154</td>\n",
       "      <td>0.001089</td>\n",
       "      <td>0.556236</td>\n",
       "      <td>0.018219</td>\n",
       "      <td>0.201529</td>\n",
       "      <td>1.279795e-24</td>\n",
       "      <td>0.060729</td>\n",
       "      <td>0.007257</td>\n",
       "      <td>4.265982e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1047</td>\n",
       "      <td>1.116042</td>\n",
       "      <td>0.438109</td>\n",
       "      <td>0.860136</td>\n",
       "      <td>0.547137</td>\n",
       "      <td>0.160597</td>\n",
       "      <td>9.165784e-27</td>\n",
       "      <td>0.729516</td>\n",
       "      <td>0.836868</td>\n",
       "      <td>3.666314e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>871</td>\n",
       "      <td>1.010937</td>\n",
       "      <td>0.932394</td>\n",
       "      <td>0.713797</td>\n",
       "      <td>0.226131</td>\n",
       "      <td>0.184693</td>\n",
       "      <td>1.073003e-20</td>\n",
       "      <td>0.376884</td>\n",
       "      <td>0.981467</td>\n",
       "      <td>3.065723e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>795</td>\n",
       "      <td>0.989432</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.532632</td>\n",
       "      <td>0.037915</td>\n",
       "      <td>0.407333</td>\n",
       "      <td>6.069822e-09</td>\n",
       "      <td>0.084256</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.213964e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>697</td>\n",
       "      <td>0.925349</td>\n",
       "      <td>0.776935</td>\n",
       "      <td>0.843455</td>\n",
       "      <td>0.624397</td>\n",
       "      <td>3.813977</td>\n",
       "      <td>1.735665e-44</td>\n",
       "      <td>0.734585</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>1.157110e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>554</td>\n",
       "      <td>0.826667</td>\n",
       "      <td>0.460277</td>\n",
       "      <td>0.416404</td>\n",
       "      <td>0.027279</td>\n",
       "      <td>0.229651</td>\n",
       "      <td>8.001289e-12</td>\n",
       "      <td>0.077939</td>\n",
       "      <td>0.836868</td>\n",
       "      <td>2.000322e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>454</td>\n",
       "      <td>1.997844</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>1.510976</td>\n",
       "      <td>0.097231</td>\n",
       "      <td>0.200021</td>\n",
       "      <td>7.642308e-11</td>\n",
       "      <td>0.194461</td>\n",
       "      <td>0.002492</td>\n",
       "      <td>1.698291e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>354</td>\n",
       "      <td>1.049345</td>\n",
       "      <td>0.792376</td>\n",
       "      <td>0.436736</td>\n",
       "      <td>0.121018</td>\n",
       "      <td>0.232567</td>\n",
       "      <td>8.607329e-08</td>\n",
       "      <td>0.220033</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>1.434555e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>266</td>\n",
       "      <td>0.335891</td>\n",
       "      <td>0.021310</td>\n",
       "      <td>0.887646</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.498185</td>\n",
       "      <td>7.312715e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.085240</td>\n",
       "      <td>8.603194e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>186</td>\n",
       "      <td>1.127417</td>\n",
       "      <td>0.715198</td>\n",
       "      <td>1.064241</td>\n",
       "      <td>0.813045</td>\n",
       "      <td>0.196921</td>\n",
       "      <td>6.227493e-05</td>\n",
       "      <td>0.903383</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>8.896418e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>157</td>\n",
       "      <td>0.577684</td>\n",
       "      <td>0.425910</td>\n",
       "      <td>1.268719</td>\n",
       "      <td>0.603295</td>\n",
       "      <td>0.234532</td>\n",
       "      <td>6.879887e-04</td>\n",
       "      <td>0.734585</td>\n",
       "      <td>0.836868</td>\n",
       "      <td>9.173182e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>150</td>\n",
       "      <td>0.762855</td>\n",
       "      <td>0.688923</td>\n",
       "      <td>0.784920</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.562200</td>\n",
       "      <td>1.301102e-05</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.919039</td>\n",
       "      <td>2.001696e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           n_genes   gwas_or    gwas_p   twas_or    twas_p     de_or  \\\n",
       "module_id                                                              \n",
       "0             4932  1.195380  0.017789  1.515280  0.000007  2.970279   \n",
       "1             3681  0.977286  0.827135  0.569069  0.000021  0.343346   \n",
       "2             3047  1.048770  0.602387  1.686867  0.000001  1.204706   \n",
       "3             2688  1.110855  0.270985  1.294817  0.032309  0.957064   \n",
       "4             1533  1.046916  0.697807  0.839276  0.401404  0.908283   \n",
       "5             1488  1.170352  0.211855  1.185547  0.269657  1.559550   \n",
       "6             1353  0.471844  0.000014  0.564796  0.009668  1.069158   \n",
       "7             1157  0.722880  0.054457  0.494499  0.003970  0.183562   \n",
       "8             1103  0.545154  0.001089  0.556236  0.018219  0.201529   \n",
       "9             1047  1.116042  0.438109  0.860136  0.547137  0.160597   \n",
       "10             871  1.010937  0.932394  0.713797  0.226131  0.184693   \n",
       "11             795  0.989432  1.000000  0.532632  0.037915  0.407333   \n",
       "12             697  0.925349  0.776935  0.843455  0.624397  3.813977   \n",
       "13             554  0.826667  0.460277  0.416404  0.027279  0.229651   \n",
       "14             454  1.997844  0.000249  1.510976  0.097231  0.200021   \n",
       "15             354  1.049345  0.792376  0.436736  0.121018  0.232567   \n",
       "16             266  0.335891  0.021310  0.887646  1.000000  0.498185   \n",
       "17             186  1.127417  0.715198  1.064241  0.813045  0.196921   \n",
       "18             157  0.577684  0.425910  1.268719  0.603295  0.234532   \n",
       "19             150  0.762855  0.688923  0.784920  1.000000  2.562200   \n",
       "\n",
       "                    de_p  twas_fdr_bh  gwas_fdr_bh      de_fdr_bh  \n",
       "module_id                                                          \n",
       "0          1.148426e-124     0.000073     0.085240  2.296851e-123  \n",
       "1           3.497782e-46     0.000137     0.919039   3.497782e-45  \n",
       "2           2.526951e-03     0.000030     0.919039   3.158689e-03  \n",
       "3           5.414012e-01     0.080773     0.677462   5.414012e-01  \n",
       "4           3.126862e-01     0.573434     0.919039   3.474292e-01  \n",
       "5           3.047559e-08     0.414857     0.605301   5.541017e-08  \n",
       "6           4.564584e-01     0.038673     0.000275   4.804825e-01  \n",
       "7           3.232025e-27     0.019850     0.181522   1.616013e-26  \n",
       "8           1.279795e-24     0.060729     0.007257   4.265982e-24  \n",
       "9           9.165784e-27     0.729516     0.836868   3.666314e-26  \n",
       "10          1.073003e-20     0.376884     0.981467   3.065723e-20  \n",
       "11          6.069822e-09     0.084256     1.000000   1.213964e-08  \n",
       "12          1.735665e-44     0.734585     0.919039   1.157110e-43  \n",
       "13          8.001289e-12     0.077939     0.836868   2.000322e-11  \n",
       "14          7.642308e-11     0.194461     0.002492   1.698291e-10  \n",
       "15          8.607329e-08     0.220033     0.919039   1.434555e-07  \n",
       "16          7.312715e-03     1.000000     0.085240   8.603194e-03  \n",
       "17          6.227493e-05     0.903383     0.919039   8.896418e-05  \n",
       "18          6.879887e-04     0.734585     0.836868   9.173182e-04  \n",
       "19          1.301102e-05     1.000000     0.919039   2.001696e-05  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def enrichment_rows():\n",
    "    part = cgowc.part\n",
    "    df = cgowc.expression_df\n",
    "    u = set(cgowc.expression_df.index[3:])\n",
    "    for ii in range(len(part)):\n",
    "        a = set((df.index[x] for x in part[ii]))\n",
    "        yield (ii,\n",
    "               len(part[ii]),\n",
    "               *fet(a, gwas_genes, u),\n",
    "               *fet(a, twas_genes, u),\n",
    "               *fet(a, de_genes, u),\n",
    "               )\n",
    "        \n",
    "    #print(ii, 0 in part[ii], 1 in part[ii], fet(a, gwas_genes, u), fet(a, twas_genes, u), fet(a, de_genes, u), len(part[ii]))\n",
    "    \n",
    "edf1 = pd.DataFrame.from_records(enrichment_rows(), \n",
    "                                 columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', 'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                 index='module_id')\n",
    "edf1['twas_fdr_bh'] = multipletests(edf1['twas_p'], method='fdr_bh')[1]\n",
    "edf1['gwas_fdr_bh'] = multipletests(edf1['gwas_p'], method='fdr_bh')[1]\n",
    "edf1['de_fdr_bh'] = multipletests(edf1['de_p'], method='fdr_bh')[1]\n",
    "\n",
    "edf1[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', 'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('module_enrichment.csv')\n",
    "\n",
    "edf1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:23.205033Z",
     "iopub.status.busy": "2021-09-10T13:42:23.203794Z",
     "iopub.status.idle": "2021-09-10T13:42:23.799291Z",
     "shell.execute_reply": "2021-09-10T13:42:23.798206Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genes</th>\n",
       "      <th>gwas_or</th>\n",
       "      <th>gwas_p</th>\n",
       "      <th>twas_or</th>\n",
       "      <th>twas_p</th>\n",
       "      <th>de_or</th>\n",
       "      <th>de_p</th>\n",
       "      <th>twas_fdr_bh</th>\n",
       "      <th>gwas_fdr_bh</th>\n",
       "      <th>de_fdr_bh</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>module_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4932</td>\n",
       "      <td>1.270454</td>\n",
       "      <td>0.003104</td>\n",
       "      <td>1.549287</td>\n",
       "      <td>5.867860e-06</td>\n",
       "      <td>2.961945</td>\n",
       "      <td>2.549525e-123</td>\n",
       "      <td>0.000039</td>\n",
       "      <td>0.020695</td>\n",
       "      <td>5.099050e-122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3681</td>\n",
       "      <td>1.020123</td>\n",
       "      <td>0.812521</td>\n",
       "      <td>0.519681</td>\n",
       "      <td>3.626710e-06</td>\n",
       "      <td>0.345171</td>\n",
       "      <td>1.176488e-45</td>\n",
       "      <td>0.000036</td>\n",
       "      <td>0.931078</td>\n",
       "      <td>1.176488e-44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3047</td>\n",
       "      <td>1.089280</td>\n",
       "      <td>0.381779</td>\n",
       "      <td>1.766923</td>\n",
       "      <td>3.520444e-07</td>\n",
       "      <td>1.195410</td>\n",
       "      <td>4.129949e-03</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.702226</td>\n",
       "      <td>5.162436e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2688</td>\n",
       "      <td>1.114344</td>\n",
       "      <td>0.300997</td>\n",
       "      <td>1.381112</td>\n",
       "      <td>9.880629e-03</td>\n",
       "      <td>0.951603</td>\n",
       "      <td>4.958298e-01</td>\n",
       "      <td>0.039523</td>\n",
       "      <td>0.670483</td>\n",
       "      <td>4.958298e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1533</td>\n",
       "      <td>0.838594</td>\n",
       "      <td>0.288888</td>\n",
       "      <td>0.693205</td>\n",
       "      <td>7.838740e-02</td>\n",
       "      <td>0.915592</td>\n",
       "      <td>3.534987e-01</td>\n",
       "      <td>0.174194</td>\n",
       "      <td>0.670483</td>\n",
       "      <td>3.927763e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1488</td>\n",
       "      <td>1.166873</td>\n",
       "      <td>0.252847</td>\n",
       "      <td>1.245411</td>\n",
       "      <td>1.818538e-01</td>\n",
       "      <td>1.565331</td>\n",
       "      <td>2.673936e-08</td>\n",
       "      <td>0.303090</td>\n",
       "      <td>0.670483</td>\n",
       "      <td>4.861702e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1353</td>\n",
       "      <td>0.524344</td>\n",
       "      <td>0.000442</td>\n",
       "      <td>0.585507</td>\n",
       "      <td>2.007082e-02</td>\n",
       "      <td>1.070443</td>\n",
       "      <td>4.560327e-01</td>\n",
       "      <td>0.057345</td>\n",
       "      <td>0.004420</td>\n",
       "      <td>4.800344e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1157</td>\n",
       "      <td>0.768958</td>\n",
       "      <td>0.148966</td>\n",
       "      <td>0.467050</td>\n",
       "      <td>3.603034e-03</td>\n",
       "      <td>0.184092</td>\n",
       "      <td>3.015561e-27</td>\n",
       "      <td>0.018015</td>\n",
       "      <td>0.595863</td>\n",
       "      <td>1.507781e-26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1103</td>\n",
       "      <td>0.454820</td>\n",
       "      <td>0.000199</td>\n",
       "      <td>0.532590</td>\n",
       "      <td>1.788755e-02</td>\n",
       "      <td>0.202944</td>\n",
       "      <td>2.563420e-24</td>\n",
       "      <td>0.057345</td>\n",
       "      <td>0.003987</td>\n",
       "      <td>8.544733e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1047</td>\n",
       "      <td>0.977653</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.821551</td>\n",
       "      <td>4.609709e-01</td>\n",
       "      <td>0.162616</td>\n",
       "      <td>2.728412e-26</td>\n",
       "      <td>0.658530</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.091365e-25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>871</td>\n",
       "      <td>1.078994</td>\n",
       "      <td>0.645156</td>\n",
       "      <td>0.783014</td>\n",
       "      <td>4.214106e-01</td>\n",
       "      <td>0.185456</td>\n",
       "      <td>1.543271e-20</td>\n",
       "      <td>0.648324</td>\n",
       "      <td>0.913159</td>\n",
       "      <td>4.409344e-20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>795</td>\n",
       "      <td>1.076547</td>\n",
       "      <td>0.699904</td>\n",
       "      <td>0.583337</td>\n",
       "      <td>9.202699e-02</td>\n",
       "      <td>0.408722</td>\n",
       "      <td>5.968394e-09</td>\n",
       "      <td>0.184054</td>\n",
       "      <td>0.913159</td>\n",
       "      <td>1.193679e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>697</td>\n",
       "      <td>1.025411</td>\n",
       "      <td>0.837970</td>\n",
       "      <td>0.859146</td>\n",
       "      <td>7.015372e-01</td>\n",
       "      <td>3.800434</td>\n",
       "      <td>4.331722e-44</td>\n",
       "      <td>0.892430</td>\n",
       "      <td>0.931078</td>\n",
       "      <td>2.887815e-43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>554</td>\n",
       "      <td>0.886653</td>\n",
       "      <td>0.730527</td>\n",
       "      <td>0.455637</td>\n",
       "      <td>4.576318e-02</td>\n",
       "      <td>0.230371</td>\n",
       "      <td>1.151085e-11</td>\n",
       "      <td>0.114408</td>\n",
       "      <td>0.913159</td>\n",
       "      <td>2.877712e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>454</td>\n",
       "      <td>1.270219</td>\n",
       "      <td>0.301717</td>\n",
       "      <td>1.084679</td>\n",
       "      <td>7.479606e-01</td>\n",
       "      <td>0.208002</td>\n",
       "      <td>3.064875e-10</td>\n",
       "      <td>0.892430</td>\n",
       "      <td>0.670483</td>\n",
       "      <td>6.810834e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>354</td>\n",
       "      <td>1.172825</td>\n",
       "      <td>0.565691</td>\n",
       "      <td>0.477367</td>\n",
       "      <td>1.549964e-01</td>\n",
       "      <td>0.233109</td>\n",
       "      <td>8.470326e-08</td>\n",
       "      <td>0.281812</td>\n",
       "      <td>0.913159</td>\n",
       "      <td>1.411721e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>266</td>\n",
       "      <td>0.299728</td>\n",
       "      <td>0.020586</td>\n",
       "      <td>0.971485</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>0.499903</td>\n",
       "      <td>7.306296e-03</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.102931</td>\n",
       "      <td>8.595642e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>186</td>\n",
       "      <td>0.743496</td>\n",
       "      <td>0.690608</td>\n",
       "      <td>0.701575</td>\n",
       "      <td>8.031866e-01</td>\n",
       "      <td>0.201263</td>\n",
       "      <td>8.869142e-05</td>\n",
       "      <td>0.892430</td>\n",
       "      <td>0.913159</td>\n",
       "      <td>1.267020e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>157</td>\n",
       "      <td>0.515488</td>\n",
       "      <td>0.386225</td>\n",
       "      <td>1.104572</td>\n",
       "      <td>7.864634e-01</td>\n",
       "      <td>0.235950</td>\n",
       "      <td>6.739618e-04</td>\n",
       "      <td>0.892430</td>\n",
       "      <td>0.702226</td>\n",
       "      <td>8.986157e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>150</td>\n",
       "      <td>0.908731</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.855370</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.560943</td>\n",
       "      <td>1.309510e-05</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.014631e-05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           n_genes   gwas_or    gwas_p   twas_or        twas_p     de_or  \\\n",
       "module_id                                                                  \n",
       "0             4932  1.270454  0.003104  1.549287  5.867860e-06  2.961945   \n",
       "1             3681  1.020123  0.812521  0.519681  3.626710e-06  0.345171   \n",
       "2             3047  1.089280  0.381779  1.766923  3.520444e-07  1.195410   \n",
       "3             2688  1.114344  0.300997  1.381112  9.880629e-03  0.951603   \n",
       "4             1533  0.838594  0.288888  0.693205  7.838740e-02  0.915592   \n",
       "5             1488  1.166873  0.252847  1.245411  1.818538e-01  1.565331   \n",
       "6             1353  0.524344  0.000442  0.585507  2.007082e-02  1.070443   \n",
       "7             1157  0.768958  0.148966  0.467050  3.603034e-03  0.184092   \n",
       "8             1103  0.454820  0.000199  0.532590  1.788755e-02  0.202944   \n",
       "9             1047  0.977653  1.000000  0.821551  4.609709e-01  0.162616   \n",
       "10             871  1.078994  0.645156  0.783014  4.214106e-01  0.185456   \n",
       "11             795  1.076547  0.699904  0.583337  9.202699e-02  0.408722   \n",
       "12             697  1.025411  0.837970  0.859146  7.015372e-01  3.800434   \n",
       "13             554  0.886653  0.730527  0.455637  4.576318e-02  0.230371   \n",
       "14             454  1.270219  0.301717  1.084679  7.479606e-01  0.208002   \n",
       "15             354  1.172825  0.565691  0.477367  1.549964e-01  0.233109   \n",
       "16             266  0.299728  0.020586  0.971485  1.000000e+00  0.499903   \n",
       "17             186  0.743496  0.690608  0.701575  8.031866e-01  0.201263   \n",
       "18             157  0.515488  0.386225  1.104572  7.864634e-01  0.235950   \n",
       "19             150  0.908731  1.000000  0.855370  1.000000e+00  2.560943   \n",
       "\n",
       "                    de_p  twas_fdr_bh  gwas_fdr_bh      de_fdr_bh  \n",
       "module_id                                                          \n",
       "0          2.549525e-123     0.000039     0.020695  5.099050e-122  \n",
       "1           1.176488e-45     0.000036     0.931078   1.176488e-44  \n",
       "2           4.129949e-03     0.000007     0.702226   5.162436e-03  \n",
       "3           4.958298e-01     0.039523     0.670483   4.958298e-01  \n",
       "4           3.534987e-01     0.174194     0.670483   3.927763e-01  \n",
       "5           2.673936e-08     0.303090     0.670483   4.861702e-08  \n",
       "6           4.560327e-01     0.057345     0.004420   4.800344e-01  \n",
       "7           3.015561e-27     0.018015     0.595863   1.507781e-26  \n",
       "8           2.563420e-24     0.057345     0.003987   8.544733e-24  \n",
       "9           2.728412e-26     0.658530     1.000000   1.091365e-25  \n",
       "10          1.543271e-20     0.648324     0.913159   4.409344e-20  \n",
       "11          5.968394e-09     0.184054     0.913159   1.193679e-08  \n",
       "12          4.331722e-44     0.892430     0.931078   2.887815e-43  \n",
       "13          1.151085e-11     0.114408     0.913159   2.877712e-11  \n",
       "14          3.064875e-10     0.892430     0.670483   6.810834e-10  \n",
       "15          8.470326e-08     0.281812     0.913159   1.411721e-07  \n",
       "16          7.306296e-03     1.000000     0.102931   8.595642e-03  \n",
       "17          8.869142e-05     0.892430     0.913159   1.267020e-04  \n",
       "18          6.739618e-04     0.892430     0.702226   8.986157e-04  \n",
       "19          1.309510e-05     1.000000     1.000000   2.014631e-05  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def enrichment_rows_nomhc():\n",
    "    part = cgowc.part\n",
    "    df = cgowc.expression_df\n",
    "    u = set(cgowc.expression_df.index[3:]) - mhc_genes\n",
    "    for ii in range(len(part)):\n",
    "        a = set((df.index[x] for x in part[ii])) - mhc_genes\n",
    "        yield (ii,\n",
    "               len(part[ii]),\n",
    "               *fet(a, gwas_genes - mhc_genes, u),\n",
    "               *fet(a, twas_genes - mhc_genes, u),\n",
    "               *fet(a, de_genes - mhc_genes, u),\n",
    "              )\n",
    "        \n",
    "    #print(ii, 0 in part[ii], 1 in part[ii], fet(a, gwas_genes, u), fet(a, twas_genes, u), fet(a, de_genes, u), len(part[ii]))\n",
    "    \n",
    "edf2 = pd.DataFrame.from_records(enrichment_rows_nomhc(), \n",
    "                                columns=['module_id', 'n_genes', 'gwas_or', 'gwas_p', 'twas_or', 'twas_p', 'de_or', 'de_p'],\n",
    "                                index='module_id')\n",
    "edf2['twas_fdr_bh'] = multipletests(edf2['twas_p'], method='fdr_bh')[1]\n",
    "edf2['gwas_fdr_bh'] = multipletests(edf2['gwas_p'], method='fdr_bh')[1]\n",
    "edf2['de_fdr_bh'] = multipletests(edf2['de_p'], method='fdr_bh')[1]\n",
    "\n",
    "edf2[['n_genes', 'gwas_or', 'gwas_p', 'gwas_fdr_bh', 'twas_or', 'twas_p', 'twas_fdr_bh', 'de_or', 'de_p', 'de_fdr_bh']].to_csv('module_enrichment_excluding_mhc_region.csv')\n",
    "\n",
    "edf2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GO enrichment for each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:42:23.826672Z",
     "iopub.status.busy": "2021-09-10T13:42:23.825510Z",
     "iopub.status.idle": "2021-09-10T13:46:11.187400Z",
     "shell.execute_reply": "2021-09-10T13:46:11.186811Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 4806 IDs ... 94%  4,412 of  4,685 study items found in association\n",
      " 97%  4,685 of  4,806 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   7,819 GO terms are associated with  4,412 of  4,806 study items\n",
      "  METHOD fdr_bh:\n",
      "     387 GO terms found significant (< 0.05=alpha) (361 enriched +  26 purified): statsmodels fdr_bh\n",
      "   3,261 study items associated with significant GO IDs (enriched)\n",
      "     583 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 4806 IDs ... 98%  4,583 of  4,685 study items found in association\n",
      " 97%  4,685 of  4,806 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "   1,296 GO terms are associated with  4,583 of  4,806 study items\n",
      "  METHOD fdr_bh:\n",
      "     297 GO terms found significant (< 0.05=alpha) (284 enriched +  13 purified): statsmodels fdr_bh\n",
      "   4,412 study items associated with significant GO IDs (enriched)\n",
      "     492 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 4806 IDs ... 95%  4,459 of  4,685 study items found in association\n",
      " 97%  4,685 of  4,806 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "   2,456 GO terms are associated with  4,459 of  4,806 study items\n",
      "  METHOD fdr_bh:\n",
      "     163 GO terms found significant (< 0.05=alpha) (143 enriched +  20 purified): statsmodels fdr_bh\n",
      "   4,070 study items associated with significant GO IDs (enriched)\n",
      "     329 study items associated with significant GO IDs (purified)\n",
      "    847 items WROTE: module0_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1869 IDs ... 88%    588 of    672 study items found in association\n",
      " 36%    672 of  1,869 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,463 GO terms are associated with    588 of  1,869 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  5 enriched +   1 purified): statsmodels fdr_bh\n",
      "     229 study items associated with significant GO IDs (enriched)\n",
      "       8 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1869 IDs ... 84%    562 of    672 study items found in association\n",
      " 36%    672 of  1,869 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     331 GO terms are associated with    562 of  1,869 study items\n",
      "  METHOD fdr_bh:\n",
      "      14 GO terms found significant (< 0.05=alpha) (  2 enriched +  12 purified): statsmodels fdr_bh\n",
      "     109 study items associated with significant GO IDs (enriched)\n",
      "     334 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1869 IDs ... 64%    430 of    672 study items found in association\n",
      " 36%    672 of  1,869 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     498 GO terms are associated with    430 of  1,869 study items\n",
      "  METHOD fdr_bh:\n",
      "       5 GO terms found significant (< 0.05=alpha) (  3 enriched +   2 purified): statsmodels fdr_bh\n",
      "      61 study items associated with significant GO IDs (enriched)\n",
      "     240 study items associated with significant GO IDs (purified)\n",
      "     25 items WROTE: module1_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2851 IDs ... 91%  2,426 of  2,671 study items found in association\n",
      " 94%  2,671 of  2,851 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   4,940 GO terms are associated with  2,426 of  2,851 study items\n",
      "  METHOD fdr_bh:\n",
      "      57 GO terms found significant (< 0.05=alpha) ( 31 enriched +  26 purified): statsmodels fdr_bh\n",
      "     951 study items associated with significant GO IDs (enriched)\n",
      "     244 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2851 IDs ... 95%  2,537 of  2,671 study items found in association\n",
      " 94%  2,671 of  2,851 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     914 GO terms are associated with  2,537 of  2,851 study items\n",
      "  METHOD fdr_bh:\n",
      "      87 GO terms found significant (< 0.05=alpha) ( 59 enriched +  28 purified): statsmodels fdr_bh\n",
      "   2,124 study items associated with significant GO IDs (enriched)\n",
      "     847 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2851 IDs ... 95%  2,526 of  2,671 study items found in association\n",
      " 94%  2,671 of  2,851 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "   1,688 GO terms are associated with  2,526 of  2,851 study items\n",
      "  METHOD fdr_bh:\n",
      "      35 GO terms found significant (< 0.05=alpha) ( 16 enriched +  19 purified): statsmodels fdr_bh\n",
      "   2,209 study items associated with significant GO IDs (enriched)\n",
      "     137 study items associated with significant GO IDs (purified)\n",
      "    179 items WROTE: module2_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 2024 IDs ... 88%  1,443 of  1,633 study items found in association\n",
      " 81%  1,633 of  2,024 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   3,370 GO terms are associated with  1,443 of  2,024 study items\n",
      "  METHOD fdr_bh:\n",
      "      11 GO terms found significant (< 0.05=alpha) (  4 enriched +   7 purified): statsmodels fdr_bh\n",
      "     280 study items associated with significant GO IDs (enriched)\n",
      "      85 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 2024 IDs ... 94%  1,536 of  1,633 study items found in association\n",
      " 81%  1,633 of  2,024 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     664 GO terms are associated with  1,536 of  2,024 study items\n",
      "  METHOD fdr_bh:\n",
      "      21 GO terms found significant (< 0.05=alpha) (  5 enriched +  16 purified): statsmodels fdr_bh\n",
      "     685 study items associated with significant GO IDs (enriched)\n",
      "     529 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 2024 IDs ... 92%  1,507 of  1,633 study items found in association\n",
      " 81%  1,633 of  2,024 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "   1,270 GO terms are associated with  1,507 of  2,024 study items\n",
      "  METHOD fdr_bh:\n",
      "      11 GO terms found significant (< 0.05=alpha) (  6 enriched +   5 purified): statsmodels fdr_bh\n",
      "     364 study items associated with significant GO IDs (enriched)\n",
      "      14 study items associated with significant GO IDs (purified)\n",
      "     43 items WROTE: module3_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1384 IDs ... 92%  1,188 of  1,294 study items found in association\n",
      " 93%  1,294 of  1,384 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   4,356 GO terms are associated with  1,188 of  1,384 study items\n",
      "  METHOD fdr_bh:\n",
      "     110 GO terms found significant (< 0.05=alpha) (105 enriched +   5 purified): statsmodels fdr_bh\n",
      "     706 study items associated with significant GO IDs (enriched)\n",
      "      10 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1384 IDs ... 97%  1,249 of  1,294 study items found in association\n",
      " 93%  1,294 of  1,384 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     555 GO terms are associated with  1,249 of  1,384 study items\n",
      "  METHOD fdr_bh:\n",
      "      30 GO terms found significant (< 0.05=alpha) ( 28 enriched +   2 purified): statsmodels fdr_bh\n",
      "   1,063 study items associated with significant GO IDs (enriched)\n",
      "      41 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1384 IDs ... 94%  1,210 of  1,294 study items found in association\n",
      " 93%  1,294 of  1,384 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "   1,231 GO terms are associated with  1,210 of  1,384 study items\n",
      "  METHOD fdr_bh:\n",
      "      27 GO terms found significant (< 0.05=alpha) ( 23 enriched +   4 purified): statsmodels fdr_bh\n",
      "     968 study items associated with significant GO IDs (enriched)\n",
      "      55 study items associated with significant GO IDs (purified)\n",
      "    167 items WROTE: module4_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 1375 IDs ... 91%  1,164 of  1,281 study items found in association\n",
      " 93%  1,281 of  1,375 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   3,377 GO terms are associated with  1,164 of  1,375 study items\n",
      "  METHOD fdr_bh:\n",
      "       8 GO terms found significant (< 0.05=alpha) (  1 enriched +   7 purified): statsmodels fdr_bh\n",
      "       5 study items associated with significant GO IDs (enriched)\n",
      "      25 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 1375 IDs ... 96%  1,235 of  1,281 study items found in association\n",
      " 93%  1,281 of  1,375 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     638 GO terms are associated with  1,235 of  1,375 study items\n",
      "  METHOD fdr_bh:\n",
      "      22 GO terms found significant (< 0.05=alpha) ( 13 enriched +   9 purified): statsmodels fdr_bh\n",
      "     902 study items associated with significant GO IDs (enriched)\n",
      "     422 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 1375 IDs ... 93%  1,191 of  1,281 study items found in association\n",
      " 93%  1,281 of  1,375 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "   1,056 GO terms are associated with  1,191 of  1,375 study items\n",
      "  METHOD fdr_bh:\n",
      "       7 GO terms found significant (< 0.05=alpha) (  4 enriched +   3 purified): statsmodels fdr_bh\n",
      "     994 study items associated with significant GO IDs (enriched)\n",
      "       3 study items associated with significant GO IDs (purified)\n",
      "     37 items WROTE: module5_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 779 IDs ... 89%    409 of    461 study items found in association\n",
      " 59%    461 of    779 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,724 GO terms are associated with    409 of    779 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  6 enriched +   0 purified): statsmodels fdr_bh\n",
      "      47 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 779 IDs ... 94%    434 of    461 study items found in association\n",
      " 59%    461 of    779 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     308 GO terms are associated with    434 of    779 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  7 enriched +   2 purified): statsmodels fdr_bh\n",
      "     172 study items associated with significant GO IDs (enriched)\n",
      "      99 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 779 IDs ... 90%    415 of    461 study items found in association\n",
      " 59%    461 of    779 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     565 GO terms are associated with    415 of    779 study items\n",
      "  METHOD fdr_bh:\n",
      "       5 GO terms found significant (< 0.05=alpha) (  4 enriched +   1 purified): statsmodels fdr_bh\n",
      "      19 study items associated with significant GO IDs (enriched)\n",
      "      10 study items associated with significant GO IDs (purified)\n",
      "     20 items WROTE: module6_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 493 IDs ... 89%    108 of    122 study items found in association\n",
      " 25%    122 of    493 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     393 GO terms are associated with    108 of    493 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      21 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 493 IDs ... 80%     98 of    122 study items found in association\n",
      " 25%    122 of    493 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     100 GO terms are associated with     98 of    493 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  0 enriched +   1 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       8 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 493 IDs ... 72%     88 of    122 study items found in association\n",
      " 25%    122 of    493 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     165 GO terms are associated with     88 of    493 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  0 enriched +   1 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "      45 study items associated with significant GO IDs (purified)\n",
      "      3 items WROTE: module7_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 711 IDs ... 88%    428 of    485 study items found in association\n",
      " 68%    485 of    711 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   2,094 GO terms are associated with    428 of    711 study items\n",
      "  METHOD fdr_bh:\n",
      "       3 GO terms found significant (< 0.05=alpha) (  3 enriched +   0 purified): statsmodels fdr_bh\n",
      "      24 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 711 IDs ... 95%    460 of    485 study items found in association\n",
      " 68%    485 of    711 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     359 GO terms are associated with    460 of    711 study items\n",
      "  METHOD fdr_bh:\n",
      "      12 GO terms found significant (< 0.05=alpha) (  9 enriched +   3 purified): statsmodels fdr_bh\n",
      "     260 study items associated with significant GO IDs (enriched)\n",
      "     118 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 711 IDs ... 92%    448 of    485 study items found in association\n",
      " 68%    485 of    711 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     594 GO terms are associated with    448 of    711 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  4 enriched +   0 purified): statsmodels fdr_bh\n",
      "      26 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "     19 items WROTE: module8_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 491 IDs ... 85%    221 of    260 study items found in association\n",
      " 53%    260 of    491 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,013 GO terms are associated with    221 of    491 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 491 IDs ... 90%    233 of    260 study items found in association\n",
      " 53%    260 of    491 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     223 GO terms are associated with    233 of    491 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 491 IDs ... 84%    218 of    260 study items found in association\n",
      " 53%    260 of    491 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     318 GO terms are associated with    218 of    491 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      14 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      1 items WROTE: module9_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 462 IDs ... 88%    202 of    230 study items found in association\n",
      " 50%    230 of    462 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     875 GO terms are associated with    202 of    462 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 462 IDs ... 92%    211 of    230 study items found in association\n",
      " 50%    230 of    462 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     206 GO terms are associated with    211 of    462 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 462 IDs ... 87%    201 of    230 study items found in association\n",
      " 50%    230 of    462 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     301 GO terms are associated with    201 of    462 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING module10_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 428 IDs ... 88%    224 of    254 study items found in association\n",
      " 59%    254 of    428 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,253 GO terms are associated with    224 of    428 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 428 IDs ... 94%    238 of    254 study items found in association\n",
      " 59%    254 of    428 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     227 GO terms are associated with    238 of    428 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 428 IDs ... 92%    233 of    254 study items found in association\n",
      " 59%    254 of    428 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     353 GO terms are associated with    233 of    428 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING module11_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 513 IDs ... 90%    341 of    378 study items found in association\n",
      " 74%    378 of    513 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,536 GO terms are associated with    341 of    513 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  9 enriched +   0 purified): statsmodels fdr_bh\n",
      "      51 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 513 IDs ... 94%    357 of    378 study items found in association\n",
      " 74%    378 of    513 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     328 GO terms are associated with    357 of    513 study items\n",
      "  METHOD fdr_bh:\n",
      "      12 GO terms found significant (< 0.05=alpha) ( 10 enriched +   2 purified): statsmodels fdr_bh\n",
      "     179 study items associated with significant GO IDs (enriched)\n",
      "      78 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 513 IDs ... 91%    344 of    378 study items found in association\n",
      " 74%    378 of    513 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     533 GO terms are associated with    344 of    513 study items\n",
      "  METHOD fdr_bh:\n",
      "       4 GO terms found significant (< 0.05=alpha) (  3 enriched +   1 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       7 study items associated with significant GO IDs (purified)\n",
      "     25 items WROTE: module12_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 250 IDs ... 83%     60 of     72 study items found in association\n",
      " 29%     72 of    250 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     300 GO terms are associated with     60 of    250 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 250 IDs ... 93%     67 of     72 study items found in association\n",
      " 29%     72 of    250 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "      94 GO terms are associated with     67 of    250 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 250 IDs ... 74%     53 of     72 study items found in association\n",
      " 29%     72 of    250 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     118 GO terms are associated with     53 of    250 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING module13_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 386 IDs ... 91%    301 of    329 study items found in association\n",
      " 85%    329 of    386 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "   1,898 GO terms are associated with    301 of    386 study items\n",
      "  METHOD fdr_bh:\n",
      "     197 GO terms found significant (< 0.05=alpha) (195 enriched +   2 purified): statsmodels fdr_bh\n",
      "     237 study items associated with significant GO IDs (enriched)\n",
      "      10 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 386 IDs ... 99%    325 of    329 study items found in association\n",
      " 85%    329 of    386 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     271 GO terms are associated with    325 of    386 study items\n",
      "  METHOD fdr_bh:\n",
      "      41 GO terms found significant (< 0.05=alpha) ( 41 enriched +   0 purified): statsmodels fdr_bh\n",
      "     263 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 386 IDs ... 94%    309 of    329 study items found in association\n",
      " 85%    329 of    386 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     507 GO terms are associated with    309 of    386 study items\n",
      "  METHOD fdr_bh:\n",
      "      34 GO terms found significant (< 0.05=alpha) ( 34 enriched +   0 purified): statsmodels fdr_bh\n",
      "     269 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "    272 items WROTE: module14_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 162 IDs ... 80%     57 of     71 study items found in association\n",
      " 44%     71 of    162 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     223 GO terms are associated with     57 of    162 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 162 IDs ... 92%     65 of     71 study items found in association\n",
      " 44%     71 of    162 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "      71 GO terms are associated with     65 of    162 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 162 IDs ... 80%     57 of     71 study items found in association\n",
      " 44%     71 of    162 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "      95 GO terms are associated with     57 of    162 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING module15_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 229 IDs ... 82%    152 of    186 study items found in association\n",
      " 81%    186 of    229 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     842 GO terms are associated with    152 of    229 study items\n",
      "  METHOD fdr_bh:\n",
      "      19 GO terms found significant (< 0.05=alpha) ( 19 enriched +   0 purified): statsmodels fdr_bh\n",
      "      59 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 229 IDs ... 91%    170 of    186 study items found in association\n",
      " 81%    186 of    229 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     183 GO terms are associated with    170 of    229 study items\n",
      "  METHOD fdr_bh:\n",
      "      20 GO terms found significant (< 0.05=alpha) ( 16 enriched +   4 purified): statsmodels fdr_bh\n",
      "      76 study items associated with significant GO IDs (enriched)\n",
      "      51 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 229 IDs ... 84%    157 of    186 study items found in association\n",
      " 81%    186 of    229 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     224 GO terms are associated with    157 of    229 study items\n",
      "  METHOD fdr_bh:\n",
      "       6 GO terms found significant (< 0.05=alpha) (  5 enriched +   1 purified): statsmodels fdr_bh\n",
      "      19 study items associated with significant GO IDs (enriched)\n",
      "       1 study items associated with significant GO IDs (purified)\n",
      "     45 items WROTE: module16_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 157 IDs ... 99%    132 of    133 study items found in association\n",
      " 85%    133 of    157 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     307 GO terms are associated with    132 of    157 study items\n",
      "  METHOD fdr_bh:\n",
      "      15 GO terms found significant (< 0.05=alpha) ( 14 enriched +   1 purified): statsmodels fdr_bh\n",
      "     110 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 157 IDs ... 98%    131 of    133 study items found in association\n",
      " 85%    133 of    157 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     115 GO terms are associated with    131 of    157 study items\n",
      "  METHOD fdr_bh:\n",
      "      26 GO terms found significant (< 0.05=alpha) ( 23 enriched +   3 purified): statsmodels fdr_bh\n",
      "     124 study items associated with significant GO IDs (enriched)\n",
      "      14 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 157 IDs ... 68%     90 of    133 study items found in association\n",
      " 85%    133 of    157 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     100 GO terms are associated with     90 of    157 study items\n",
      "  METHOD fdr_bh:\n",
      "       9 GO terms found significant (< 0.05=alpha) (  9 enriched +   0 purified): statsmodels fdr_bh\n",
      "      70 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "     50 items WROTE: module17_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 97 IDs ... 80%     33 of     41 study items found in association\n",
      " 42%     41 of     97 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     196 GO terms are associated with     33 of     97 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 97 IDs ... 90%     37 of     41 study items found in association\n",
      " 42%     41 of     97 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "      50 GO terms are associated with     37 of     97 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 97 IDs ... 83%     34 of     41 study items found in association\n",
      " 42%     41 of     97 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "      70 GO terms are associated with     34 of     97 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      0 items. NOT WRITING module18_go_enrichment.tsv\n",
      "\n",
      "Run BP Gene Ontology Analysis: current study set of 115 IDs ... 88%     92 of    104 study items found in association\n",
      " 90%    104 of    115 study items found in population(20652)\n",
      "Calculating 12,417 uncorrected p-values using fisher_scipy_stats\n",
      "  12,417 GO terms are associated with 18,501 of 20,652 population items\n",
      "     474 GO terms are associated with     92 of    115 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run CC Gene Ontology Analysis: current study set of 115 IDs ... 95%     99 of    104 study items found in association\n",
      " 90%    104 of    115 study items found in population(20652)\n",
      "Calculating 1,754 uncorrected p-values using fisher_scipy_stats\n",
      "   1,754 GO terms are associated with 19,433 of 20,652 population items\n",
      "     133 GO terms are associated with     99 of    115 study items\n",
      "  METHOD fdr_bh:\n",
      "       1 GO terms found significant (< 0.05=alpha) (  1 enriched +   0 purified): statsmodels fdr_bh\n",
      "      12 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "\n",
      "Run MF Gene Ontology Analysis: current study set of 115 IDs ... 91%     95 of    104 study items found in association\n",
      " 90%    104 of    115 study items found in population(20652)\n",
      "Calculating 4,419 uncorrected p-values using fisher_scipy_stats\n",
      "   4,419 GO terms are associated with 18,194 of 20,652 population items\n",
      "     200 GO terms are associated with     95 of    115 study items\n",
      "  METHOD fdr_bh:\n",
      "       0 GO terms found significant (< 0.05=alpha) (  0 enriched +   0 purified): statsmodels fdr_bh\n",
      "       0 study items associated with significant GO IDs (enriched)\n",
      "       0 study items associated with significant GO IDs (purified)\n",
      "      1 items WROTE: module19_go_enrichment.tsv\n"
     ]
    }
   ],
   "source": [
    "# Run pipeline for each cluster\n",
    "for cluster_id in range(len(cgowc.part)):\n",
    "    cgowc.pipeline(cluster_id, 'module%d' % cluster_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-10T13:46:11.193407Z",
     "iopub.status.busy": "2021-09-10T13:46:11.192539Z",
     "iopub.status.idle": "2021-09-10T13:46:11.196757Z",
     "shell.execute_reply": "2021-09-10T13:46:11.196260Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ceph/projects/v4_phase3_paper/analysis/gnvae/disvae/model/embedding/leiden/_m'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
