{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting eQTLs, increase font sizes\n",
    "\n",
    "### Kynon Jade Benjamin and Apu√£ Paquola"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import functools\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from plotnine import *\n",
    "from pandas_plink import read_plink\n",
    "from warnings import filterwarnings\n",
    "from matplotlib.cbook import mplDeprecation\n",
    "\n",
    "filterwarnings(\"ignore\",category=mplDeprecation)\n",
    "filterwarnings('ignore', category=UserWarning, module='plotnine.*')\n",
    "filterwarnings('ignore', category=DeprecationWarning, module='plotnine.*')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configuration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature = \"junctions\"; tissue = \"caudate\"\n",
    "config = {\n",
    "    'biomart_file': '../_h/biomart.csv',\n",
    "    'residual_expression_file': \"/ceph/projects/v4_phase3_paper/analysis/eqtl_analysis/all/%s/\" % feature+\\\n",
    "    \"expression_gct/covariates/residualized_expression/_m/%s_residualized_expression.csv\" % feature,\n",
    "    'residual_expression_file2': \"/ceph/projects/v4_phase3_paper/analysis/eqtl_analysis/all/%s/\" % feature+\\\n",
    "    \"expression_gct/prepare_expression/_m/%s.expression.bed.gz\" % feature,\n",
    "    'phenotype_file': '/ceph/projects/v4_phase3_paper/inputs/phenotypes/_m/merged_phenotypes.csv',\n",
    "    'plink_file_prefix': '/ceph/projects/v4_phase3_paper/inputs/genotypes/_m/LIBD_Brain_TopMed',\n",
    "    'eqtl_output_file': '/ceph/projects/v4_phase3_paper/analysis/eqtl_analysis/all/%s/' % feature +\\\n",
    "    'expression_gct/prepare_expression/fastqtl_permutation/_m/Brainseq_LIBD.genes.txt.gz',\n",
    "    'gwas_snp_file': '../../summary_table/_m/Brainseq_LIBD_caudate_4features_PGC2.signifpairs.txt.gz'\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def feature_map(feature):\n",
    "    return {\"genes\": \"Gene\", \"transcripts\": \"Transcript\", \n",
    "            \"exons\": \"Exon\", \"junctions\": \"Junction\"}[feature]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Expression functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_biomart_df():\n",
    "    biomart = pd.read_csv(config['biomart_file'], index_col=0)\n",
    "    biomart['description'] = biomart['description'].str.replace('\\[Source.*$','', regex=True)\n",
    "    return biomart\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df():\n",
    "    return pd.read_csv(config['residual_expression_file'], index_col=0).transpose()\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df_v2():\n",
    "    return pd.read_csv(config[\"residual_expression_file2\"], sep='\\t', index_col=3).drop([\"#chr\", \"start\", \"end\"], axis=1).transpose()\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_pheno_df():\n",
    "    pheno_df = pd.read_csv(config['phenotype_file']).set_index(\"BrNum\").loc[:, [\"RNum\", \"Sex\", \"Dx\", \"Region\"]]\n",
    "    return pheno_df[(pheno_df[\"Region\"] == \"Caudate\")].drop(\"Region\", axis=1)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_expression_and_pheno_df():\n",
    "    return pd.merge(get_pheno_df(), get_residual_expression_df(),\n",
    "                    left_index=True, right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gene_id_df():\n",
    "    return pd.DataFrame({'gene_id': get_residual_expression_df().columns,\n",
    "                         'ensembl_gene_id': get_residual_expression_df().columns.str.replace('\\..+$','', regex=True)})\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_info_from_symbol(gene_symbol):\n",
    "    return get_biomart_df()[get_biomart_df()['external_gene_name']==gene_symbol]\\\n",
    "        .merge(get_gene_id_df(), on='ensembl_gene_id', how='left')\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def gene_id_from_symbol(gene_symbol):\n",
    "    df = gene_info_from_symbol(gene_symbol)\n",
    "    assert df.shape[0] == 1\n",
    "    return df[['gene_id']].iloc[0].values[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRD2 functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_drd2_junction_annotation(junction_id):\n",
    "    return {\n",
    "        'chr11:113424683-113474229(-)': \"DRD2 junction 1L-2\",\n",
    "        \"chr11:113424683-113475075(-)\": \"DRD2 junction 1-2\",\n",
    "        \"chr11:113418137-113424366(-)\": \"DRD2 junction 2-3\",\n",
    "        \"chr11:113417000-113418026(-)\": \"DRD2 junction 3-4\",\n",
    "        \"chr11:113415612-113416862(-)\": \"DRD2 junction 4-5\",\n",
    "        \"chr11:113414462-113415420(-)\": \"DRD2 junction 5-6\",\n",
    "        \"chr11:113412884-113415420(-)\": \"DRD2 junction 5-7\",\n",
    "        \"chr11:113412884-113414374(-)\": \"DRD2 junction 6-7\",\n",
    "        \"chr11:113410921-113412555(-)\": \"DRD2 junction 7-8\",}[junction_id]\n",
    "\n",
    "\n",
    "def get_drd2_junctions():\n",
    "    with subprocess.Popen(\"\"\"cat <(head -1 /ceph/projects/v4_phase3_paper/analysis/differential_expression/_m/junctions/diffExpr_szVctl_full.txt) \\\n",
    "    <(grep -i drd2 /ceph/projects/v4_phase3_paper/analysis/differential_expression/_m/junctions/diffExpr_szVctl_full.txt)\"\"\", \n",
    "                          shell=True, stdout=subprocess.PIPE) as p:\n",
    "        drd2_df = pd.read_csv(p.stdout, sep='\\t', index_col=0)\n",
    "    return drd2_df.reset_index().rename(columns={\"index\": \"Feature\"})\n",
    "\n",
    "\n",
    "def get_drd2():\n",
    "    drdj = get_drd2_junctions()[get_drd2_junctions()[\"gencodeTx\"].str.contains(\"ENST00000362072.7|ENST00000346454.7\")]\n",
    "    return get_eqtl_df()[(get_eqtl_df()[\"gene_id\"].isin(drdj.Feature))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotype and eQTL functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def letter_snp(number, a0, a1):\n",
    "    '''\n",
    "    Example:\n",
    "    letter_snp(0, 'A', 'G') is 'AA'\n",
    "    letter_snp(1, 'A', 'G') is 'AG'\n",
    "    letter_snp(2, 'A', 'G') is 'GG'\n",
    "    '''\n",
    "    if np.isnan(number):\n",
    "        return np.nan\n",
    "    if len(a0)==1 and len(a1)==1:\n",
    "        sep = ''\n",
    "    else:\n",
    "        sep = ' '\n",
    "    return sep.join(sorted([a0]*int(number) + [a1]*(2-int(number))))\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_plink_tuple():\n",
    "    '''\n",
    "    Usage: (bim, fam, bed) = get_plink_tuple()\n",
    "    '''\n",
    "    return read_plink(config['plink_file_prefix'])\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_eqtl_df():\n",
    "    \"\"\"\n",
    "    This function loads permutation eQTL analysis.\n",
    "    \"\"\"\n",
    "    return pd.read_csv(config[\"eqtl_output_file\"], sep='\\t')\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_snps():\n",
    "    \"\"\"\n",
    "    This function load the GWAS SNPs that have been mapped to BrainSeq.\n",
    "    \"\"\"\n",
    "    gwas_df = pd.read_csv(config['gwas_snp_file'], sep='\\t', index_col=0)\n",
    "    return gwas_df[(gwas_df[\"Type\"] == feature_map(feature))]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_risk_allele(snp_id):\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['OR'].iloc[0] > 1:\n",
    "        ra = gwas_snp['A1'].iloc[0]\n",
    "    else:\n",
    "        ra = gwas_snp['A2'].iloc[0]\n",
    "    return ra\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def subset_bed():\n",
    "    \"\"\"\n",
    "    This subsets the bed and bim file and returns the new subsetted\n",
    "    data with shared brain_ids.\n",
    "    \n",
    "    This is to speed up accessing the bed file.\n",
    "    \"\"\"\n",
    "    (bim, fam, bed) = get_plink_tuple()\n",
    "    brain_ids = list(set(get_expression_and_pheno_df().index).intersection(set(fam['fid'])))\n",
    "    fam_pos = list(fam[(fam[\"fid\"].isin(brain_ids))].drop_duplicates(subset=\"fid\").loc[:, 'i'])\n",
    "    unique_snps0 = get_eqtl_df().variant_id.unique()\n",
    "    unique_snps1 = get_gwas_snps()[get_gwas_snps()[\"gene_id\"].isin(get_drd2().gene_id)].reset_index().variant_id.unique()\n",
    "    unique_snps = list(set(unique_snps0).union(set(unique_snps1)))\n",
    "    snp_info = bim[(bim[\"snp\"].isin(unique_snps))].copy()\n",
    "    snp_pos = list(snp_info.loc[:, \"i\"])\n",
    "    new_bed = bed[snp_pos].compute()[:,fam_pos]\n",
    "    new_bim = bim[(bim[\"i\"].isin(snp_pos))].reset_index(drop=True)\n",
    "    new_bim['ii'] = new_bim.index\n",
    "    return new_bed, new_bim, brain_ids\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the same as in the plink files.\n",
    "    \n",
    "    Example: \n",
    "    get_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             0              GG    0\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             2              AA    2\\nAA\n",
    "    Br5182             2              AA    2\\nAA\n",
    "    '''\n",
    "    bed, bim, brain_ids = subset_bed()\n",
    "    snp_info = bim[bim['snp']==snp_id]\n",
    "    snp_pos = snp_info.iloc[0]['ii']\n",
    "    dfsnp = pd.DataFrame(bed[[snp_pos]], columns=brain_ids, index=[snp_id + '_num']).transpose().dropna()\n",
    "    my_letter_snp = functools.partial(letter_snp, a0=snp_info.iloc[0]['a0'], a1=snp_info.iloc[0]['a1'])\n",
    "    # the 2 - in next line is to workaround a possible bug in pandas_plink? a1 and a0 inverted\n",
    "    dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']].astype('int')\n",
    "    dfsnp[snp_id + '_letter'] = dfsnp[snp_id + '_num'].apply(my_letter_snp)\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_ordered_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the number of risk alleles according to GWAS.\n",
    "    \n",
    "    Example: \n",
    "    get_gwas_ordered_snp_df('rs653953', 'genes').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             2              GG    2\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             0              AA    0\\nAA\n",
    "    Br5182             0              AA    0\\nAA\n",
    "    '''\n",
    "    pgc = get_gwas_snps()\n",
    "    dfsnp = get_snp_df(snp_id).copy()\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    if gwas_snp['pgc2_a1_same_as_our_counted'].iloc[0]:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            pass\n",
    "        else:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "    else:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "        else:\n",
    "            pass\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func):\n",
    "    pheno_columns = list(get_pheno_df().columns)\n",
    "    expr_df = get_expression_and_pheno_df()[pheno_columns + [gene_id]]\n",
    "    snp_df =  snp_df_func(snp_id)\n",
    "    return expr_df.merge(snp_df, left_index=True, right_index=True)\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot_impl(snp_id, gene_id, snp_df_func):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id)) \\\n",
    "    + geom_boxplot(fill=\"red\", alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(width=0.25, stroke=0, alpha=0.6) \\\n",
    "    + ylim(y0, y1) + theme_bw(base_size=15) \\\n",
    "    + theme(panel_grid=element_blank(), \n",
    "            axis_title=element_text(face=\"bold\"))\n",
    "    return p\n",
    "    \n",
    "\n",
    "def simple_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_snp_df)\n",
    "\n",
    "\n",
    "def simple_gwas_ordered_snp_expression_plot(snp_id, gene_id):\n",
    "    return simple_snp_expression_plot_impl(snp_id, gene_id, get_gwas_ordered_snp_df)\n",
    "\n",
    "\n",
    "def get_gene_symbol(gene_id, biomart=get_biomart_df()):\n",
    "    ensge = re.sub('\\..+$','', gene_id)\n",
    "    ggg = biomart[biomart['ensembl_gene_id']==ensge]\n",
    "    if ggg.shape[0]==0:\n",
    "        return '', ''\n",
    "    gs = ggg['external_gene_name'].values[0]\n",
    "    de = ggg['description'].values[0]\n",
    "    if type(de)!=str:\n",
    "        de = ''\n",
    "    de = re.sub('\\[Source:.*$','',de)\n",
    "    return gs, de\n",
    "\n",
    "\n",
    "def get_gwas_snp(snp_id):\n",
    "    gwas = get_gwas_snps().reset_index()\n",
    "    r = gwas[gwas['variant_id']==snp_id]\n",
    "    assert len(r) == 1\n",
    "    return r\n",
    "\n",
    "\n",
    "def gwas_annotation(snp_id):\n",
    "    return 'SZ GWAS pvalue: %.1e' % get_gwas_snp(snp_id).iloc[0]['P']\n",
    "\n",
    "\n",
    "def eqtl_annotation(snp_id, gene_id):\n",
    "    r = get_eqtl_df()[(get_eqtl_df()['variant_id']==snp_id) & \n",
    "                      (get_eqtl_df()['gene_id']==gene_id)]\n",
    "    try:\n",
    "        assert len(r)==1\n",
    "        return 'eQTL q-value: %.1e' % r.iloc[0]['qval']\n",
    "    except AssertionError:\n",
    "        eqtl_df = get_gwas_snps().reset_index()\n",
    "        r = eqtl_df[(eqtl_df['variant_id']==snp_id) & (eqtl_df['gene_id']==gene_id)]\n",
    "        return 'eQTL nominal p-value: %.1e' % r.iloc[0]['pval_nominal']\n",
    "    \n",
    "\n",
    "\n",
    "def risk_allele_annotation(snp_id):\n",
    "    return 'SZ risk allele: %s' % get_risk_allele(snp_id)\n",
    "\n",
    "\n",
    "def annotated_eqtl_plot(snp_id, gene_id, title_prefix=[]):\n",
    "    p = simple_snp_expression_plot(snp_id, gene_id)\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join(title_prefix + [eqtl_annotation(snp_id, gene_id)])\n",
    "    p += ggtitle(title) + ylab('Residualized Expression') \n",
    "    return p\n",
    "\n",
    "\n",
    "def gwas_annotated_eqtl_plot(snp_id, gene_id, title_prefix=[]):\n",
    "    p = simple_gwas_ordered_snp_expression_plot(snp_id, gene_id)\n",
    "    gene_symbol, gene_description = get_gene_symbol(gene_id)\n",
    "    title =\"\\n\".join(title_prefix + [eqtl_annotation(snp_id, gene_id),\n",
    "                     gwas_annotation(snp_id),\n",
    "                     risk_allele_annotation(snp_id)\n",
    "                     ])\n",
    "    p += ggtitle(title) + ylab('Residualized Expression') \n",
    "    return p\n",
    "\n",
    "\n",
    "def save_plot(p, fn):\n",
    "    for ext in ['png', 'pdf', 'svg']:\n",
    "        p.save(fn + '.' + ext)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot eQTLs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRD2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drd2_df = get_drd2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in drd2_df.reset_index(drop=True).itertuples():\n",
    "    en = get_drd2_junction_annotation(x.gene_id).replace(' ','_').replace(\"-\", \"_\")\n",
    "    anno = get_drd2_junction_annotation(x.gene_id)\n",
    "    filename = \"drd2_eqtl_{}\".format(en)\n",
    "    p = annotated_eqtl_plot(x.variant_id, x.gene_id, [anno, x.gene_id])\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id, anno)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Top 5 eQTL with GWAS significant index SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drd2_gwas_eqtl_df = get_gwas_snps().reset_index().merge(drd2_df.loc[:, [\"gene_id\", \"qval\"]], on=[\"gene_id\"]).groupby([\"gene_id\", \"is_index_snp\"]).first().reset_index()\n",
    "drd2_gwas_eqtl_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in drd2_gwas_eqtl_df[(drd2_gwas_eqtl_df[\"is_index_snp\"])].itertuples():\n",
    "    en = get_drd2_junction_annotation(x.gene_id).replace(' ','_').replace(\"-\", \"_\")\n",
    "    anno = get_drd2_junction_annotation(x.gene_id)\n",
    "    filename = \"drd2_eqtl_gwas_index_snp_{}\".format(en)\n",
    "    p = gwas_annotated_eqtl_plot(x.variant_id, x.gene_id, [anno, x.gene_id])\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id, anno)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in drd2_gwas_eqtl_df[~(drd2_gwas_eqtl_df[\"is_index_snp\"])].itertuples():\n",
    "    en = get_drd2_junction_annotation(x.gene_id).replace(' ','_').replace(\"-\", \"_\")\n",
    "    anno = get_drd2_junction_annotation(x.gene_id)\n",
    "    filename = \"drd2_eqtl_gwas_{}\".format(en)\n",
    "    p = gwas_annotated_eqtl_plot(x.variant_id, x.gene_id, [anno, x.gene_id])\n",
    "    print(filename, x.Index, x.variant_id, x.gene_id, anno)\n",
    "    print(p)\n",
    "    save_plot(p, filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
