{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Enrichment and Overlap of PGC2+CLOZUK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import os, errno\n",
    "import functools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from plotnine import *\n",
    "from pandas_plink import read_plink\n",
    "from warnings import filterwarnings\n",
    "from matplotlib.cbook import mplDeprecation\n",
    "from scipy.stats import fisher_exact, binom_test\n",
    "\n",
    "filterwarnings(\"ignore\",category=mplDeprecation)\n",
    "filterwarnings('ignore', category=UserWarning, module='plotnine.*')\n",
    "filterwarnings('ignore', category=DeprecationWarning, module='plotnine.*')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Config and Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "config = {\n",
    "    'biomart_file': '../_h/biomart.csv',\n",
    "    'phenotype_file': '/ceph/projects/v4_phase3_paper/inputs/phenotypes/_m/merged_phenotypes.csv',\n",
    "    'plink_file_prefix': '/ceph/projects/v4_phase3_paper/inputs/genotypes/_m/LIBD_Brain_TopMed',\n",
    "    'gwas_snp_file': '/ceph/projects/v4_phase3_paper/inputs/sz_gwas/pgc2_clozuk/map_phase3/_m/libd_hg38_pgc2sz_snps.tsv'\n",
    "}\n",
    "\n",
    "config_feature = {\n",
    "    'de_file': '../../differential_expression/_m/junctions/diffExpr_szVctl_full.txt',\n",
    "    'residual_expression_file': '../../differential_expression/_m/junctions/residualized_expression.tsv',\n",
    "    'fastqtl_output_file': '../../eqtl/caudate/summary_table/_m/Brainseq_LIBD_caudate_4features.signifpairs.txt.gz',\n",
    "}\n",
    "\n",
    "feature = \"junctions\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def feature_map(feature):\n",
    "    return {\"genes\": \"Gene\", \"transcripts\": \"Transcript\", \n",
    "            \"exons\": \"Exon\", \"junctions\": \"Junction\"}[feature]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_de_df():\n",
    "    \"\"\"\n",
    "    Load DE analysis\n",
    "    \"\"\"\n",
    "    return pd.read_csv(config_feature['de_file'], sep='\\t', index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_eqtl_df():\n",
    "    eqtl_df = pd.read_csv(config_feature['fastqtl_output_file'], sep='\\t')\n",
    "    return eqtl_df[(eqtl_df[\"Type\"] == feature_map(feature))]\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_snps():\n",
    "    return pd.read_csv(config['gwas_snp_file'], sep='\\t', index_col=0, low_memory=False)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_integration_df():\n",
    "    return get_gwas_snps().merge(get_eqtl_df(), left_on='our_snp_id', right_on='variant_id', \n",
    "                                 suffixes=['_PGC2', '_eQTL'])\\\n",
    "                          .merge(get_de_df(), left_on='gene_id', right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_residual_expression_df():\n",
    "    return pd.read_csv(config_feature['residual_expression_file'], \n",
    "                       sep='\\t', index_col=0).transpose()\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_pheno_df():\n",
    "    return pd.read_csv(config['phenotype_file'], index_col=0)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_biomart_df():\n",
    "    biomart = pd.read_csv(config['biomart_file'])\n",
    "    biomart['description'] = biomart['description'].str.replace('\\[Source.*$','', regex=True)\n",
    "    return biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def agree_direction(row):\n",
    "    return [-1, 1][row['pgc2_a1_same_as_our_counted']] * np.sign(row['OR'] - 1) * np.sign(row['slope']) * np.sign(row['t'])\n",
    "\n",
    "\n",
    "def save_plot(p, fn):\n",
    "    for ext in ['png', 'pdf', 'svg']:\n",
    "        p.save(fn + '.' + ext)\n",
    "        \n",
    "\n",
    "def letter_snp(number, a0, a1):\n",
    "    '''\n",
    "    Example:\n",
    "    letter_snp(0, 'A', 'G') is 'AA'\n",
    "    letter_snp(1, 'A', 'G') is 'AG'\n",
    "    letter_snp(2, 'A', 'G') is 'GG'\n",
    "    \n",
    "    '''\n",
    "    if np.isnan(number):\n",
    "        return np.nan\n",
    "    if len(a0)==1 and len(a1)==1:\n",
    "        sep = ''\n",
    "    else:\n",
    "        sep = ' '\n",
    "    return sep.join(sorted([a0]*int(number) + [a1]*(2-int(number))))\n",
    "\n",
    "\n",
    "def get_gwas_snp(snp_id):\n",
    "    gwas = get_gwas_snps()\n",
    "    r = gwas[gwas['our_snp_id']==snp_id]\n",
    "    assert len(r) == 1\n",
    "    return r"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "@functools.lru_cache()\n",
    "def get_expression_and_pheno_df():\n",
    "    return pd.merge(get_pheno_df(), get_residual_expression_df(), left_index=True, right_index=True)\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_plink_tuple():\n",
    "    '''\n",
    "    Usage: (bim, fam, bed) = get_plink_tuple()\n",
    "    '''\n",
    "    return read_plink(config['plink_file_prefix'])\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def subset_bed():\n",
    "    \"\"\"\n",
    "    This subsets the bed and bim file and returns the new subsetted\n",
    "    data with shared brain_ids.\n",
    "    \n",
    "    This is to speed up accessing the bed file.\n",
    "    \"\"\"\n",
    "    (bim, fam, bed) = get_plink_tuple()\n",
    "    brain_ids = list(set(get_expression_and_pheno_df()['BrNum']).intersection(set(fam['fid'])))\n",
    "    fam_pos = list(fam[(fam[\"fid\"].isin(brain_ids))].drop_duplicates(subset=\"fid\").loc[:, 'i'])\n",
    "    unique_snps = get_eqtl_df().variant_id.unique()\n",
    "    snp_info = bim[(bim[\"snp\"].isin(unique_snps))].copy()\n",
    "    snp_pos = list(snp_info.loc[:, \"i\"])\n",
    "    new_bed = bed[snp_pos].compute()[:,fam_pos]\n",
    "    new_bim = bim[(bim[\"i\"].isin(snp_pos))].reset_index(drop=True)\n",
    "    new_bim['ii'] = new_bim.index\n",
    "    return new_bed, new_bim, brain_ids\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the same as in the plink files.\n",
    "    \n",
    "    Example: \n",
    "    get_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             0              GG    0\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             2              AA    2\\nAA\n",
    "    Br5182             2              AA    2\\nAA\n",
    "    '''\n",
    "    bed, bim, brain_ids = subset_bed()\n",
    "    snp_info = bim[bim['snp']==snp_id]\n",
    "    snp_pos = snp_info.iloc[0]['ii']\n",
    "    dfsnp = pd.DataFrame(bed[[snp_pos]], columns=brain_ids, index=[snp_id + '_num']).transpose().dropna()\n",
    "    my_letter_snp = functools.partial(letter_snp, a0=snp_info.iloc[0]['a0'], a1=snp_info.iloc[0]['a1'])\n",
    "    # the 2 - in next line is to workaround a possible bug in pandas_plink? a1 and a0 inverted\n",
    "    dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']].astype('int')\n",
    "    dfsnp[snp_id + '_letter'] = dfsnp[snp_id + '_num'].apply(my_letter_snp)\n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_gwas_ordered_snp_df(snp_id):\n",
    "    '''\n",
    "    Returns a dataframe containing the genotype on snp snp_id.\n",
    "    The allele count is the number of risk alleles according to GWAS.\n",
    "    \n",
    "    Example: \n",
    "    get_gwas_ordered_snp_df('rs653953').head(5)\n",
    "    \n",
    "            rs653953_num rs653953_letter rs653953\n",
    "    Br5168             2              GG    2\\nGG\n",
    "    Br2582             1              AG    1\\nAG\n",
    "    Br2378             1              AG    1\\nAG\n",
    "    Br5155             0              AA    0\\nAA\n",
    "    Br5182             0              AA    0\\nAA\n",
    "    '''\n",
    "    pgc = get_gwas_snps()\n",
    "    dfsnp = get_snp_df(snp_id).copy()\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    \n",
    "    if gwas_snp['pgc2_a1_same_as_our_counted'].iloc[0]:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            pass\n",
    "        else:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "    else:\n",
    "        if gwas_snp['OR'].iloc[0] > 1:\n",
    "            dfsnp[[snp_id + '_num']] = 2 - dfsnp[[snp_id + '_num']]\n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "    dfsnp[snp_id] = (dfsnp[snp_id + '_num'].astype('str') + '\\n' + \n",
    "                     dfsnp[snp_id + '_letter'].astype('str')).astype('category')\n",
    "    return dfsnp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_symbol(gene_id, biomart=get_biomart_df()):\n",
    "    ensge = re.sub('\\..+$','', gene_id)\n",
    "    ggg = biomart[biomart['ensembl_gene_id']==ensge]\n",
    "    if ggg.shape[0]==0:\n",
    "        return '', ''\n",
    "\n",
    "    gs = ggg['external_gene_name'].values[0]\n",
    "    de = ggg['description'].values[0]\n",
    "    if type(de)!=str:\n",
    "        de = ''\n",
    "        \n",
    "    de = re.sub('\\[Source:.*$','',de)\n",
    "    return gs, de\n",
    "\n",
    "\n",
    "@functools.lru_cache()\n",
    "def get_risk_allele(snp_id):\n",
    "    gwas_snp = get_gwas_snp(snp_id)\n",
    "    \n",
    "    if gwas_snp['OR'].iloc[0] > 1:\n",
    "        ra = gwas_snp['A1'].iloc[0]\n",
    "    else:\n",
    "        ra = gwas_snp['A2'].iloc[0]\n",
    "    \n",
    "    return ra\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func):\n",
    "    pheno_columns = list(get_pheno_df().columns)\n",
    "    expr_df = get_expression_and_pheno_df()[pheno_columns + [gene_id]]\n",
    "    snp_df =  snp_df_func(snp_id)\n",
    "    return expr_df.merge(snp_df, left_on='BrNum', right_index=True)\n",
    "\n",
    "\n",
    "def simple_snp_expression_pheno_plot_impl(snp_id, gene_id, snp_df_func, pheno_var):\n",
    "    df = get_snp_gene_pheno_df(snp_id, gene_id, snp_df_func)\n",
    "    df['Dx'] = df.Dx.astype('category').cat.rename_categories({'Control': 'CTL', 'Schizo': 'SZ'})\n",
    "    y0 = df[gene_id].quantile(.01) - 0.26\n",
    "    y1 = df[gene_id].quantile(.99) + 0.26\n",
    "    pjd = position_jitterdodge(jitter_width=0.27)\n",
    "    p = ggplot(df, aes(x=snp_id, y=gene_id, fill=pheno_var)) \\\n",
    "    + geom_boxplot(alpha=0.4, outlier_alpha=0) \\\n",
    "    + geom_jitter(position=pjd, stroke=0, alpha=0.6) + ylim(y0, y1) \\\n",
    "    + labs(y='Residualized expression', fill='Diagnosis') \\\n",
    "    + theme_bw(base_size=20)\\\n",
    "    + theme(legend_title=element_text(face='bold'), \n",
    "            panel_grid_major=element_blank(), \n",
    "            panel_grid_minor=element_blank())\n",
    "    return p\n",
    "\n",
    "\n",
    "def simple_gwas_ordered_snp_expression_pheno_plot(snp_id, gene_id, pheno_var):\n",
    "    return simple_snp_expression_pheno_plot_impl(snp_id, gene_id, get_gwas_ordered_snp_df, pheno_var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gwas_annotation(snp_id):\n",
    "    return 'SZ GWAS pvalue: %.1e' % get_gwas_snp(snp_id).iloc[0]['P']\n",
    "\n",
    "\n",
    "def eqtl_annotation(snp_id, gene_id):\n",
    "    r = get_eqtl_df()[(get_eqtl_df()['variant_id']==snp_id) & \n",
    "                      (get_eqtl_df()['gene_id']==gene_id)]\n",
    "    assert len(r)==1\n",
    "    return 'eQTL nominal p-value: %.1e' % r.iloc[0]['pval_nominal']\n",
    "\n",
    "\n",
    "def de_annotation(gene_id):\n",
    "    de_df = get_de_df()\n",
    "    de_df['Feature'] = de_df.index\n",
    "    g = de_df[(de_df['Feature'] == gene_id)]\n",
    "    return 'DE adj.P.Val: %.3f' % g.iloc[0]['adj.P.Val']\n",
    "\n",
    "\n",
    "def risk_allele_annotation(snp_id):\n",
    "    return 'SZ risk allele: %s' % get_risk_allele(snp_id)\n",
    "\n",
    "\n",
    "def gwas_annotated_eqtl_pheno_plot(snp_id, gene_id, pheno_var):\n",
    "    p = simple_gwas_ordered_snp_expression_pheno_plot(snp_id, gene_id, pheno_var)\n",
    "    de_df = get_de_df()\n",
    "    de_df['Feature'] = de_df.index\n",
    "    de_df = de_df[(de_df['Feature'] == gene_id)]\n",
    "    gene_symbol = de_df.iloc[0]['newGeneSymbol']\n",
    "    \n",
    "    title =\"\\n\".join([gene_symbol,\n",
    "                      gene_id,\n",
    "                      gwas_annotation(snp_id),\n",
    "                      risk_allele_annotation(snp_id),\n",
    "                      eqtl_annotation(snp_id, gene_id), \n",
    "                      de_annotation(gene_id)])\n",
    "    \n",
    "    p += ggtitle(title) \n",
    "    return p"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Junctions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    os.makedirs(feature)\n",
    "except OSError as e:\n",
    "    if e.errno != errno.EEXIST:\n",
    "        raise"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Enrichment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Integrate DEG with PGC2+CLOZUK SNPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jbenja13/.local/lib/python3.9/site-packages/numpy/lib/arraysetops.py:583: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(3190613, 57)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft = get_integration_df()\n",
    "dft.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[5196, 74475], [122145, 2988797]]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(1.7071801613298905, 4.8917286718013116e-254)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "agreement = {1: 'Yes', -1: 'No', 0: 0}\n",
    "dft['agree_direction'] = dft.apply(agree_direction, axis=1)\n",
    "dft.agree_direction = [agreement[item] for item in dft['agree_direction']]\n",
    "\n",
    "table =  [[np.sum((dft['P']<5e-8) & ((dft['adj.P.Val']<.05))),\n",
    "           np.sum((dft['P']<5e-8) & ((dft['adj.P.Val']>=.05)))],\n",
    "          [np.sum((dft['P']>=5e-8) & ((dft['adj.P.Val']<.05))),\n",
    "           np.sum((dft['P']>=5e-8) & ((dft['adj.P.Val']>=.05)))]]\n",
    "print(table)\n",
    "fisher_exact(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>No</td>\n",
       "      <td>295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Yes</td>\n",
       "      <td>4901</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  agree_direction     0\n",
       "0              No   295\n",
       "1             Yes  4901"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dft1 = dft[(dft['P']<5e-8) & ((dft['adj.P.Val']<.05))]\n",
    "df = dft1.groupby('agree_direction').size().reset_index()\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "binom_test(df[0].iloc[1], df[0].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "dft2 = dft[(dft['P']<=5e-8) & (dft['adj.P.Val'] < 0.05)].copy()\n",
    "dft2['risk_allele'] = dft2['our_snp_id'].apply(get_risk_allele)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "direction = {-1: 'Down', 1: 'Up'}\n",
    "boolean_conv = {True: 1, False: -1}\n",
    "dft2.pgc2_a1_same_as_our_counted = [boolean_conv[item] for item in dft2['pgc2_a1_same_as_our_counted']]\n",
    "dft2['eqtl_gwas_dir'] = [direction[item] for item in np.sign(dft2['pgc2_a1_same_as_our_counted']) * np.sign(dft2['slope']) * np.sign(dft2['OR'] - 1)]\n",
    "dft2['de_dir'] = [direction[item] for item in np.sign(dft2['t'])]\n",
    "dft2['eqtl_slope'] = np.sign(dft2['pgc2_a1_same_as_our_counted']) * np.sign(dft2['OR'] - 1) * dft2['slope']\n",
    "dft2 = dft2[['gene_id', 'newGeneSymbol', 'variant_id', 'A1', 'A2', 'risk_allele', 'OR', \n",
    "             'P', 'pval_nominal', 'adj.P.Val', 'logFC', 't', 'eqtl_slope', \n",
    "             'de_dir', 'eqtl_gwas_dir', 'agree_direction']].rename(columns={\"newGeneSymbol\": \"Symbol\"})\n",
    "dft2['Symbol'].fillna(dft2['gene_id'], inplace=True)\n",
    "dft2.to_csv('%s/integration_by_symbol.txt' % feature, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "agree_direction\n",
       "No     10\n",
       "Yes    11\n",
       "dtype: int64"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = dft2.groupby(['Symbol']).first().reset_index().sort_values('P')\n",
    "df2.groupby('agree_direction').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrN</th>\n",
       "      <th>our_snp_id</th>\n",
       "      <th>cm</th>\n",
       "      <th>pos</th>\n",
       "      <th>our_counted</th>\n",
       "      <th>our_alt</th>\n",
       "      <th>chrom</th>\n",
       "      <th>SNP</th>\n",
       "      <th>Freq.A1</th>\n",
       "      <th>CHR</th>\n",
       "      <th>...</th>\n",
       "      <th>endExon</th>\n",
       "      <th>newGeneID</th>\n",
       "      <th>isFusion</th>\n",
       "      <th>logFC</th>\n",
       "      <th>AveExpr</th>\n",
       "      <th>de_t</th>\n",
       "      <th>P.Value</th>\n",
       "      <th>de_FDR</th>\n",
       "      <th>B</th>\n",
       "      <th>agree_direction</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>newGeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HLA-C</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:30770669:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>30770669</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs3095336:30738446:G:A</td>\n",
       "      <td>0.9085</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>182798.0</td>\n",
       "      <td>ENSG00000204525.15</td>\n",
       "      <td>False</td>\n",
       "      <td>-1.434498</td>\n",
       "      <td>-1.062368</td>\n",
       "      <td>-3.191459</td>\n",
       "      <td>1.533957e-03</td>\n",
       "      <td>0.037406</td>\n",
       "      <td>-1.448857</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4B</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:31530467:G:GC</td>\n",
       "      <td>0</td>\n",
       "      <td>31530467</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs9281523:31498244:G:GC</td>\n",
       "      <td>0.0825</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>184126.0</td>\n",
       "      <td>ENSG00000224389.8</td>\n",
       "      <td>False</td>\n",
       "      <td>0.273063</td>\n",
       "      <td>1.175654</td>\n",
       "      <td>3.088621</td>\n",
       "      <td>2.159476e-03</td>\n",
       "      <td>0.046496</td>\n",
       "      <td>-1.588252</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZSCAN26</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:27772521:C:A</td>\n",
       "      <td>0</td>\n",
       "      <td>27772521</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs760587:27740300:C:A</td>\n",
       "      <td>0.7803</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSG00000197062.11</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.261770</td>\n",
       "      <td>0.397751</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>5.341323e-04</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.337857</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZFYVE21</th>\n",
       "      <td>14</td>\n",
       "      <td>chr14:103633871:T:A</td>\n",
       "      <td>0</td>\n",
       "      <td>103633871</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>chr14</td>\n",
       "      <td>rs12147645:104100208:T:A</td>\n",
       "      <td>0.7028</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>386973.0</td>\n",
       "      <td>ENSG00000100711.13</td>\n",
       "      <td>False</td>\n",
       "      <td>0.137121</td>\n",
       "      <td>2.583755</td>\n",
       "      <td>4.331524</td>\n",
       "      <td>1.899676e-05</td>\n",
       "      <td>0.002033</td>\n",
       "      <td>2.543318</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CKB</th>\n",
       "      <td>14</td>\n",
       "      <td>chr14:103710761:G:T</td>\n",
       "      <td>0</td>\n",
       "      <td>103710761</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr14</td>\n",
       "      <td>rs4900592:104177098:G:T</td>\n",
       "      <td>0.7048</td>\n",
       "      <td>14</td>\n",
       "      <td>...</td>\n",
       "      <td>386769.0</td>\n",
       "      <td>ENSG00000166165.12</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.094478</td>\n",
       "      <td>6.957920</td>\n",
       "      <td>-4.201159</td>\n",
       "      <td>3.316100e-05</td>\n",
       "      <td>0.002976</td>\n",
       "      <td>1.672783</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BAG6</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:31610995:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>31610995</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs2844477:31578772:T:C</td>\n",
       "      <td>0.5974</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>183144.0</td>\n",
       "      <td>ENSG00000204463.12</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.111484</td>\n",
       "      <td>2.471439</td>\n",
       "      <td>-3.769427</td>\n",
       "      <td>1.897734e-04</td>\n",
       "      <td>0.009575</td>\n",
       "      <td>0.440874</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NGEF</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2:232926898:C:G</td>\n",
       "      <td>0</td>\n",
       "      <td>232926898</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>chr2</td>\n",
       "      <td>rs1878289:233791608:C:G</td>\n",
       "      <td>0.5596</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>91921.0</td>\n",
       "      <td>ENSG00000066248.14</td>\n",
       "      <td>False</td>\n",
       "      <td>0.115109</td>\n",
       "      <td>5.582328</td>\n",
       "      <td>3.772815</td>\n",
       "      <td>1.873065e-04</td>\n",
       "      <td>0.009518</td>\n",
       "      <td>0.159948</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLCH2</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:2440958:A:G</td>\n",
       "      <td>0</td>\n",
       "      <td>2440958</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr1</td>\n",
       "      <td>rs6688934:2372397:A:G</td>\n",
       "      <td>0.5080</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1980.0</td>\n",
       "      <td>ENSG00000149527.17</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.254239</td>\n",
       "      <td>-0.320352</td>\n",
       "      <td>-3.941689</td>\n",
       "      <td>9.639145e-05</td>\n",
       "      <td>0.006178</td>\n",
       "      <td>0.891407</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HLA-DPB1</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:33020111:A:G</td>\n",
       "      <td>0</td>\n",
       "      <td>33020111</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs17214290:32987888:A:G</td>\n",
       "      <td>0.9563</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>184945.0</td>\n",
       "      <td>ENSG00000223865.10</td>\n",
       "      <td>False</td>\n",
       "      <td>0.452891</td>\n",
       "      <td>-1.123835</td>\n",
       "      <td>3.159284</td>\n",
       "      <td>1.708866e-03</td>\n",
       "      <td>0.040099</td>\n",
       "      <td>-1.508125</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DRD2</th>\n",
       "      <td>11</td>\n",
       "      <td>chr11:113447023:G:C</td>\n",
       "      <td>0</td>\n",
       "      <td>113447023</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>chr11</td>\n",
       "      <td>rs17601612:113317745:G:C</td>\n",
       "      <td>0.6272</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>320189.0</td>\n",
       "      <td>ENSG00000149295.13</td>\n",
       "      <td>False</td>\n",
       "      <td>0.302443</td>\n",
       "      <td>-0.618759</td>\n",
       "      <td>4.079084</td>\n",
       "      <td>5.516630e-05</td>\n",
       "      <td>0.004208</td>\n",
       "      <td>1.310019</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ME1</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:83463743:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>83463743</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs2324418:84173462:T:C</td>\n",
       "      <td>0.6720</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>192328.0</td>\n",
       "      <td>ENSG00000065833.8</td>\n",
       "      <td>False</td>\n",
       "      <td>0.114156</td>\n",
       "      <td>2.960380</td>\n",
       "      <td>3.949190</td>\n",
       "      <td>9.353690e-05</td>\n",
       "      <td>0.006069</td>\n",
       "      <td>1.033709</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PACSIN3</th>\n",
       "      <td>11</td>\n",
       "      <td>chr11:46729945:T:TAGG</td>\n",
       "      <td>0</td>\n",
       "      <td>46729945</td>\n",
       "      <td>TAGG</td>\n",
       "      <td>T</td>\n",
       "      <td>chr11</td>\n",
       "      <td>rs3136476:46751495:T:TAGG</td>\n",
       "      <td>0.1581</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>302524.0</td>\n",
       "      <td>ENSG00000165912.15</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.186281</td>\n",
       "      <td>0.339457</td>\n",
       "      <td>-3.419199</td>\n",
       "      <td>6.964737e-04</td>\n",
       "      <td>0.022512</td>\n",
       "      <td>-0.560671</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IP6K3</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:33741539:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>33741539</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs791904:33709316:G:A</td>\n",
       "      <td>0.8062</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>185634.0</td>\n",
       "      <td>ENSG00000161896.11</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.370340</td>\n",
       "      <td>-0.703164</td>\n",
       "      <td>-4.042487</td>\n",
       "      <td>6.410590e-05</td>\n",
       "      <td>0.004709</td>\n",
       "      <td>1.082038</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCCB</th>\n",
       "      <td>3</td>\n",
       "      <td>chr3:136088767:G:T</td>\n",
       "      <td>0</td>\n",
       "      <td>136088767</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>chr3</td>\n",
       "      <td>rs9810089:135807609:G:T</td>\n",
       "      <td>0.4692</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>118526.0</td>\n",
       "      <td>ENSG00000114054.13</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.105287</td>\n",
       "      <td>2.929883</td>\n",
       "      <td>-3.448890</td>\n",
       "      <td>6.263035e-04</td>\n",
       "      <td>0.020973</td>\n",
       "      <td>-0.700648</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDFIP2</th>\n",
       "      <td>13</td>\n",
       "      <td>chr13:79285321:A:C</td>\n",
       "      <td>0</td>\n",
       "      <td>79285321</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr13</td>\n",
       "      <td>rs9545047:79859456:A:C</td>\n",
       "      <td>0.6074</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>365438.0</td>\n",
       "      <td>ENSG00000102471.13</td>\n",
       "      <td>False</td>\n",
       "      <td>0.081578</td>\n",
       "      <td>3.780114</td>\n",
       "      <td>3.311700</td>\n",
       "      <td>1.016623e-03</td>\n",
       "      <td>0.028837</td>\n",
       "      <td>-1.232955</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NDRG4</th>\n",
       "      <td>16</td>\n",
       "      <td>chr16:58505281:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>58505281</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>chr16</td>\n",
       "      <td>rs8056039:58539185:G:A</td>\n",
       "      <td>0.4761</td>\n",
       "      <td>16</td>\n",
       "      <td>...</td>\n",
       "      <td>429234.0</td>\n",
       "      <td>ENSG00000103034.14</td>\n",
       "      <td>False</td>\n",
       "      <td>0.171047</td>\n",
       "      <td>2.478321</td>\n",
       "      <td>3.324829</td>\n",
       "      <td>9.712452e-04</td>\n",
       "      <td>0.027942</td>\n",
       "      <td>-1.056353</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RERE</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:8361143:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>8361143</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr1</td>\n",
       "      <td>rs13596:8421203:T:C</td>\n",
       "      <td>0.3181</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>3461.0</td>\n",
       "      <td>ENSG00000142599.17</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.207443</td>\n",
       "      <td>0.327161</td>\n",
       "      <td>-3.460156</td>\n",
       "      <td>6.014509e-04</td>\n",
       "      <td>0.020434</td>\n",
       "      <td>-0.439841</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MPP6</th>\n",
       "      <td>7</td>\n",
       "      <td>chr7:24737470:CA:C</td>\n",
       "      <td>0</td>\n",
       "      <td>24737470</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>chr7</td>\n",
       "      <td>rs146678232:24777089:CA:C</td>\n",
       "      <td>0.1720</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>206629.0</td>\n",
       "      <td>ENSG00000105926.15</td>\n",
       "      <td>False</td>\n",
       "      <td>0.079785</td>\n",
       "      <td>4.509116</td>\n",
       "      <td>3.603734</td>\n",
       "      <td>3.556283e-04</td>\n",
       "      <td>0.014432</td>\n",
       "      <td>-0.338258</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REEP2</th>\n",
       "      <td>5</td>\n",
       "      <td>chr5:138362610:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>138362610</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>chr5</td>\n",
       "      <td>rs10055995:137698299:C:T</td>\n",
       "      <td>0.4066</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>168231.0</td>\n",
       "      <td>ENSG00000132563.15</td>\n",
       "      <td>False</td>\n",
       "      <td>0.141226</td>\n",
       "      <td>3.947112</td>\n",
       "      <td>5.772797</td>\n",
       "      <td>1.626890e-08</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>9.119410</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C4A</th>\n",
       "      <td>6</td>\n",
       "      <td>chr6:31495800:T:C</td>\n",
       "      <td>0</td>\n",
       "      <td>31495800</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>chr6</td>\n",
       "      <td>rs2534680:31463577:T:C</td>\n",
       "      <td>0.7376</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>184035.0</td>\n",
       "      <td>ENSG00000244731.7</td>\n",
       "      <td>False</td>\n",
       "      <td>0.349910</td>\n",
       "      <td>1.399810</td>\n",
       "      <td>3.265208</td>\n",
       "      <td>1.193640e-03</td>\n",
       "      <td>0.031770</td>\n",
       "      <td>-1.095988</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VRK2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr2:57867738:A:C</td>\n",
       "      <td>0</td>\n",
       "      <td>57867738</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>chr2</td>\n",
       "      <td>rs2717068:58094873:A:C</td>\n",
       "      <td>0.3966</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>61902.0</td>\n",
       "      <td>ENSG00000028116.16</td>\n",
       "      <td>False</td>\n",
       "      <td>-0.148130</td>\n",
       "      <td>2.995787</td>\n",
       "      <td>-3.135273</td>\n",
       "      <td>1.851197e-03</td>\n",
       "      <td>0.042204</td>\n",
       "      <td>-1.684538</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21 rows Ã— 57 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              chrN             our_snp_id  cm        pos our_counted our_alt  \\\n",
       "newGeneSymbol                                                                  \n",
       "HLA-C            6      chr6:30770669:G:A   0   30770669           A       G   \n",
       "C4B              6     chr6:31530467:G:GC   0   31530467          GC       G   \n",
       "ZSCAN26          6      chr6:27772521:C:A   0   27772521           A       C   \n",
       "ZFYVE21         14    chr14:103633871:T:A   0  103633871           A       T   \n",
       "CKB             14    chr14:103710761:G:T   0  103710761           T       G   \n",
       "BAG6             6      chr6:31610995:T:C   0   31610995           C       T   \n",
       "NGEF             2     chr2:232926898:C:G   0  232926898           G       C   \n",
       "PLCH2            1       chr1:2440958:A:G   0    2440958           G       A   \n",
       "HLA-DPB1         6      chr6:33020111:A:G   0   33020111           G       A   \n",
       "DRD2            11    chr11:113447023:G:C   0  113447023           C       G   \n",
       "ME1              6      chr6:83463743:T:C   0   83463743           C       T   \n",
       "PACSIN3         11  chr11:46729945:T:TAGG   0   46729945        TAGG       T   \n",
       "IP6K3            6      chr6:33741539:G:A   0   33741539           A       G   \n",
       "PCCB             3     chr3:136088767:G:T   0  136088767           T       G   \n",
       "NDFIP2          13     chr13:79285321:A:C   0   79285321           C       A   \n",
       "NDRG4           16     chr16:58505281:G:A   0   58505281           A       G   \n",
       "RERE             1       chr1:8361143:T:C   0    8361143           C       T   \n",
       "MPP6             7     chr7:24737470:CA:C   0   24737470           C      CA   \n",
       "REEP2            5     chr5:138362610:C:T   0  138362610           T       C   \n",
       "C4A              6      chr6:31495800:T:C   0   31495800           C       T   \n",
       "VRK2             2      chr2:57867738:A:C   0   57867738           C       A   \n",
       "\n",
       "               chrom                        SNP  Freq.A1  CHR  ...   endExon  \\\n",
       "newGeneSymbol                                                  ...             \n",
       "HLA-C           chr6     rs3095336:30738446:G:A   0.9085    6  ...  182798.0   \n",
       "C4B             chr6    rs9281523:31498244:G:GC   0.0825    6  ...  184126.0   \n",
       "ZSCAN26         chr6      rs760587:27740300:C:A   0.7803    6  ...       NaN   \n",
       "ZFYVE21        chr14   rs12147645:104100208:T:A   0.7028   14  ...  386973.0   \n",
       "CKB            chr14    rs4900592:104177098:G:T   0.7048   14  ...  386769.0   \n",
       "BAG6            chr6     rs2844477:31578772:T:C   0.5974    6  ...  183144.0   \n",
       "NGEF            chr2    rs1878289:233791608:C:G   0.5596    2  ...   91921.0   \n",
       "PLCH2           chr1      rs6688934:2372397:A:G   0.5080    1  ...    1980.0   \n",
       "HLA-DPB1        chr6    rs17214290:32987888:A:G   0.9563    6  ...  184945.0   \n",
       "DRD2           chr11   rs17601612:113317745:G:C   0.6272   11  ...  320189.0   \n",
       "ME1             chr6     rs2324418:84173462:T:C   0.6720    6  ...  192328.0   \n",
       "PACSIN3        chr11  rs3136476:46751495:T:TAGG   0.1581   11  ...  302524.0   \n",
       "IP6K3           chr6      rs791904:33709316:G:A   0.8062    6  ...  185634.0   \n",
       "PCCB            chr3    rs9810089:135807609:G:T   0.4692    3  ...  118526.0   \n",
       "NDFIP2         chr13     rs9545047:79859456:A:C   0.6074   13  ...  365438.0   \n",
       "NDRG4          chr16     rs8056039:58539185:G:A   0.4761   16  ...  429234.0   \n",
       "RERE            chr1        rs13596:8421203:T:C   0.3181    1  ...    3461.0   \n",
       "MPP6            chr7  rs146678232:24777089:CA:C   0.1720    7  ...  206629.0   \n",
       "REEP2           chr5   rs10055995:137698299:C:T   0.4066    5  ...  168231.0   \n",
       "C4A             chr6     rs2534680:31463577:T:C   0.7376    6  ...  184035.0   \n",
       "VRK2            chr2     rs2717068:58094873:A:C   0.3966    2  ...   61902.0   \n",
       "\n",
       "                        newGeneID isFusion     logFC   AveExpr      de_t  \\\n",
       "newGeneSymbol                                                              \n",
       "HLA-C          ENSG00000204525.15    False -1.434498 -1.062368 -3.191459   \n",
       "C4B             ENSG00000224389.8    False  0.273063  1.175654  3.088621   \n",
       "ZSCAN26        ENSG00000197062.11    False -0.261770  0.397751 -3.493012   \n",
       "ZFYVE21        ENSG00000100711.13    False  0.137121  2.583755  4.331524   \n",
       "CKB            ENSG00000166165.12    False -0.094478  6.957920 -4.201159   \n",
       "BAG6           ENSG00000204463.12    False -0.111484  2.471439 -3.769427   \n",
       "NGEF           ENSG00000066248.14    False  0.115109  5.582328  3.772815   \n",
       "PLCH2          ENSG00000149527.17    False -0.254239 -0.320352 -3.941689   \n",
       "HLA-DPB1       ENSG00000223865.10    False  0.452891 -1.123835  3.159284   \n",
       "DRD2           ENSG00000149295.13    False  0.302443 -0.618759  4.079084   \n",
       "ME1             ENSG00000065833.8    False  0.114156  2.960380  3.949190   \n",
       "PACSIN3        ENSG00000165912.15    False -0.186281  0.339457 -3.419199   \n",
       "IP6K3          ENSG00000161896.11    False -0.370340 -0.703164 -4.042487   \n",
       "PCCB           ENSG00000114054.13    False -0.105287  2.929883 -3.448890   \n",
       "NDFIP2         ENSG00000102471.13    False  0.081578  3.780114  3.311700   \n",
       "NDRG4          ENSG00000103034.14    False  0.171047  2.478321  3.324829   \n",
       "RERE           ENSG00000142599.17    False -0.207443  0.327161 -3.460156   \n",
       "MPP6           ENSG00000105926.15    False  0.079785  4.509116  3.603734   \n",
       "REEP2          ENSG00000132563.15    False  0.141226  3.947112  5.772797   \n",
       "C4A             ENSG00000244731.7    False  0.349910  1.399810  3.265208   \n",
       "VRK2           ENSG00000028116.16    False -0.148130  2.995787 -3.135273   \n",
       "\n",
       "                    P.Value    de_FDR         B  agree_direction  \n",
       "newGeneSymbol                                                     \n",
       "HLA-C          1.533957e-03  0.037406 -1.448857              Yes  \n",
       "C4B            2.159476e-03  0.046496 -1.588252              Yes  \n",
       "ZSCAN26        5.341323e-04  0.018932 -0.337857              Yes  \n",
       "ZFYVE21        1.899676e-05  0.002033  2.543318              Yes  \n",
       "CKB            3.316100e-05  0.002976  1.672783              Yes  \n",
       "BAG6           1.897734e-04  0.009575  0.440874               No  \n",
       "NGEF           1.873065e-04  0.009518  0.159948               No  \n",
       "PLCH2          9.639145e-05  0.006178  0.891407              Yes  \n",
       "HLA-DPB1       1.708866e-03  0.040099 -1.508125              Yes  \n",
       "DRD2           5.516630e-05  0.004208  1.310019               No  \n",
       "ME1            9.353690e-05  0.006069  1.033709               No  \n",
       "PACSIN3        6.964737e-04  0.022512 -0.560671               No  \n",
       "IP6K3          6.410590e-05  0.004709  1.082038               No  \n",
       "PCCB           6.263035e-04  0.020973 -0.700648              Yes  \n",
       "NDFIP2         1.016623e-03  0.028837 -1.232955               No  \n",
       "NDRG4          9.712452e-04  0.027942 -1.056353               No  \n",
       "RERE           6.014509e-04  0.020434 -0.439841               No  \n",
       "MPP6           3.556283e-04  0.014432 -0.338258              Yes  \n",
       "REEP2          1.626890e-08  0.000011  9.119410              Yes  \n",
       "C4A            1.193640e-03  0.031770 -1.095988              Yes  \n",
       "VRK2           1.851197e-03  0.042204 -1.684538               No  \n",
       "\n",
       "[21 rows x 57 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.set_index('Symbol').rename(columns={'t': 'de_t', 'P': 'GWAS_P', 'pval_nominal': 'eQTL_pvalue', \n",
    "                                        'adj.P.Val': 'de_FDR'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plot with PGC2 risk allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for xx in range(df2.shape[0]):\n",
    "    gg = gwas_annotated_eqtl_pheno_plot(df2.iloc[xx, :].variant_id, \n",
    "                                        df2.iloc[xx, :].gene_id, 'Dx')\n",
    "    print(gg)\n",
    "    label = '%s/eqtl_gwas_%s' % (feature, df2.iloc[xx, :].Symbol)\n",
    "    save_plot(gg, label)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
