{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate supplemental data for TWAS, caudate, across all features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:53.012316Z",
     "iopub.status.busy": "2022-03-16T14:39:53.009398Z",
     "iopub.status.idle": "2022-03-16T14:39:53.541520Z",
     "shell.execute_reply": "2022-03-16T14:39:53.540191Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## With MHC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:53.552046Z",
     "iopub.status.busy": "2022-03-16T14:39:53.551227Z",
     "iopub.status.idle": "2022-03-16T14:39:53.869437Z",
     "shell.execute_reply": "2022-03-16T14:39:53.870160Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>254</th>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.330293</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>10.971295</td>\n",
       "      <td>5.251485e-28</td>\n",
       "      <td>2.668805e-24</td>\n",
       "      <td>2.668805e-24</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4439</th>\n",
       "      <td>ENSG00000219891</td>\n",
       "      <td>ENSG00000219891</td>\n",
       "      <td>ZSCAN12P1</td>\n",
       "      <td>0.219714</td>\n",
       "      <td>chr6:27837477:A:C</td>\n",
       "      <td>chr6:27883095:G:A</td>\n",
       "      <td>10.689750</td>\n",
       "      <td>1.136772e-26</td>\n",
       "      <td>2.888538e-23</td>\n",
       "      <td>5.777075e-23</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature        ensemblID         ID       HSQ  \\\n",
       "254   ENSG00000244731  ENSG00000244731        C4A  0.330293   \n",
       "4439  ENSG00000219891  ENSG00000219891  ZSCAN12P1  0.219714   \n",
       "\n",
       "           BEST.GWAS.ID            EQTL.ID     TWAS.Z        TWAS.P  \\\n",
       "254   chr6:32219860:G:T  chr6:31902549:G:A  10.971295  5.251485e-28   \n",
       "4439  chr6:27837477:A:C  chr6:27883095:G:A  10.689750  1.136772e-26   \n",
       "\n",
       "               FDR    Bonferroni  Type  \n",
       "254   2.668805e-24  2.668805e-24  Gene  \n",
       "4439  2.888538e-23  5.777075e-23  Gene  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'gene_weights/fusion/summary_stats/_m/fusion_associations.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/genes/diffExpr_szVctl_full.txt', sep='\\t')\n",
    "genes = annot[['ensemblID']].merge(genes, left_on='ensemblID', right_on='FILE')\n",
    "genes = genes[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "genes['Type'] = 'Gene'\n",
    "genes.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "genes.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:53.881623Z",
     "iopub.status.busy": "2022-03-16T14:39:53.880683Z",
     "iopub.status.idle": "2022-03-16T14:39:55.441743Z",
     "shell.execute_reply": "2022-03-16T14:39:55.439791Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3106</th>\n",
       "      <td>ENST00000496659</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.208279</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31942164:A:G</td>\n",
       "      <td>11.008695</td>\n",
       "      <td>3.469941e-28</td>\n",
       "      <td>3.078185e-24</td>\n",
       "      <td>3.078185e-24</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>ENST00000428956</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.313207</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>10.781964</td>\n",
       "      <td>4.188461e-27</td>\n",
       "      <td>1.857792e-23</td>\n",
       "      <td>3.715584e-23</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature        ensemblID   ID       HSQ       BEST.GWAS.ID  \\\n",
       "3106  ENST00000496659  ENSG00000244731  C4A  0.208279  chr6:32219860:G:T   \n",
       "291   ENST00000428956  ENSG00000244731  C4A  0.313207  chr6:32219860:G:T   \n",
       "\n",
       "                EQTL.ID     TWAS.Z        TWAS.P           FDR    Bonferroni  \\\n",
       "3106  chr6:31942164:A:G  11.008695  3.469941e-28  3.078185e-24  3.078185e-24   \n",
       "291   chr6:31902549:G:A  10.781964  4.188461e-27  1.857792e-23  3.715584e-23   \n",
       "\n",
       "            Type  \n",
       "3106  Transcript  \n",
       "291   Transcript  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'transcript_weights/fusion/summary_stats/_m/fusion_associations.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/transcripts/diffExpr_szVctl_full.txt', sep='\\t')\n",
    "annot['ensemblID'] = annot.gene_id.str.replace('\\\\..*', '', regex=True)\n",
    "annot['FILE'] = annot.transcript_id.str.replace('\\\\..*', '', regex=True)\n",
    "trans = annot[['ensemblID', 'FILE']].merge(trans, on='FILE')\n",
    "trans = trans[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "trans['Type'] = 'Transcript'\n",
    "trans.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "trans.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:55.453076Z",
     "iopub.status.busy": "2022-03-16T14:39:55.451684Z",
     "iopub.status.idle": "2022-03-16T14:39:58.321870Z",
     "shell.execute_reply": "2022-03-16T14:39:58.320619Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34301</th>\n",
       "      <td>e377846</td>\n",
       "      <td>ENSG00000186470</td>\n",
       "      <td>BTN3A2</td>\n",
       "      <td>0.136590</td>\n",
       "      <td>chr6:26463347:G:T</td>\n",
       "      <td>chr6:26336344:G:A</td>\n",
       "      <td>11.162478</td>\n",
       "      <td>6.223219e-29</td>\n",
       "      <td>2.397931e-24</td>\n",
       "      <td>2.397931e-24</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34737</th>\n",
       "      <td>e384610</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.387274</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>11.064041</td>\n",
       "      <td>1.874569e-28</td>\n",
       "      <td>3.611545e-24</td>\n",
       "      <td>7.223090e-24</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature        ensemblID      ID       HSQ       BEST.GWAS.ID  \\\n",
       "34301  e377846  ENSG00000186470  BTN3A2  0.136590  chr6:26463347:G:T   \n",
       "34737  e384610  ENSG00000244731     C4A  0.387274  chr6:32219860:G:T   \n",
       "\n",
       "                 EQTL.ID     TWAS.Z        TWAS.P           FDR    Bonferroni  \\\n",
       "34301  chr6:26336344:G:A  11.162478  6.223219e-29  2.397931e-24  2.397931e-24   \n",
       "34737  chr6:31902549:G:A  11.064041  1.874569e-28  3.611545e-24  7.223090e-24   \n",
       "\n",
       "       Type  \n",
       "34301  Exon  \n",
       "34737  Exon  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'exon_weights/fusion/summary_stats/_m/fusion_associations.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/exons/diffExpr_szVctl_full.txt', \n",
    "                    sep='\\t', index_col=0)\n",
    "exons = annot[['ensemblID']].merge(exons, left_index=True, right_on='FILE')\n",
    "exons = exons[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "exons['Type'] = 'Exon'\n",
    "exons.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "exons.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:58.333811Z",
     "iopub.status.busy": "2022-03-16T14:39:58.332908Z",
     "iopub.status.idle": "2022-03-16T14:39:59.080363Z",
     "shell.execute_reply": "2022-03-16T14:39:59.079499Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10530</th>\n",
       "      <td>chr6:31996601-31996828(+)</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.375134</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>10.811784</td>\n",
       "      <td>3.027249e-27</td>\n",
       "      <td>2.030200e-23</td>\n",
       "      <td>3.972356e-23</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10528</th>\n",
       "      <td>chr6:31996112-31996206(+)</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.410304</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>10.809773</td>\n",
       "      <td>3.094346e-27</td>\n",
       "      <td>2.030200e-23</td>\n",
       "      <td>4.060401e-23</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Feature        ensemblID   ID       HSQ  \\\n",
       "10530  chr6:31996601-31996828(+)  ENSG00000244731  C4A  0.375134   \n",
       "10528  chr6:31996112-31996206(+)  ENSG00000244731  C4A  0.410304   \n",
       "\n",
       "            BEST.GWAS.ID            EQTL.ID     TWAS.Z        TWAS.P  \\\n",
       "10530  chr6:32219860:G:T  chr6:31902549:G:A  10.811784  3.027249e-27   \n",
       "10528  chr6:32219860:G:T  chr6:31902549:G:A  10.809773  3.094346e-27   \n",
       "\n",
       "                FDR    Bonferroni      Type  \n",
       "10530  2.030200e-23  3.972356e-23  Junction  \n",
       "10528  2.030200e-23  4.060401e-23  Junction  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = pd.read_csv('../../_m/jxn_annotation.tsv', sep='\\t', index_col=1)\n",
    "annot[\"gene_id\"] = annot.index\n",
    "juncs = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'junction_weights/fusion/summary_stats/_m/fusion_associations.txt', sep='\\t')\n",
    "juncs = pd.merge(annot, juncs, left_on='JxnID', right_on='FILE')\n",
    "juncs = juncs[['gene_id', 'ensemblID', 'Symbol', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "juncs['Type'] = 'Junction'\n",
    "juncs.rename(columns={'Symbol': 'ID', 'gene_id': 'Feature'}, inplace=True)\n",
    "juncs.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:59.092591Z",
     "iopub.status.busy": "2022-03-16T14:39:59.087734Z",
     "iopub.status.idle": "2022-03-16T14:39:59.153830Z",
     "shell.execute_reply": "2022-03-16T14:39:59.153006Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65607, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000248587</td>\n",
       "      <td>ENSG00000248587</td>\n",
       "      <td>GDNF-AS1</td>\n",
       "      <td>0.182975</td>\n",
       "      <td>chr5:37821119:C:A</td>\n",
       "      <td>chr5:37981540:A:C</td>\n",
       "      <td>-0.345122</td>\n",
       "      <td>0.730003</td>\n",
       "      <td>0.889658</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000185052</td>\n",
       "      <td>ENSG00000185052</td>\n",
       "      <td>SLC24A3</td>\n",
       "      <td>0.255074</td>\n",
       "      <td>chr20:19648189:C:T</td>\n",
       "      <td>chr20:19138350:A:G</td>\n",
       "      <td>0.400819</td>\n",
       "      <td>0.688553</td>\n",
       "      <td>0.871279</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature        ensemblID        ID       HSQ        BEST.GWAS.ID  \\\n",
       "0  ENSG00000248587  ENSG00000248587  GDNF-AS1  0.182975   chr5:37821119:C:A   \n",
       "1  ENSG00000185052  ENSG00000185052   SLC24A3  0.255074  chr20:19648189:C:T   \n",
       "\n",
       "              EQTL.ID    TWAS.Z    TWAS.P       FDR  Bonferroni  Type  \n",
       "0   chr5:37981540:A:C -0.345122  0.730003  0.889658         1.0  Gene  \n",
       "1  chr20:19138350:A:G  0.400819  0.688553  0.871279         1.0  Gene  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([genes, trans, exons, juncs], axis=0)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:39:59.164633Z",
     "iopub.status.busy": "2022-03-16T14:39:59.163556Z",
     "iopub.status.idle": "2022-03-16T14:40:01.580275Z",
     "shell.execute_reply": "2022-03-16T14:40:01.579494Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv('BrainSeq_Phase3_Caudate_TWAS_associations_allFeatures.txt.gz', index=False, header=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Without MHC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:01.592211Z",
     "iopub.status.busy": "2022-03-16T14:40:01.591149Z",
     "iopub.status.idle": "2022-03-16T14:40:01.924780Z",
     "shell.execute_reply": "2022-03-16T14:40:01.923810Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4665</th>\n",
       "      <td>ENSG00000163938</td>\n",
       "      <td>ENSG00000163938</td>\n",
       "      <td>GNL3</td>\n",
       "      <td>0.410009</td>\n",
       "      <td>chr3:52781889:T:C</td>\n",
       "      <td>chr3:52588070:G:A</td>\n",
       "      <td>9.415273</td>\n",
       "      <td>4.718536e-21</td>\n",
       "      <td>2.348887e-17</td>\n",
       "      <td>2.348887e-17</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3609</th>\n",
       "      <td>ENSG00000166159</td>\n",
       "      <td>ENSG00000166159</td>\n",
       "      <td>LRTM2</td>\n",
       "      <td>0.239365</td>\n",
       "      <td>chr12:2221292:C:T</td>\n",
       "      <td>chr12:2224318:C:T</td>\n",
       "      <td>-9.064394</td>\n",
       "      <td>1.252984e-19</td>\n",
       "      <td>3.118677e-16</td>\n",
       "      <td>6.237353e-16</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature        ensemblID     ID       HSQ       BEST.GWAS.ID  \\\n",
       "4665  ENSG00000163938  ENSG00000163938   GNL3  0.410009  chr3:52781889:T:C   \n",
       "3609  ENSG00000166159  ENSG00000166159  LRTM2  0.239365  chr12:2221292:C:T   \n",
       "\n",
       "                EQTL.ID    TWAS.Z        TWAS.P           FDR    Bonferroni  \\\n",
       "4665  chr3:52588070:G:A  9.415273  4.718536e-21  2.348887e-17  2.348887e-17   \n",
       "3609  chr12:2224318:C:T -9.064394  1.252984e-19  3.118677e-16  6.237353e-16   \n",
       "\n",
       "      Type  \n",
       "4665  Gene  \n",
       "3609  Gene  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'gene_weights/fusion/summary_stats/_m/fusion_associations_noMHC.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/genes/diffExpr_szVctl_full.txt', sep='\\t')\n",
    "genes = annot[['ensemblID']].merge(genes, left_on='ensemblID', right_on='FILE')\n",
    "genes = genes[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "genes['Type'] = 'Gene'\n",
    "genes.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "genes.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:01.935289Z",
     "iopub.status.busy": "2022-03-16T14:40:01.934489Z",
     "iopub.status.idle": "2022-03-16T14:40:03.537027Z",
     "shell.execute_reply": "2022-03-16T14:40:03.538031Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4596</th>\n",
       "      <td>ENST00000394799</td>\n",
       "      <td>ENSG00000163938</td>\n",
       "      <td>GNL3</td>\n",
       "      <td>0.122586</td>\n",
       "      <td>chr3:52781889:T:C</td>\n",
       "      <td>chr3:52799789:C:A</td>\n",
       "      <td>8.983223</td>\n",
       "      <td>2.629480e-19</td>\n",
       "      <td>2.283966e-15</td>\n",
       "      <td>2.283966e-15</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1396</th>\n",
       "      <td>ENST00000315580</td>\n",
       "      <td>ENSG00000182196</td>\n",
       "      <td>ARL6IP4</td>\n",
       "      <td>0.482633</td>\n",
       "      <td>chr12:123148383:G:A</td>\n",
       "      <td>chr12:122973072:C:T</td>\n",
       "      <td>-8.699604</td>\n",
       "      <td>3.330436e-18</td>\n",
       "      <td>1.446408e-14</td>\n",
       "      <td>2.892817e-14</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature        ensemblID       ID       HSQ  \\\n",
       "4596  ENST00000394799  ENSG00000163938     GNL3  0.122586   \n",
       "1396  ENST00000315580  ENSG00000182196  ARL6IP4  0.482633   \n",
       "\n",
       "             BEST.GWAS.ID              EQTL.ID    TWAS.Z        TWAS.P  \\\n",
       "4596    chr3:52781889:T:C    chr3:52799789:C:A  8.983223  2.629480e-19   \n",
       "1396  chr12:123148383:G:A  chr12:122973072:C:T -8.699604  3.330436e-18   \n",
       "\n",
       "               FDR    Bonferroni        Type  \n",
       "4596  2.283966e-15  2.283966e-15  Transcript  \n",
       "1396  1.446408e-14  2.892817e-14  Transcript  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'transcript_weights/fusion/summary_stats/_m/fusion_associations_noMHC.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/transcripts/diffExpr_szVctl_full.txt', sep='\\t')\n",
    "annot['ensemblID'] = annot.gene_id.str.replace('\\\\..*', '', regex=True)\n",
    "annot['FILE'] = annot.transcript_id.str.replace('\\\\..*', '', regex=True)\n",
    "trans = annot[['ensemblID', 'FILE']].merge(trans, on='FILE')\n",
    "trans = trans[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "trans['Type'] = 'Transcript'\n",
    "trans.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "trans.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:03.549904Z",
     "iopub.status.busy": "2022-03-16T14:40:03.548716Z",
     "iopub.status.idle": "2022-03-16T14:40:06.828205Z",
     "shell.execute_reply": "2022-03-16T14:40:06.827109Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34301</th>\n",
       "      <td>e377846</td>\n",
       "      <td>ENSG00000186470</td>\n",
       "      <td>BTN3A2</td>\n",
       "      <td>0.136590</td>\n",
       "      <td>chr6:26463347:G:T</td>\n",
       "      <td>chr6:26336344:G:A</td>\n",
       "      <td>11.162478</td>\n",
       "      <td>6.223219e-29</td>\n",
       "      <td>2.397931e-24</td>\n",
       "      <td>2.397931e-24</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34737</th>\n",
       "      <td>e384610</td>\n",
       "      <td>ENSG00000244731</td>\n",
       "      <td>C4A</td>\n",
       "      <td>0.387274</td>\n",
       "      <td>chr6:32219860:G:T</td>\n",
       "      <td>chr6:31902549:G:A</td>\n",
       "      <td>11.064041</td>\n",
       "      <td>1.874569e-28</td>\n",
       "      <td>3.611545e-24</td>\n",
       "      <td>7.223090e-24</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       Feature        ensemblID      ID       HSQ       BEST.GWAS.ID  \\\n",
       "34301  e377846  ENSG00000186470  BTN3A2  0.136590  chr6:26463347:G:T   \n",
       "34737  e384610  ENSG00000244731     C4A  0.387274  chr6:32219860:G:T   \n",
       "\n",
       "                 EQTL.ID     TWAS.Z        TWAS.P           FDR    Bonferroni  \\\n",
       "34301  chr6:26336344:G:A  11.162478  6.223219e-29  2.397931e-24  2.397931e-24   \n",
       "34737  chr6:31902549:G:A  11.064041  1.874569e-28  3.611545e-24  7.223090e-24   \n",
       "\n",
       "       Type  \n",
       "34301  Exon  \n",
       "34737  Exon  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'exon_weights/fusion/summary_stats/_m/fusion_associations.txt', sep='\\t')\n",
    "annot = pd.read_csv('../../../../differential_expression/_m/exons/diffExpr_szVctl_full.txt', \n",
    "                    sep='\\t', index_col=0)\n",
    "exons = annot[['ensemblID']].merge(exons, left_index=True, right_on='FILE')\n",
    "exons = exons[['FILE', 'ensemblID', 'ID', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "exons['Type'] = 'Exon'\n",
    "exons.rename(columns={'FILE': 'Feature'}, inplace=True)\n",
    "exons.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:06.839914Z",
     "iopub.status.busy": "2022-03-16T14:40:06.838553Z",
     "iopub.status.idle": "2022-03-16T14:40:07.790696Z",
     "shell.execute_reply": "2022-03-16T14:40:07.791510Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8293</th>\n",
       "      <td>chr3:52690705-52690944(+)</td>\n",
       "      <td>ENSG00000163938</td>\n",
       "      <td>GNL3</td>\n",
       "      <td>0.062694</td>\n",
       "      <td>chr3:52781889:T:C</td>\n",
       "      <td>chr3:52507237:A:G</td>\n",
       "      <td>9.632857</td>\n",
       "      <td>5.809151e-22</td>\n",
       "      <td>7.466501e-18</td>\n",
       "      <td>7.466501e-18</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8295</th>\n",
       "      <td>chr3:52693808-52694036(+)</td>\n",
       "      <td>ENSG00000163938</td>\n",
       "      <td>GNL3</td>\n",
       "      <td>0.114983</td>\n",
       "      <td>chr3:52781889:T:C</td>\n",
       "      <td>chr3:52594040:C:A</td>\n",
       "      <td>9.401695</td>\n",
       "      <td>5.369125e-21</td>\n",
       "      <td>3.450468e-17</td>\n",
       "      <td>6.900936e-17</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        Feature        ensemblID    ID       HSQ  \\\n",
       "8293  chr3:52690705-52690944(+)  ENSG00000163938  GNL3  0.062694   \n",
       "8295  chr3:52693808-52694036(+)  ENSG00000163938  GNL3  0.114983   \n",
       "\n",
       "           BEST.GWAS.ID            EQTL.ID    TWAS.Z        TWAS.P  \\\n",
       "8293  chr3:52781889:T:C  chr3:52507237:A:G  9.632857  5.809151e-22   \n",
       "8295  chr3:52781889:T:C  chr3:52594040:C:A  9.401695  5.369125e-21   \n",
       "\n",
       "               FDR    Bonferroni      Type  \n",
       "8293  7.466501e-18  7.466501e-18  Junction  \n",
       "8295  3.450468e-17  6.900936e-17  Junction  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot = pd.read_csv('../../_m/jxn_annotation.tsv', sep='\\t', index_col=1)\n",
    "annot[\"gene_id\"] = annot.index\n",
    "juncs = pd.read_csv('/ceph/projects/v4_phase3_paper/analysis/twas_ea/'+\\\n",
    "                    'junction_weights/fusion/summary_stats/_m/fusion_associations_noMHC.txt', sep='\\t')\n",
    "juncs = pd.merge(annot, juncs, left_on='JxnID', right_on='FILE')\n",
    "juncs = juncs[['gene_id', 'ensemblID', 'Symbol', 'HSQ', 'BEST.GWAS.ID', 'EQTL.ID', \n",
    "               'TWAS.Z', 'TWAS.P', 'FDR', 'Bonferroni']]\n",
    "juncs['Type'] = 'Junction'\n",
    "juncs.rename(columns={'Symbol': 'ID', 'gene_id': 'Feature'}, inplace=True)\n",
    "juncs.sort_values('TWAS.P').head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:07.800017Z",
     "iopub.status.busy": "2022-03-16T14:40:07.799212Z",
     "iopub.status.idle": "2022-03-16T14:40:07.861404Z",
     "shell.execute_reply": "2022-03-16T14:40:07.860614Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(65049, 11)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>ID</th>\n",
       "      <th>HSQ</th>\n",
       "      <th>BEST.GWAS.ID</th>\n",
       "      <th>EQTL.ID</th>\n",
       "      <th>TWAS.Z</th>\n",
       "      <th>TWAS.P</th>\n",
       "      <th>FDR</th>\n",
       "      <th>Bonferroni</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000248587</td>\n",
       "      <td>ENSG00000248587</td>\n",
       "      <td>GDNF-AS1</td>\n",
       "      <td>0.182975</td>\n",
       "      <td>chr5:37821119:C:A</td>\n",
       "      <td>chr5:37981540:A:C</td>\n",
       "      <td>-0.345122</td>\n",
       "      <td>0.730003</td>\n",
       "      <td>0.892644</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000185052</td>\n",
       "      <td>ENSG00000185052</td>\n",
       "      <td>SLC24A3</td>\n",
       "      <td>0.255074</td>\n",
       "      <td>chr20:19648189:C:T</td>\n",
       "      <td>chr20:19138350:A:G</td>\n",
       "      <td>0.400819</td>\n",
       "      <td>0.688553</td>\n",
       "      <td>0.875014</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Feature        ensemblID        ID       HSQ        BEST.GWAS.ID  \\\n",
       "0  ENSG00000248587  ENSG00000248587  GDNF-AS1  0.182975   chr5:37821119:C:A   \n",
       "1  ENSG00000185052  ENSG00000185052   SLC24A3  0.255074  chr20:19648189:C:T   \n",
       "\n",
       "              EQTL.ID    TWAS.Z    TWAS.P       FDR  Bonferroni  Type  \n",
       "0   chr5:37981540:A:C -0.345122  0.730003  0.892644         1.0  Gene  \n",
       "1  chr20:19138350:A:G  0.400819  0.688553  0.875014         1.0  Gene  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([genes, trans, exons, juncs], axis=0)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-03-16T14:40:07.869993Z",
     "iopub.status.busy": "2022-03-16T14:40:07.869153Z",
     "iopub.status.idle": "2022-03-16T14:40:10.170043Z",
     "shell.execute_reply": "2022-03-16T14:40:10.170732Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv('BrainSeq_Phase3_Caudate_TWAS_associations_allFeatures_noMHC.txt.gz', \n",
    "          index=False, header=True, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
