{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate supplemental data of all significant integration analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:28.722094Z",
     "iopub.status.busy": "2021-09-07T20:40:28.720166Z",
     "iopub.status.idle": "2021-09-07T20:40:29.415678Z",
     "shell.execute_reply": "2021-09-07T20:40:29.413932Z"
    }
   },
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.431148Z",
     "iopub.status.busy": "2021-09-07T20:40:29.429255Z",
     "iopub.status.idle": "2021-09-07T20:40:29.545017Z",
     "shell.execute_reply": "2021-09-07T20:40:29.546310Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>pgc2_a1_same_as_our_counted</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28651576:A:G</td>\n",
       "      <td>rs34724414</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.275700</td>\n",
       "      <td>1.120000e-43</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28662914:G:A</td>\n",
       "      <td>rs13194504</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.275700</td>\n",
       "      <td>1.180000e-43</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28468237:TAAACATA:T</td>\n",
       "      <td>rs138024639</td>\n",
       "      <td>T</td>\n",
       "      <td>TAAACATA</td>\n",
       "      <td>0.770435</td>\n",
       "      <td>2.595000e-43</td>\n",
       "      <td>0.000917</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>True</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28594470:TG:T</td>\n",
       "      <td>rs34119086</td>\n",
       "      <td>T</td>\n",
       "      <td>TG</td>\n",
       "      <td>0.787730</td>\n",
       "      <td>5.296000e-42</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>True</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28608366:T:A</td>\n",
       "      <td>rs4947336</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>1.268900</td>\n",
       "      <td>7.480000e-42</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id  Symbol                variant_id         rsid A1  \\\n",
       "0  ENSG00000137185.11  ZSCAN9         chr6:28651576:A:G   rs34724414  A   \n",
       "1  ENSG00000137185.11  ZSCAN9         chr6:28662914:G:A   rs13194504  G   \n",
       "2  ENSG00000137185.11  ZSCAN9  chr6:28468237:TAAACATA:T  rs138024639  T   \n",
       "3  ENSG00000137185.11  ZSCAN9        chr6:28594470:TG:T   rs34119086  T   \n",
       "4  ENSG00000137185.11  ZSCAN9         chr6:28608366:T:A    rs4947336  T   \n",
       "\n",
       "         A2        OR             P  pval_nominal  adj.P.Val    logFC  \\\n",
       "0         G  1.275700  1.120000e-43      0.001482   0.022627 -0.07713   \n",
       "1         A  1.275700  1.180000e-43      0.001482   0.022627 -0.07713   \n",
       "2  TAAACATA  0.770435  2.595000e-43      0.000917   0.022627 -0.07713   \n",
       "3        TG  0.787730  5.296000e-42      0.001482   0.022627 -0.07713   \n",
       "4         A  1.268900  7.480000e-42      0.001482   0.022627 -0.07713   \n",
       "\n",
       "          t  eqtl_slope  de_dir  eqtl_gwas_dir agree_direction  \\\n",
       "0 -3.126621          -1      -1             -1             Yes   \n",
       "1 -3.126621          -1      -1             -1             Yes   \n",
       "2 -3.126621          -1      -1             -1             Yes   \n",
       "3 -3.126621          -1      -1             -1             Yes   \n",
       "4 -3.126621          -1      -1             -1             Yes   \n",
       "\n",
       "   pgc2_a1_same_as_our_counted Feature_Type  \n",
       "0                        False         Gene  \n",
       "1                        False         Gene  \n",
       "2                         True         Gene  \n",
       "3                         True         Gene  \n",
       "4                        False         Gene  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('../../_m/genes/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "genes['Feature_Type'] = 'Gene'\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.557686Z",
     "iopub.status.busy": "2021-09-07T20:40:29.555606Z",
     "iopub.status.idle": "2021-09-07T20:40:29.594127Z",
     "shell.execute_reply": "2021-09-07T20:40:29.595416Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>pgc2_a1_same_as_our_counted</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000418983.1</td>\n",
       "      <td>HCG4</td>\n",
       "      <td>chr6:29854872:C:T</td>\n",
       "      <td>rs2517857</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.166700</td>\n",
       "      <td>1.170000e-22</td>\n",
       "      <td>7.560360e-05</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.251395</td>\n",
       "      <td>3.568189</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000418983.1</td>\n",
       "      <td>HCG4</td>\n",
       "      <td>chr6:29638984:G:A</td>\n",
       "      <td>rs3095268</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.146100</td>\n",
       "      <td>4.090000e-22</td>\n",
       "      <td>1.498240e-04</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.251395</td>\n",
       "      <td>3.568189</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000418983.1</td>\n",
       "      <td>HCG4</td>\n",
       "      <td>chr6:29956359:AG:A</td>\n",
       "      <td>rs5875221</td>\n",
       "      <td>A</td>\n",
       "      <td>AG</td>\n",
       "      <td>0.859074</td>\n",
       "      <td>1.624000e-21</td>\n",
       "      <td>3.937150e-05</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.251395</td>\n",
       "      <td>3.568189</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>True</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000418983.1</td>\n",
       "      <td>HCG4</td>\n",
       "      <td>chr6:29852420:A:G</td>\n",
       "      <td>rs9258609</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.159100</td>\n",
       "      <td>1.860000e-21</td>\n",
       "      <td>1.103540e-04</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.251395</td>\n",
       "      <td>3.568189</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000418983.1</td>\n",
       "      <td>HCG4</td>\n",
       "      <td>chr6:29850949:A:G</td>\n",
       "      <td>rs2523761</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.156400</td>\n",
       "      <td>2.000000e-21</td>\n",
       "      <td>1.462340e-09</td>\n",
       "      <td>0.028722</td>\n",
       "      <td>0.251395</td>\n",
       "      <td>3.568189</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Transcript</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             gene_id Symbol          variant_id       rsid A1  A2        OR  \\\n",
       "0  ENST00000418983.1   HCG4   chr6:29854872:C:T  rs2517857  C   T  1.166700   \n",
       "1  ENST00000418983.1   HCG4   chr6:29638984:G:A  rs3095268  G   A  1.146100   \n",
       "2  ENST00000418983.1   HCG4  chr6:29956359:AG:A  rs5875221  A  AG  0.859074   \n",
       "3  ENST00000418983.1   HCG4   chr6:29852420:A:G  rs9258609  A   G  1.159100   \n",
       "4  ENST00000418983.1   HCG4   chr6:29850949:A:G  rs2523761  A   G  1.156400   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val     logFC         t  eqtl_slope  \\\n",
       "0  1.170000e-22  7.560360e-05   0.028722  0.251395  3.568189           1   \n",
       "1  4.090000e-22  1.498240e-04   0.028722  0.251395  3.568189           1   \n",
       "2  1.624000e-21  3.937150e-05   0.028722  0.251395  3.568189           1   \n",
       "3  1.860000e-21  1.103540e-04   0.028722  0.251395  3.568189           1   \n",
       "4  2.000000e-21  1.462340e-09   0.028722  0.251395  3.568189           1   \n",
       "\n",
       "   de_dir  eqtl_gwas_dir agree_direction  pgc2_a1_same_as_our_counted  \\\n",
       "0       1              1             Yes                        False   \n",
       "1       1              1             Yes                        False   \n",
       "2       1              1             Yes                         True   \n",
       "3       1              1             Yes                        False   \n",
       "4       1              1             Yes                        False   \n",
       "\n",
       "  Feature_Type  \n",
       "0   Transcript  \n",
       "1   Transcript  \n",
       "2   Transcript  \n",
       "3   Transcript  \n",
       "4   Transcript  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans = pd.read_csv('../../_m/transcripts/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID', 'gene_name': 'Symbol'})\n",
    "trans['Feature_Type'] = 'Transcript'\n",
    "trans.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.609765Z",
     "iopub.status.busy": "2021-09-07T20:40:29.604112Z",
     "iopub.status.idle": "2021-09-07T20:40:29.663416Z",
     "shell.execute_reply": "2021-09-07T20:40:29.661678Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>pgc2_a1_same_as_our_counted</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>e378566</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28744470:A:G</td>\n",
       "      <td>rs1233578</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.22280</td>\n",
       "      <td>4.320000e-44</td>\n",
       "      <td>7.294750e-10</td>\n",
       "      <td>0.018675</td>\n",
       "      <td>-0.084735</td>\n",
       "      <td>-3.408383</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>e378558</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28744470:A:G</td>\n",
       "      <td>rs1233578</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.22280</td>\n",
       "      <td>4.320000e-44</td>\n",
       "      <td>1.080300e-09</td>\n",
       "      <td>0.019079</td>\n",
       "      <td>-0.084425</td>\n",
       "      <td>-3.399939</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>e378566</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28716406:T:G</td>\n",
       "      <td>rs9295778</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.82449</td>\n",
       "      <td>4.520000e-42</td>\n",
       "      <td>6.790160e-10</td>\n",
       "      <td>0.018675</td>\n",
       "      <td>-0.084735</td>\n",
       "      <td>-3.408383</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>e378558</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28716406:T:G</td>\n",
       "      <td>rs9295778</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.82449</td>\n",
       "      <td>4.520000e-42</td>\n",
       "      <td>8.899240e-10</td>\n",
       "      <td>0.019079</td>\n",
       "      <td>-0.084425</td>\n",
       "      <td>-3.399939</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>e378566</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28721895:C:A</td>\n",
       "      <td>rs9468413</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.82511</td>\n",
       "      <td>6.860000e-42</td>\n",
       "      <td>6.790160e-10</td>\n",
       "      <td>0.018675</td>\n",
       "      <td>-0.084735</td>\n",
       "      <td>-3.408383</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Exon</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_id   Symbol         variant_id       rsid A1 A2       OR  \\\n",
       "0  e378566  ZSCAN26  chr6:28744470:A:G  rs1233578  A  G  1.22280   \n",
       "1  e378558  ZSCAN26  chr6:28744470:A:G  rs1233578  A  G  1.22280   \n",
       "2  e378566  ZSCAN26  chr6:28716406:T:G  rs9295778  T  G  0.82449   \n",
       "3  e378558  ZSCAN26  chr6:28716406:T:G  rs9295778  T  G  0.82449   \n",
       "4  e378566  ZSCAN26  chr6:28721895:C:A  rs9468413  C  A  0.82511   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val     logFC         t  eqtl_slope  \\\n",
       "0  4.320000e-44  7.294750e-10   0.018675 -0.084735 -3.408383          -1   \n",
       "1  4.320000e-44  1.080300e-09   0.019079 -0.084425 -3.399939          -1   \n",
       "2  4.520000e-42  6.790160e-10   0.018675 -0.084735 -3.408383          -1   \n",
       "3  4.520000e-42  8.899240e-10   0.019079 -0.084425 -3.399939          -1   \n",
       "4  6.860000e-42  6.790160e-10   0.018675 -0.084735 -3.408383          -1   \n",
       "\n",
       "   de_dir  eqtl_gwas_dir agree_direction  pgc2_a1_same_as_our_counted  \\\n",
       "0      -1             -1             Yes                        False   \n",
       "1      -1             -1             Yes                        False   \n",
       "2      -1             -1             Yes                        False   \n",
       "3      -1             -1             Yes                        False   \n",
       "4      -1             -1             Yes                        False   \n",
       "\n",
       "  Feature_Type  \n",
       "0         Exon  \n",
       "1         Exon  \n",
       "2         Exon  \n",
       "3         Exon  \n",
       "4         Exon  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons = pd.read_csv('../../_m/exons/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "exons['Feature_Type'] = 'Exon'\n",
    "exons.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.674040Z",
     "iopub.status.busy": "2021-09-07T20:40:29.672191Z",
     "iopub.status.idle": "2021-09-07T20:40:29.739690Z",
     "shell.execute_reply": "2021-09-07T20:40:29.738090Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>pgc2_a1_same_as_our_counted</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr6:28271935-28272021(+)</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28744470:A:G</td>\n",
       "      <td>rs1233578</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.222800</td>\n",
       "      <td>4.320000e-44</td>\n",
       "      <td>9.439080e-11</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.26177</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6:28271935-28272021(+)</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28468237:TAAACATA:T</td>\n",
       "      <td>rs138024639</td>\n",
       "      <td>T</td>\n",
       "      <td>TAAACATA</td>\n",
       "      <td>0.770435</td>\n",
       "      <td>2.595000e-43</td>\n",
       "      <td>2.249520e-04</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.26177</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>True</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr6:28271935-28272021(+)</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28716406:T:G</td>\n",
       "      <td>rs9295778</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.824490</td>\n",
       "      <td>4.520000e-42</td>\n",
       "      <td>1.737900e-10</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.26177</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr6:28271935-28272021(+)</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28721895:C:A</td>\n",
       "      <td>rs9468413</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.825110</td>\n",
       "      <td>6.860000e-42</td>\n",
       "      <td>1.737900e-10</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.26177</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr6:28271935-28272021(+)</td>\n",
       "      <td>ZSCAN26</td>\n",
       "      <td>chr6:28426903:C:T</td>\n",
       "      <td>rs13201681</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1.265200</td>\n",
       "      <td>1.320000e-41</td>\n",
       "      <td>2.249520e-04</td>\n",
       "      <td>0.018932</td>\n",
       "      <td>-0.26177</td>\n",
       "      <td>-3.493012</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Junction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     gene_id   Symbol                variant_id         rsid  \\\n",
       "0  chr6:28271935-28272021(+)  ZSCAN26         chr6:28744470:A:G    rs1233578   \n",
       "1  chr6:28271935-28272021(+)  ZSCAN26  chr6:28468237:TAAACATA:T  rs138024639   \n",
       "2  chr6:28271935-28272021(+)  ZSCAN26         chr6:28716406:T:G    rs9295778   \n",
       "3  chr6:28271935-28272021(+)  ZSCAN26         chr6:28721895:C:A    rs9468413   \n",
       "4  chr6:28271935-28272021(+)  ZSCAN26         chr6:28426903:C:T   rs13201681   \n",
       "\n",
       "  A1        A2        OR             P  pval_nominal  adj.P.Val    logFC  \\\n",
       "0  A         G  1.222800  4.320000e-44  9.439080e-11   0.018932 -0.26177   \n",
       "1  T  TAAACATA  0.770435  2.595000e-43  2.249520e-04   0.018932 -0.26177   \n",
       "2  T         G  0.824490  4.520000e-42  1.737900e-10   0.018932 -0.26177   \n",
       "3  C         A  0.825110  6.860000e-42  1.737900e-10   0.018932 -0.26177   \n",
       "4  C         T  1.265200  1.320000e-41  2.249520e-04   0.018932 -0.26177   \n",
       "\n",
       "          t  eqtl_slope  de_dir  eqtl_gwas_dir agree_direction  \\\n",
       "0 -3.493012          -1      -1             -1             Yes   \n",
       "1 -3.493012          -1      -1             -1             Yes   \n",
       "2 -3.493012          -1      -1             -1             Yes   \n",
       "3 -3.493012          -1      -1             -1             Yes   \n",
       "4 -3.493012          -1      -1             -1             Yes   \n",
       "\n",
       "   pgc2_a1_same_as_our_counted Feature_Type  \n",
       "0                        False     Junction  \n",
       "1                         True     Junction  \n",
       "2                        False     Junction  \n",
       "3                        False     Junction  \n",
       "4                        False     Junction  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "juncs = pd.read_csv('../../_m/junctions/integration_by_symbol.txt', sep='\\t').rename(columns={'gene': 'Feature_ID'})\n",
    "juncs['Feature_Type'] = 'Junction'\n",
    "juncs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.749456Z",
     "iopub.status.busy": "2021-09-07T20:40:29.747759Z",
     "iopub.status.idle": "2021-09-07T20:40:29.783021Z",
     "shell.execute_reply": "2021-09-07T20:40:29.784236Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(13169, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>rsid</th>\n",
       "      <th>A1</th>\n",
       "      <th>A2</th>\n",
       "      <th>OR</th>\n",
       "      <th>P</th>\n",
       "      <th>pval_nominal</th>\n",
       "      <th>adj.P.Val</th>\n",
       "      <th>logFC</th>\n",
       "      <th>t</th>\n",
       "      <th>eqtl_slope</th>\n",
       "      <th>de_dir</th>\n",
       "      <th>eqtl_gwas_dir</th>\n",
       "      <th>agree_direction</th>\n",
       "      <th>pgc2_a1_same_as_our_counted</th>\n",
       "      <th>Feature_Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28651576:A:G</td>\n",
       "      <td>rs34724414</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.2757</td>\n",
       "      <td>1.120000e-43</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000137185.11</td>\n",
       "      <td>ZSCAN9</td>\n",
       "      <td>chr6:28662914:G:A</td>\n",
       "      <td>rs13194504</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.2757</td>\n",
       "      <td>1.180000e-43</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.022627</td>\n",
       "      <td>-0.07713</td>\n",
       "      <td>-3.126621</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>Yes</td>\n",
       "      <td>False</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene_id  Symbol         variant_id        rsid A1 A2      OR  \\\n",
       "0  ENSG00000137185.11  ZSCAN9  chr6:28651576:A:G  rs34724414  A  G  1.2757   \n",
       "1  ENSG00000137185.11  ZSCAN9  chr6:28662914:G:A  rs13194504  G  A  1.2757   \n",
       "\n",
       "              P  pval_nominal  adj.P.Val    logFC         t  eqtl_slope  \\\n",
       "0  1.120000e-43      0.001482   0.022627 -0.07713 -3.126621          -1   \n",
       "1  1.180000e-43      0.001482   0.022627 -0.07713 -3.126621          -1   \n",
       "\n",
       "   de_dir  eqtl_gwas_dir agree_direction  pgc2_a1_same_as_our_counted  \\\n",
       "0      -1             -1             Yes                        False   \n",
       "1      -1             -1             Yes                        False   \n",
       "\n",
       "  Feature_Type  \n",
       "0         Gene  \n",
       "1         Gene  "
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([genes, trans, exons, juncs], axis=0, sort=False)\n",
    "print(df.shape)\n",
    "df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-07T20:40:29.794023Z",
     "iopub.status.busy": "2021-09-07T20:40:29.792294Z",
     "iopub.status.idle": "2021-09-07T20:40:30.027148Z",
     "shell.execute_reply": "2021-09-07T20:40:30.025521Z"
    }
   },
   "outputs": [],
   "source": [
    "df.to_csv('BrainSeq_Phase3_Caudate_GWAS_DE_eQTL_Integration.txt.gz', index=False, header=True, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
